

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>moments.LD.Parsing &mdash; moments 1.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../_static/jupyter-sphinx.css" />
      <link rel="stylesheet" type="text/css" href="../../../_static/thebelab.css" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=fc837d61"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../../../_static/thebelab-helper.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            moments
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">SFS</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../sfs/sfs.html">The Site Frequency Spectrum</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../sfs/parsing.html">Parsing the SFS from a VCF file</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../sfs/inference.html">SFS Inference</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Linkage disequilibrium</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../ld/ld.html">Multi-population LD statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ld/parsing.html">Parsing LD statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../ld/inference.html">Inferring demography with LD</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Extensions</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../extensions/demes.html">Specifying models with <code class="docutils literal notranslate"><span class="pre">demes</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../extensions/two_locus.html">Two-locus frequency spectrum</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../extensions/triallele.html">Triallele frequency spectrum</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/diversity.html">Demography and genetic diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/dfe.html">DFE inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/recombination.html">Linkage disequilibrium and recombination</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/two_locus_selection.html">Selection at two loci</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Data types and functions</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../api/api_moments.html">API for site frequency spectra</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api/api_ld.html">API for linkage disequilibrium</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api/api_demes.html">API for <code class="docutils literal notranslate"><span class="pre">demes</span></code> functions</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">moments</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">moments.LD.Parsing</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for moments.LD.Parsing</h1><div class="highlight"><pre>
<span></span><span class="n">_imported_h5py</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">_imported_allel</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">_imported_pandas</span> <span class="o">=</span> <span class="mi">0</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">datetime</span><span class="w"> </span><span class="kn">import</span> <span class="n">datetime</span>


<span class="k">def</span><span class="w"> </span><span class="nf">current_time</span><span class="p">():</span>
    <span class="k">return</span> <span class="s2">&quot; [&quot;</span> <span class="o">+</span> <span class="n">datetime</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">(),</span> <span class="s2">&quot;%Y-%m-</span><span class="si">%d</span><span class="s2"> %H:%M:%S&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;]&quot;</span>


<span class="k">try</span><span class="p">:</span>
    <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;NUMEXPR_MAX_THREADS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;272&quot;</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">allel</span>

    <span class="n">_imported_allel</span> <span class="o">=</span> <span class="mi">1</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="k">pass</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">h5py</span>

    <span class="n">_imported_h5py</span> <span class="o">=</span> <span class="mi">1</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="k">pass</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span>

    <span class="n">_imported_pandas</span> <span class="o">=</span> <span class="mi">1</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="k">pass</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">.</span><span class="w"> </span><span class="kn">import</span> <span class="n">Util</span>


<span class="k">def</span><span class="w"> </span><span class="nf">check_imports</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">_imported_allel</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ImportError</span><span class="p">(</span><span class="s2">&quot;Failed to import scikit-allel which is needed for Parsing&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">_imported_h5py</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ImportError</span><span class="p">(</span><span class="s2">&quot;Failed to import h5py which is needed for Parsing&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">_imported_pandas</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ImportError</span><span class="p">(</span><span class="s2">&quot;Failed to import pandas which is needed for Parsing&quot;</span><span class="p">)</span>


<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">collections</span><span class="w"> </span><span class="kn">import</span> <span class="n">Counter</span><span class="p">,</span> <span class="n">defaultdict</span>

<span class="c1"># from . import stats_from_genotype_counts as sgc</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">.</span><span class="w"> </span><span class="kn">import</span> <span class="n">stats_from_haplotype_counts</span> <span class="k">as</span> <span class="n">shc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">itertools</span>

<span class="n">ld_extensions</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">genotype_calculations</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">gcs</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">genotype_calculations_multipop</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">gcs_mp</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">sparse_tallying</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">spt</span>

    <span class="n">ld_extensions</span> <span class="o">=</span> <span class="mi">1</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="k">pass</span>

<span class="c1"># turn off UserWarnings from allel</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>

<span class="k">if</span> <span class="n">_imported_allel</span><span class="p">:</span>
    <span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s2">&quot;&#39;GQ&#39; FORMAT&quot;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">UserWarning</span>
    <span class="p">)</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_load_h5</span><span class="p">(</span><span class="n">vcf_file</span><span class="p">,</span> <span class="n">report</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="n">check_imports</span><span class="p">()</span>
    <span class="c1">## open the h5 callset, create if doesn&#39;t exist</span>
    <span class="c1">## note that if the h5 file exists, but isn&#39;t properly written,</span>
    <span class="c1">## you will need to delete and recreate</span>
    <span class="c1">## saves h5 callset as same name and path,</span>
    <span class="c1">## but with h5 extension instead of vcf or vcf.gz</span>
    <span class="n">h5_file_path</span> <span class="o">=</span> <span class="n">vcf_file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.vcf&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;.h5&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">callset</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">h5_file_path</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="p">(</span><span class="ne">OSError</span><span class="p">,</span> <span class="ne">IOError</span><span class="p">):</span>  <span class="c1"># IOError merged into OSError in python 3</span>
        <span class="k">if</span> <span class="n">report</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">current_time</span><span class="p">(),</span> <span class="s2">&quot;creating and saving h5 file&quot;</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
        <span class="n">allel</span><span class="o">.</span><span class="n">vcf_to_hdf5</span><span class="p">(</span>
            <span class="n">vcf_file</span><span class="p">,</span>
            <span class="n">h5_file_path</span><span class="p">,</span>
            <span class="n">fields</span><span class="o">=</span><span class="s2">&quot;*&quot;</span><span class="p">,</span>
            <span class="n">exclude_fields</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;calldata/GQ&quot;</span><span class="p">],</span>
            <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">callset</span> <span class="o">=</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">h5_file_path</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">callset</span>


<div class="viewcode-block" id="get_genotypes">
<a class="viewcode-back" href="../../../api/api_ld.html#moments.LD.Parsing.get_genotypes">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_genotypes</span><span class="p">(</span>
    <span class="n">vcf_file</span><span class="p">,</span> <span class="n">bed_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">chromosome</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">min_bp</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_h5</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">report</span><span class="o">=</span><span class="kc">True</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a vcf file, we extract the biallelic SNP genotypes. If bed_file is None,</span>
<span class="sd">    we use all valid variants. Otherwise we filter genotypes by the given bed file.</span>
<span class="sd">    If chromosome is given, filters to keep snps only in given chrom (useful for vcfs</span>
<span class="sd">    spanning multiple chromosomes).</span>

<span class="sd">    If use_h5 is True, we try to load the h5 file, which has the same path/name as</span>
<span class="sd">    vcf_file, but with {fname}.h5 instead of {fname}.vcf or {fname}.vcf.gz. If the h5</span>
<span class="sd">    file does not exist, we create it and save it as {fname}.h5</span>

<span class="sd">    Returns (biallelic positions, biallelic genotypes, biallelic allele counts,</span>
<span class="sd">    sampled ids).</span>

<span class="sd">    :param vcf_file: A VCF-formatted file.</span>
<span class="sd">    :type vcf_file: str</span>
<span class="sd">    :param bed_file: A bed file specifying regions to compute statistics from. The</span>
<span class="sd">        chromosome name formatting must match the chromosome name formatting of the</span>
<span class="sd">        input VCF (i.e., both carry the leading &quot;chr&quot; or both omit it).</span>
<span class="sd">    :type bed_file: str, optional</span>
<span class="sd">    :param min_bp: only used with bed file, filters out features that are smaller</span>
<span class="sd">        than ``min_bp``.</span>
<span class="sd">    :type min_bp: int, optional</span>
<span class="sd">    :param chromosome: Chromosome to compute LD statistics from.</span>
<span class="sd">    :type chromosome: int or str, optional</span>
<span class="sd">    :param use_h5: If use_h5 is True, we try to load the h5 file, which has the</span>
<span class="sd">        same path/name as vcf_file, but with .h5 instead of .vcf or .vcf.gz extension.</span>
<span class="sd">        If the h5 file does not exist, we create it and save it with .h5 extension.</span>
<span class="sd">        Defaults to True.</span>
<span class="sd">    :type use_h5: bool, optional</span>
<span class="sd">    :param report: Prints progress updates if True, silent otherwise. Defaults to True.</span>
<span class="sd">    :type report: bool, optional</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">check_imports</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">use_h5</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">callset</span> <span class="o">=</span> <span class="n">_load_h5</span><span class="p">(</span><span class="n">vcf_file</span><span class="p">,</span> <span class="n">report</span><span class="o">=</span><span class="n">report</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">callset</span> <span class="o">=</span> <span class="n">allel</span><span class="o">.</span><span class="n">read_vcf</span><span class="p">(</span><span class="n">vcf_file</span><span class="p">)</span>

    <span class="n">all_genotypes</span> <span class="o">=</span> <span class="n">allel</span><span class="o">.</span><span class="n">GenotypeChunkedArray</span><span class="p">(</span><span class="n">callset</span><span class="p">[</span><span class="s2">&quot;calldata/GT&quot;</span><span class="p">])</span>
    <span class="n">all_positions</span> <span class="o">=</span> <span class="n">callset</span><span class="p">[</span><span class="s2">&quot;variants/POS&quot;</span><span class="p">][:]</span>

    <span class="c1"># if there are multiple chromosomes in the VCF, we require a specified chromosome</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">all_chromosomes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">c</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">callset</span><span class="p">[</span><span class="s2">&quot;variants/CHROM&quot;</span><span class="p">][:]])</span>
    <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
        <span class="n">all_chromosomes</span> <span class="o">=</span> <span class="n">callset</span><span class="p">[</span><span class="s2">&quot;variants/CHROM&quot;</span><span class="p">][:]</span>

    <span class="n">num_chromosomes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">all_chromosomes</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">num_chromosomes</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">chromosome</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">chromosome</span> <span class="o">=</span> <span class="n">all_chromosomes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">num_chromosomes</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">chromosome</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;The input VCF has more than one chromosome present. &quot;</span>
                <span class="s2">&quot;The `chromosome` must be specified.&quot;</span>
            <span class="p">)</span>

    <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">all_chromosomes</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;The specified chromosome, </span><span class="si">{</span><span class="n">chromosome</span><span class="si">}</span><span class="s2">, was not found among &quot;</span>
            <span class="s2">&quot;sites in the VCF. Double check the input chromosome name.&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">num_chromosomes</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">in_chromosome</span> <span class="o">=</span> <span class="n">all_chromosomes</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)</span>
        <span class="n">all_positions</span> <span class="o">=</span> <span class="n">all_positions</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">in_chromosome</span><span class="p">)</span>
        <span class="n">all_genotypes</span> <span class="o">=</span> <span class="n">all_genotypes</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">in_chromosome</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">bed_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">bed_file_data</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">bed_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">bed_chromosomes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">bed_file_data</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">bed_lefts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">bed_file_data</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">bed_rights</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">bed_file_data</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>

        <span class="c1"># only keep rows of bed file that match our chromosome</span>
        <span class="n">bed_chrom_filter</span> <span class="o">=</span> <span class="n">bed_chromosomes</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)</span>
        <span class="n">bed_chromosomes</span> <span class="o">=</span> <span class="n">bed_chromosomes</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">bed_chrom_filter</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">bed_chromosomes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;No regions of the bed file matched the input chromosome. &quot;</span>
                <span class="s2">&quot;Check that chromosome names match between bed file and VCF.&quot;</span>
            <span class="p">)</span>
        <span class="n">bed_lefts</span> <span class="o">=</span> <span class="n">bed_lefts</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">bed_chrom_filter</span><span class="p">)</span>
        <span class="n">bed_rights</span> <span class="o">=</span> <span class="n">bed_rights</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">bed_chrom_filter</span><span class="p">)</span>

        <span class="c1"># if a minimum length of feature is specified, only keep long enough features</span>
        <span class="k">if</span> <span class="n">min_bp</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">bp_filter</span> <span class="o">=</span> <span class="n">bed_rights</span> <span class="o">-</span> <span class="n">bed_lefts</span> <span class="o">&gt;=</span> <span class="n">min_bp</span>
            <span class="n">bed_lefts</span> <span class="o">=</span> <span class="n">bed_lefts</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">bp_filter</span><span class="p">)</span>
            <span class="n">bed_rights</span> <span class="o">=</span> <span class="n">bed_rights</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">bp_filter</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">bed_lefts</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;No features in bed file were at least </span><span class="si">{min_bp}</span><span class="s2"> in length.&quot;</span>
                <span class="p">)</span>

        <span class="n">in_bed</span> <span class="o">=</span> <span class="n">all_positions</span> <span class="o">&lt;</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">left</span><span class="p">,</span> <span class="n">right</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">bed_lefts</span><span class="p">,</span> <span class="n">bed_rights</span><span class="p">):</span>
            <span class="n">in_bed</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span>
                <span class="n">in_bed</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">all_positions</span> <span class="o">&gt;=</span> <span class="n">left</span><span class="p">,</span> <span class="n">all_positions</span> <span class="o">&lt;</span> <span class="n">right</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="n">all_positions</span> <span class="o">=</span> <span class="n">all_positions</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">in_bed</span><span class="p">)</span>
        <span class="n">all_genotypes</span> <span class="o">=</span> <span class="n">all_genotypes</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">in_bed</span><span class="p">)</span>

    <span class="n">all_genotypes_012</span> <span class="o">=</span> <span class="n">all_genotypes</span><span class="o">.</span><span class="n">to_n_alt</span><span class="p">(</span><span class="n">fill</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># count alleles and only keep biallelic positions</span>
    <span class="n">allele_counts</span> <span class="o">=</span> <span class="n">all_genotypes</span><span class="o">.</span><span class="n">count_alleles</span><span class="p">()</span>
    <span class="n">is_biallelic</span> <span class="o">=</span> <span class="n">allele_counts</span><span class="o">.</span><span class="n">is_biallelic_01</span><span class="p">()</span>
    <span class="n">biallelic_positions</span> <span class="o">=</span> <span class="n">all_positions</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">is_biallelic</span><span class="p">)</span>

    <span class="n">biallelic_genotypes_012</span> <span class="o">=</span> <span class="n">all_genotypes_012</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">is_biallelic</span><span class="p">)</span>

    <span class="n">biallelic_allele_counts</span> <span class="o">=</span> <span class="n">allele_counts</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">is_biallelic</span><span class="p">)</span>
    <span class="n">biallelic_genotypes</span> <span class="o">=</span> <span class="n">all_genotypes</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">is_biallelic</span><span class="p">)</span>

    <span class="n">relevant_column</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">False</span><span class="p">]</span> <span class="o">*</span> <span class="n">biallelic_allele_counts</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">relevant_column</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">biallelic_allele_counts</span> <span class="o">=</span> <span class="n">biallelic_allele_counts</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">relevant_column</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># protect against variable encoding of sample id strings</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">sample_ids</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">sid</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span> <span class="k">for</span> <span class="n">sid</span> <span class="ow">in</span> <span class="n">callset</span><span class="p">[</span><span class="s2">&quot;samples&quot;</span><span class="p">]])</span>
    <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
        <span class="n">sample_ids</span> <span class="o">=</span> <span class="n">callset</span><span class="p">[</span><span class="s2">&quot;samples&quot;</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">biallelic_positions</span><span class="p">,</span> <span class="n">biallelic_genotypes</span><span class="p">,</span> <span class="n">biallelic_allele_counts</span><span class="p">,</span> <span class="n">sample_ids</span></div>



<span class="k">def</span><span class="w"> </span><span class="nf">_assign_r_pos</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">rec_map</span><span class="p">):</span>
    <span class="n">rs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">ii</span><span class="p">,</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">positions</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">rec_map</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
            <span class="n">rs</span><span class="p">[</span><span class="n">ii</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">rec_map</span><span class="p">[</span><span class="mi">1</span><span class="p">])[</span><span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">pos</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">rec_map</span><span class="p">[</span><span class="mi">0</span><span class="p">]))[</span><span class="mi">0</span><span class="p">]]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1">## for now, if outside rec map, assign to nearest point,</span>
            <span class="c1">## but later want to drop these positions</span>
            <span class="k">if</span> <span class="n">pos</span> <span class="o">&lt;</span> <span class="n">rec_map</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                <span class="n">rs</span><span class="p">[</span><span class="n">ii</span><span class="p">]</span> <span class="o">=</span> <span class="n">rec_map</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">pos</span> <span class="o">&gt;</span> <span class="n">rec_map</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
                <span class="n">rs</span><span class="p">[</span><span class="n">ii</span><span class="p">]</span> <span class="o">=</span> <span class="n">rec_map</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">map_ii</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">pos</span> <span class="o">&gt;=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">rec_map</span><span class="p">[</span><span class="mi">0</span><span class="p">]))[</span><span class="mi">0</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">l</span> <span class="o">=</span> <span class="n">rec_map</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">map_ii</span><span class="p">]</span>
                <span class="n">r</span> <span class="o">=</span> <span class="n">rec_map</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">map_ii</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
                <span class="n">v_l</span> <span class="o">=</span> <span class="n">rec_map</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">map_ii</span><span class="p">]</span>
                <span class="n">v_r</span> <span class="o">=</span> <span class="n">rec_map</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">map_ii</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
                <span class="n">rs</span><span class="p">[</span><span class="n">ii</span><span class="p">]</span> <span class="o">=</span> <span class="n">v_l</span> <span class="o">+</span> <span class="p">(</span><span class="n">v_r</span> <span class="o">-</span> <span class="n">v_l</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">pos</span> <span class="o">-</span> <span class="n">l</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">r</span> <span class="o">-</span> <span class="n">l</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">rs</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_assign_recombination_rates</span><span class="p">(</span>
    <span class="n">positions</span><span class="p">,</span> <span class="n">map_file</span><span class="p">,</span> <span class="n">map_name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">map_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cM</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">report</span><span class="o">=</span><span class="kc">True</span>
<span class="p">):</span>
    <span class="c1">## map_sep is now deprecated. we split by any white space</span>
    <span class="k">if</span> <span class="n">map_file</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Need to pass a recombination map file. Otherwise can bin by physical distance.&quot;</span>
        <span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">rec_map</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">map_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">)</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Error loading recombination map.&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>

    <span class="k">if</span> <span class="s2">&quot;Position(bp)&quot;</span> <span class="ow">in</span> <span class="n">rec_map</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">map_positions</span> <span class="o">=</span> <span class="n">rec_map</span><span class="p">[</span><span class="s2">&quot;Position(bp)&quot;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">map_positions</span> <span class="o">=</span> <span class="n">rec_map</span><span class="p">[</span><span class="n">rec_map</span><span class="o">.</span><span class="n">keys</span><span class="p">()[</span><span class="mi">0</span><span class="p">]]</span>
    <span class="k">if</span> <span class="n">map_name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># we use the first map column</span>
        <span class="k">if</span> <span class="n">report</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">current_time</span><span class="p">(),</span> <span class="s2">&quot;No recombination map name given, trying Map(cM).&quot;</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
        <span class="n">map_name</span> <span class="o">=</span> <span class="s2">&quot;Map(cM)&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">map_values</span> <span class="o">=</span> <span class="n">rec_map</span><span class="p">[</span><span class="n">map_name</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="n">current_time</span><span class="p">(),</span>
            <span class="sa">f</span><span class="s2">&quot;WARNING: map_name did not match </span><span class="si">{</span><span class="n">map_name</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="s2">&quot; in recombination map file. Using the first column.&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">map_values</span> <span class="o">=</span> <span class="n">rec_map</span><span class="p">[</span><span class="n">rec_map</span><span class="o">.</span><span class="n">keys</span><span class="p">()[</span><span class="mi">1</span><span class="p">]]</span>

    <span class="c1"># for positions sticking out the end of the map,</span>
    <span class="c1"># they take the value of the closest position</span>
    <span class="c1"># ideally, you&#39;d filter these out</span>

    <span class="k">if</span> <span class="n">cM</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">map_values</span> <span class="o">/=</span> <span class="mi">100</span>

    <span class="n">pos_rs</span> <span class="o">=</span> <span class="n">_assign_r_pos</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="p">[</span><span class="n">map_positions</span><span class="p">,</span> <span class="n">map_values</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">pos_rs</span>


<span class="c1">## We store a Genotype matrix, which has size L x n, in sparse format</span>
<span class="c1">## Indexed by position in Genotype array (0...L-1)</span>
<span class="c1">## and G_dict[i] = {1: J, 2: K}</span>
<span class="c1">## where J and K are sets of the diploid individual indices that</span>
<span class="c1">## have genotypes 1 or 2</span>
<span class="c1">## If there is any missing data, we also store set of individuals with -1&#39;s</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_sparsify_genotype_matrix</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
    <span class="n">G_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">missing</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">missing</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">G</span><span class="p">)):</span>
        <span class="n">G_dict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="mi">1</span><span class="p">:</span> <span class="nb">set</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">G</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]),</span>
            <span class="mi">2</span><span class="p">:</span> <span class="nb">set</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">G</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">==</span> <span class="mi">2</span><span class="p">)[</span><span class="mi">0</span><span class="p">]),</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="n">missing</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">G_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">G</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">G_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">G_dict</span><span class="p">,</span> <span class="n">missing</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_sparsify_haplotype_matrix</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
    <span class="n">G_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">missing</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">missing</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">G</span><span class="p">)):</span>
        <span class="n">G_dict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="mi">1</span><span class="p">:</span> <span class="nb">set</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">G</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]),</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="n">missing</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">G_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">G</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="p">:]</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">G_dict</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">G_dict</span><span class="p">,</span> <span class="n">missing</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_check_valid_genotype_matrix</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">genotypes</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">genotypes</span><span class="p">:</span>
        <span class="c1"># 0, 1, 2 for genotype values, -1 for missing data</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="n">G</span> <span class="o">==</span> <span class="mi">1</span><span class="p">),</span> <span class="n">G</span> <span class="o">==</span> <span class="mi">2</span><span class="p">),</span> <span class="n">G</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
        <span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Genotype matrix must have values of 0, 1, or 2, or -1 for missing data&quot;</span>
            <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># haplotypes: 0, 1, -1 for missing data</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="n">G</span> <span class="o">==</span> <span class="mi">1</span><span class="p">),</span> <span class="n">G</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Haplotype matrix must have values of 0 or 1, or -1 for missing data&quot;</span>
            <span class="p">)</span>

    <span class="n">L</span><span class="p">,</span> <span class="n">n</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">L</span> <span class="o">&gt;</span> <span class="mi">46340</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Genotype matrix is too large, consider parallelizing LD calc&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="compute_pairwise_stats">
<a class="viewcode-back" href="../../../api/api_ld.html#moments.LD.Parsing.compute_pairwise_stats">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">compute_pairwise_stats</span><span class="p">(</span><span class="n">Gs</span><span class="p">,</span> <span class="n">genotypes</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes :math:`D^2`, :math:`Dz`, :math:`\\pi_2`, and :math:`D` for every</span>
<span class="sd">    pair of loci within a block of SNPs, coded as a genotype matrix.</span>

<span class="sd">    :param Gs: A genotype matrix, of size L-by-n, where</span>
<span class="sd">        L is the number of loci and n is the sample size.</span>
<span class="sd">        Missing data is encoded as -1.</span>
<span class="sd">    :param genotypes: If True, use 0, 1, 2 genotypes. If False,</span>
<span class="sd">        use 0, 1 phased haplotypes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">ld_extensions</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Need to build LD cython extensions. &quot;</span>
            <span class="s2">&quot;Install moments with the flag `--ld_extensions`&quot;</span>
        <span class="p">)</span>

    <span class="n">_check_valid_genotype_matrix</span><span class="p">(</span><span class="n">Gs</span><span class="p">,</span> <span class="n">genotypes</span><span class="p">)</span>

    <span class="n">L</span><span class="p">,</span> <span class="n">n</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">Gs</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">genotypes</span><span class="p">:</span>
        <span class="n">G_dict</span><span class="p">,</span> <span class="n">any_missing</span> <span class="o">=</span> <span class="n">_sparsify_genotype_matrix</span><span class="p">(</span><span class="n">Gs</span><span class="p">)</span>
        <span class="n">Counts</span> <span class="o">=</span> <span class="n">spt</span><span class="o">.</span><span class="n">count_genotypes_sparse</span><span class="p">(</span><span class="n">G_dict</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">missing</span><span class="o">=</span><span class="n">any_missing</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">G_dict</span><span class="p">,</span> <span class="n">any_missing</span> <span class="o">=</span> <span class="n">_sparsify_haplotype_matrix</span><span class="p">(</span><span class="n">Gs</span><span class="p">)</span>
        <span class="n">Counts</span> <span class="o">=</span> <span class="n">spt</span><span class="o">.</span><span class="n">count_haplotypes_sparse</span><span class="p">(</span><span class="n">G_dict</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">missing</span><span class="o">=</span><span class="n">any_missing</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">genotypes</span><span class="p">:</span>
        <span class="n">D</span> <span class="o">=</span> <span class="n">gcs</span><span class="o">.</span><span class="n">compute_D</span><span class="p">(</span><span class="n">Counts</span><span class="p">)</span>
        <span class="n">D2</span> <span class="o">=</span> <span class="n">gcs</span><span class="o">.</span><span class="n">compute_D2</span><span class="p">(</span><span class="n">Counts</span><span class="p">)</span>
        <span class="n">Dz</span> <span class="o">=</span> <span class="n">gcs</span><span class="o">.</span><span class="n">compute_Dz</span><span class="p">(</span><span class="n">Counts</span><span class="p">)</span>
        <span class="n">pi2</span> <span class="o">=</span> <span class="n">gcs</span><span class="o">.</span><span class="n">compute_pi2</span><span class="p">(</span><span class="n">Counts</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">D</span> <span class="o">=</span> <span class="n">shc</span><span class="o">.</span><span class="n">D</span><span class="p">(</span><span class="n">Counts</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
        <span class="n">D2</span> <span class="o">=</span> <span class="n">shc</span><span class="o">.</span><span class="n">DD</span><span class="p">([</span><span class="n">Counts</span><span class="o">.</span><span class="n">T</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
        <span class="n">Dz</span> <span class="o">=</span> <span class="n">shc</span><span class="o">.</span><span class="n">Dz</span><span class="p">([</span><span class="n">Counts</span><span class="o">.</span><span class="n">T</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
        <span class="n">pi2</span> <span class="o">=</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">([</span><span class="n">Counts</span><span class="o">.</span><span class="n">T</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">D2</span><span class="p">,</span> <span class="n">Dz</span><span class="p">,</span> <span class="n">pi2</span><span class="p">,</span> <span class="n">D</span></div>



<div class="viewcode-block" id="compute_average_stats">
<a class="viewcode-back" href="../../../api/api_ld.html#moments.LD.Parsing.compute_average_stats">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">compute_average_stats</span><span class="p">(</span><span class="n">Gs</span><span class="p">,</span> <span class="n">genotypes</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Takes the outputs of ``compute_pairwise_stats`` and returns</span>
<span class="sd">    the average value for each statistic.</span>

<span class="sd">    :param Gs: A genotype matrix, of size L-by-n, where</span>
<span class="sd">        L is the number of loci and n is the sample size.</span>
<span class="sd">        Missing data is encoded as -1.</span>
<span class="sd">    :param genotypes: If True, use 0, 1, 2 genotypes. If False,</span>
<span class="sd">        use 0, 1 phased haplotypes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">D2</span><span class="p">,</span> <span class="n">Dz</span><span class="p">,</span> <span class="n">pi2</span><span class="p">,</span> <span class="n">D</span> <span class="o">=</span> <span class="n">compute_pairwise_stats</span><span class="p">(</span><span class="n">Gs</span><span class="p">,</span> <span class="n">genotypes</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">D2</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">Dz</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">pi2</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">D</span><span class="p">)</span></div>



<div class="viewcode-block" id="compute_pairwise_stats_between">
<a class="viewcode-back" href="../../../api/api_ld.html#moments.LD.Parsing.compute_pairwise_stats_between">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">compute_pairwise_stats_between</span><span class="p">(</span><span class="n">Gs1</span><span class="p">,</span> <span class="n">Gs2</span><span class="p">,</span> <span class="n">genotypes</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes :math:`D^2`, :math:`Dz`, :math:`\\pi_2`, and :math:`D`</span>
<span class="sd">    for every pair of loci between two blocks of SNPs, coded as</span>
<span class="sd">    genotype matrices.</span>

<span class="sd">    The Gs are matrices, where rows correspond to loci and columns to individuals.</span>
<span class="sd">    Both matrices must have the same number of individuals. If Gs1 has length L1</span>
<span class="sd">    and Gs2 has length L2, we compute all pairwise counts, which has size (L1*L2, 9).</span>

<span class="sd">    We use the sparse genotype matrix representation, where</span>
<span class="sd">    we first &quot;sparsify&quot; the genotype matrix, and then count</span>
<span class="sd">    two-locus genotype configurations from that, from which</span>
<span class="sd">    we compute two-locus statistics</span>

<span class="sd">    :param Gs1: A genotype matrices, of size L1 by n, where</span>
<span class="sd">        L1 is the number of loci and n is the sample size.</span>
<span class="sd">        Missing data is encoded as -1.</span>
<span class="sd">    :param Gs2: A genotype matrices, of size L2 by n, where</span>
<span class="sd">        L1 is the number of loci and n is the sample size.</span>
<span class="sd">        Missing data is encoded as -1.</span>
<span class="sd">    :param genotypes: If True, use 0, 1, 2 genotypes. If False,</span>
<span class="sd">        use 0, 1 phased haplotypes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="p">(</span>
        <span class="n">ld_extensions</span> <span class="o">==</span> <span class="mi">1</span>
    <span class="p">),</span> <span class="s2">&quot;Need to build LD cython extensions. Install moments with the flag `--ld_extensions`&quot;</span>

    <span class="n">_check_valid_genotype_matrix</span><span class="p">(</span><span class="n">Gs1</span><span class="p">,</span> <span class="n">genotypes</span><span class="p">)</span>
    <span class="n">_check_valid_genotype_matrix</span><span class="p">(</span><span class="n">Gs2</span><span class="p">,</span> <span class="n">genotypes</span><span class="p">)</span>

    <span class="n">L1</span><span class="p">,</span> <span class="n">n1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">Gs1</span><span class="p">)</span>
    <span class="n">L2</span><span class="p">,</span> <span class="n">n2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">Gs1</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">n1</span> <span class="o">!=</span> <span class="n">n2</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;data must have same number of sequenced individuals&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">n</span> <span class="o">=</span> <span class="n">n1</span>

    <span class="k">if</span> <span class="n">genotypes</span><span class="p">:</span>
        <span class="n">G_dict1</span><span class="p">,</span> <span class="n">any_missing1</span> <span class="o">=</span> <span class="n">_sparsify_genotype_matrix</span><span class="p">(</span><span class="n">Gs1</span><span class="p">)</span>
        <span class="n">G_dict2</span><span class="p">,</span> <span class="n">any_missing2</span> <span class="o">=</span> <span class="n">_sparsify_genotype_matrix</span><span class="p">(</span><span class="n">Gs2</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">G_dict1</span><span class="p">,</span> <span class="n">any_missing1</span> <span class="o">=</span> <span class="n">_sparsify_haplotype_matrix</span><span class="p">(</span><span class="n">Gs1</span><span class="p">)</span>
        <span class="n">G_dict2</span><span class="p">,</span> <span class="n">any_missing2</span> <span class="o">=</span> <span class="n">_sparsify_haplotype_matrix</span><span class="p">(</span><span class="n">Gs2</span><span class="p">)</span>

    <span class="n">any_missing</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span><span class="n">any_missing1</span><span class="p">,</span> <span class="n">any_missing2</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">genotypes</span><span class="p">:</span>
        <span class="n">Counts</span> <span class="o">=</span> <span class="n">spt</span><span class="o">.</span><span class="n">count_genotypes_between_sparse</span><span class="p">(</span>
            <span class="n">G_dict1</span><span class="p">,</span> <span class="n">G_dict2</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">missing</span><span class="o">=</span><span class="n">any_missing</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">Counts</span> <span class="o">=</span> <span class="n">spt</span><span class="o">.</span><span class="n">count_haplotypes_between_sparse</span><span class="p">(</span>
            <span class="n">G_dict1</span><span class="p">,</span> <span class="n">G_dict2</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">missing</span><span class="o">=</span><span class="n">any_missing</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">genotypes</span><span class="p">:</span>
        <span class="n">D2</span> <span class="o">=</span> <span class="n">gcs</span><span class="o">.</span><span class="n">compute_D2</span><span class="p">(</span><span class="n">Counts</span><span class="p">)</span>
        <span class="n">Dz</span> <span class="o">=</span> <span class="n">gcs</span><span class="o">.</span><span class="n">compute_Dz</span><span class="p">(</span><span class="n">Counts</span><span class="p">)</span>
        <span class="n">pi2</span> <span class="o">=</span> <span class="n">gcs</span><span class="o">.</span><span class="n">compute_pi2</span><span class="p">(</span><span class="n">Counts</span><span class="p">)</span>
        <span class="n">D</span> <span class="o">=</span> <span class="n">gcs</span><span class="o">.</span><span class="n">compute_D</span><span class="p">(</span><span class="n">Counts</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">D</span> <span class="o">=</span> <span class="n">shc</span><span class="o">.</span><span class="n">D</span><span class="p">(</span><span class="n">Counts</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
        <span class="n">D2</span> <span class="o">=</span> <span class="n">shc</span><span class="o">.</span><span class="n">DD</span><span class="p">([</span><span class="n">Counts</span><span class="o">.</span><span class="n">T</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
        <span class="n">Dz</span> <span class="o">=</span> <span class="n">shc</span><span class="o">.</span><span class="n">Dz</span><span class="p">([</span><span class="n">Counts</span><span class="o">.</span><span class="n">T</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
        <span class="n">pi2</span> <span class="o">=</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">([</span><span class="n">Counts</span><span class="o">.</span><span class="n">T</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">D2</span><span class="p">,</span> <span class="n">Dz</span><span class="p">,</span> <span class="n">pi2</span><span class="p">,</span> <span class="n">D</span></div>



<div class="viewcode-block" id="compute_average_stats_between">
<a class="viewcode-back" href="../../../api/api_ld.html#moments.LD.Parsing.compute_average_stats_between">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">compute_average_stats_between</span><span class="p">(</span><span class="n">Gs1</span><span class="p">,</span> <span class="n">Gs2</span><span class="p">,</span> <span class="n">genotypes</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Takes the outputs of compute_pairwise_stats_between and returns</span>
<span class="sd">    the average value for each statistic.</span>

<span class="sd">    :param Gs1: A genotype matrices, of size L1 by n, where</span>
<span class="sd">        L1 is the number of loci and n is the sample size.</span>
<span class="sd">        Missing data is encoded as -1.</span>
<span class="sd">    :param Gs2: A genotype matrices, of size L2 by n, where</span>
<span class="sd">        L1 is the number of loci and n is the sample size.</span>
<span class="sd">        Missing data is encoded as -1.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">D2</span><span class="p">,</span> <span class="n">Dz</span><span class="p">,</span> <span class="n">pi2</span><span class="p">,</span> <span class="n">D</span> <span class="o">=</span> <span class="n">compute_pairwise_stats_between</span><span class="p">(</span><span class="n">Gs1</span><span class="p">,</span> <span class="n">Gs2</span><span class="p">,</span> <span class="n">genotypes</span><span class="o">=</span><span class="n">genotypes</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">D2</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">Dz</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">pi2</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">D</span><span class="p">)</span></div>



<span class="k">def</span><span class="w"> </span><span class="nf">_count_types_sparse</span><span class="p">(</span>
    <span class="n">genotypes</span><span class="p">,</span>
    <span class="n">bins</span><span class="p">,</span>
    <span class="n">sample_ids</span><span class="p">,</span>
    <span class="n">positions</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">pos_rs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">pop_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">pops</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">use_genotypes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">report</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">report_spacing</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">use_cache</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">stats_to_compute</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">normalized_by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
    <span class="k">assert</span> <span class="p">(</span>
        <span class="n">ld_extensions</span> <span class="o">==</span> <span class="mi">1</span>
    <span class="p">),</span> <span class="s2">&quot;Need to build LD cython extensions. Install moments with the flag `--ld_extensions`&quot;</span>

    <span class="n">pop_indexes</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">pops</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1">## get columns to keep, and compress data and sample_ids</span>
        <span class="n">samples</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">pop_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">)</span>
        <span class="n">cols_to_keep</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">False</span><span class="p">]</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">genotypes</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">all_samples_to_keep</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">:</span>
            <span class="n">all_samples_to_keep</span> <span class="o">+=</span> <span class="nb">list</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="n">samples</span><span class="p">[</span><span class="s2">&quot;pop&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">pop</span><span class="p">][</span><span class="s2">&quot;sample&quot;</span><span class="p">])</span>

        <span class="n">sample_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">all_samples_to_keep</span><span class="p">:</span>
            <span class="n">cols_to_keep</span><span class="p">[</span><span class="n">sample_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">s</span><span class="p">)]</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="n">genotypes_pops</span> <span class="o">=</span> <span class="n">genotypes</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">cols_to_keep</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">sample_ids_pops</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sample_list</span><span class="p">)</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">cols_to_keep</span><span class="p">))</span>

        <span class="c1">## keep only biallelic genotypes from populations in pops, discard the rest</span>
        <span class="n">allele_counts_pops</span> <span class="o">=</span> <span class="n">genotypes_pops</span><span class="o">.</span><span class="n">count_alleles</span><span class="p">()</span>
        <span class="n">is_biallelic</span> <span class="o">=</span> <span class="n">allele_counts_pops</span><span class="o">.</span><span class="n">is_biallelic_01</span><span class="p">()</span>
        <span class="n">genotypes_pops</span> <span class="o">=</span> <span class="n">genotypes_pops</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">is_biallelic</span><span class="p">)</span>

        <span class="c1">## for each population, get the indexes for each population</span>
        <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">:</span>
            <span class="n">pop_indexes</span><span class="p">[</span><span class="n">pop</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">False</span><span class="p">]</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">genotypes_pops</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
            <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">[</span><span class="n">samples</span><span class="p">[</span><span class="s2">&quot;pop&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">pop</span><span class="p">][</span><span class="s2">&quot;sample&quot;</span><span class="p">]:</span>
                <span class="n">pop_indexes</span><span class="p">[</span><span class="n">pop</span><span class="p">][</span><span class="n">sample_ids_pops</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">s</span><span class="p">)]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">pop_indexes</span><span class="p">[</span><span class="n">pop</span><span class="p">]):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;population </span><span class="si">{</span><span class="n">pop</span><span class="si">}</span><span class="s2"> has no samples in data&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
            <span class="n">pop_indexes_haps</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">:</span>
                <span class="n">pop_indexes_haps</span><span class="p">[</span><span class="n">pop</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span>
                    <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">pop_indexes</span><span class="p">[</span><span class="n">pop</span><span class="p">],</span> <span class="n">pop_indexes</span><span class="p">[</span><span class="n">pop</span><span class="p">])),</span>
                    <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">pop_indexes</span><span class="p">[</span><span class="n">pop</span><span class="p">]),),</span>
                <span class="p">)</span>

        <span class="k">if</span> <span class="n">positions</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">positions</span> <span class="o">=</span> <span class="n">positions</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">is_biallelic</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">pos_rs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">pos_rs</span> <span class="o">=</span> <span class="n">pos_rs</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">is_biallelic</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">report</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="n">current_time</span><span class="p">(),</span>
                <span class="s2">&quot;No populations given, using all samples as one population.&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
        <span class="n">pops</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ALL&quot;</span><span class="p">]</span>
        <span class="n">pop_indexes</span><span class="p">[</span><span class="s2">&quot;ALL&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="kc">True</span><span class="p">]</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">genotypes</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">genotypes_pops</span> <span class="o">=</span> <span class="n">genotypes</span>
        <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
            <span class="n">pop_indexes_haps</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">:</span>
                <span class="n">pop_indexes_haps</span><span class="p">[</span><span class="n">pop</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span>
                    <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">pop_indexes</span><span class="p">[</span><span class="n">pop</span><span class="p">],</span> <span class="n">pop_indexes</span><span class="p">[</span><span class="n">pop</span><span class="p">])),</span>
                    <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">pop_indexes</span><span class="p">[</span><span class="n">pop</span><span class="p">]),),</span>
                <span class="p">)</span>

    <span class="c1">## convert to 0,1,2 format</span>
    <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">genotypes_pops_012</span> <span class="o">=</span> <span class="n">genotypes_pops</span><span class="o">.</span><span class="n">to_n_alt</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">haplotypes_pops_01</span> <span class="o">=</span> <span class="n">genotypes_pops</span><span class="o">.</span><span class="n">to_haplotypes</span><span class="p">()</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="n">current_time</span><span class="p">(),</span>
                <span class="s2">&quot;warning: attempted to get haplotypes from phased genotypes, returned attribute error. Using input as haplotypes.&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">haplotypes_pops_01</span> <span class="o">=</span> <span class="n">genotypes_pops</span>

    <span class="k">if</span> <span class="n">pos_rs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">rs</span> <span class="o">=</span> <span class="n">pos_rs</span>
    <span class="k">elif</span> <span class="n">positions</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">rs</span> <span class="o">=</span> <span class="n">positions</span>

    <span class="n">ns</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">ns</span><span class="p">[</span><span class="n">pop</span><span class="p">]</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">pop_indexes</span><span class="p">[</span><span class="n">pop</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ns</span><span class="p">[</span><span class="n">pop</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="nb">sum</span><span class="p">(</span><span class="n">pop_indexes</span><span class="p">[</span><span class="n">pop</span><span class="p">])</span>

    <span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">bins</span><span class="p">)</span>

    <span class="c1">## split and sparsify the geno/haplo-type arrays for each population</span>
    <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">genotypes_by_pop</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">any_missing</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">:</span>
            <span class="n">temp_genotypes</span> <span class="o">=</span> <span class="n">genotypes_pops_012</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">pop_indexes</span><span class="p">[</span><span class="n">pop</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">genotypes_by_pop</span><span class="p">[</span><span class="n">pop</span><span class="p">],</span> <span class="n">this_missing</span> <span class="o">=</span> <span class="n">_sparsify_genotype_matrix</span><span class="p">(</span>
                <span class="n">temp_genotypes</span>
            <span class="p">)</span>
            <span class="n">any_missing</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span><span class="n">any_missing</span><span class="p">,</span> <span class="n">this_missing</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">haplotypes_by_pop</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">any_missing</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">:</span>
            <span class="n">temp_haplotypes</span> <span class="o">=</span> <span class="n">haplotypes_pops_01</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">pop_indexes_haps</span><span class="p">[</span><span class="n">pop</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">haplotypes_by_pop</span><span class="p">[</span><span class="n">pop</span><span class="p">],</span> <span class="n">this_missing</span> <span class="o">=</span> <span class="n">_sparsify_haplotype_matrix</span><span class="p">(</span>
                <span class="n">temp_haplotypes</span>
            <span class="p">)</span>
            <span class="n">any_missing</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span><span class="n">any_missing</span><span class="p">,</span> <span class="n">this_missing</span><span class="p">)</span>

    <span class="c1">#    if use_cache == True:</span>
    <span class="c1">#        run loop that computes type_counts cache</span>
    <span class="c1">#    else</span>
    <span class="c1">#        run loop that adds to sums</span>

    <span class="c1">## if use_cache is True, type_counts will store the number of times we</span>
    <span class="c1">## see each genotype count configuration within each bin</span>
    <span class="c1">## if use_cache is False, we add to the running total of sums of each</span>
    <span class="c1">## statistic as we count their genotypes, never storing the counts of configurations</span>
    <span class="n">bs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">bins</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">bins</span><span class="p">[</span><span class="mi">1</span><span class="p">:]))</span>
    <span class="k">if</span> <span class="n">use_cache</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">type_counts</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bs</span><span class="p">:</span>
            <span class="n">type_counts</span><span class="p">[</span><span class="n">b</span><span class="p">]</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sums</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bs</span><span class="p">:</span>
            <span class="n">sums</span><span class="p">[</span><span class="n">b</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">stat</span> <span class="ow">in</span> <span class="n">stats_to_compute</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                <span class="n">sums</span><span class="p">[</span><span class="n">b</span><span class="p">][</span><span class="n">stat</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1">## loop through left positions and pair with positions to the right</span>
    <span class="c1">## within the bin windows</span>
    <span class="c1">## this is a very inefficient, naive approach</span>
    <span class="k">for</span> <span class="n">ii</span><span class="p">,</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">rs</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]):</span>
        <span class="k">if</span> <span class="n">report</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ii</span> <span class="o">%</span> <span class="n">report_spacing</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="n">current_time</span><span class="p">(),</span>
                    <span class="s2">&quot;tallied two locus counts </span><span class="si">{0}</span><span class="s2"> of </span><span class="si">{1}</span><span class="s2"> positions&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ii</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">rs</span><span class="p">)),</span>
                <span class="p">)</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>

        <span class="c1">## loop through each bin, picking out the positions to the right</span>
        <span class="c1">## of the left locus that fall within the given bin</span>
        <span class="k">if</span> <span class="n">pos_rs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">distances</span> <span class="o">=</span> <span class="n">pos_rs</span> <span class="o">-</span> <span class="n">r</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">distances</span> <span class="o">=</span> <span class="n">positions</span> <span class="o">-</span> <span class="n">r</span>

        <span class="n">filt</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">distances</span> <span class="o">&gt;=</span> <span class="n">bs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="n">distances</span> <span class="o">&lt;</span> <span class="n">bs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]),</span>
            <span class="n">positions</span> <span class="o">!=</span> <span class="n">positions</span><span class="p">[</span><span class="n">ii</span><span class="p">],</span>
        <span class="p">)</span>
        <span class="n">filt</span><span class="p">[</span><span class="n">ii</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>  <span class="c1"># don&#39;t compare to mutations at same base pair position</span>
        <span class="n">right_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">filt</span> <span class="o">==</span> <span class="kc">True</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1">## if there are no variants within the bin&#39;s distance to the right,</span>
        <span class="c1">## continue to next bin</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">right_indices</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="n">right_start</span> <span class="o">=</span> <span class="n">right_indices</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">right_end</span> <span class="o">=</span> <span class="n">right_indices</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>

        <span class="k">if</span> <span class="n">use_cache</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
            <span class="n">counts_ii</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bs</span><span class="p">:</span>
                <span class="n">counts_ii</span><span class="p">[</span><span class="n">b</span><span class="p">]</span> <span class="o">=</span> <span class="p">[[]</span> <span class="k">for</span> <span class="n">pop_ind</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pops</span><span class="p">))]</span>

        <span class="c1">## loop through right loci and count two-locus genotypes</span>
        <span class="k">for</span> <span class="n">jj</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">right_start</span><span class="p">,</span> <span class="n">right_end</span><span class="p">):</span>
            <span class="c1"># get the bin that this pair belongs to</span>
            <span class="n">r_dist</span> <span class="o">=</span> <span class="n">distances</span><span class="p">[</span><span class="n">jj</span><span class="p">]</span>
            <span class="n">bin_ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">r_dist</span> <span class="o">&gt;=</span> <span class="n">bins</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">b</span> <span class="o">=</span> <span class="n">bs</span><span class="p">[</span><span class="n">bin_ind</span><span class="p">]</span>

            <span class="c1"># count genotypes within each population</span>
            <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                <span class="n">cs</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="n">spt</span><span class="o">.</span><span class="n">tally_sparse</span><span class="p">(</span>
                            <span class="n">genotypes_by_pop</span><span class="p">[</span><span class="n">pop</span><span class="p">][</span><span class="n">ii</span><span class="p">],</span>
                            <span class="n">genotypes_by_pop</span><span class="p">[</span><span class="n">pop</span><span class="p">][</span><span class="n">jj</span><span class="p">],</span>
                            <span class="n">ns</span><span class="p">[</span><span class="n">pop</span><span class="p">],</span>
                            <span class="n">any_missing</span><span class="p">,</span>
                        <span class="p">)</span>
                        <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops</span>
                    <span class="p">]</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cs</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="n">spt</span><span class="o">.</span><span class="n">tally_sparse_haplotypes</span><span class="p">(</span>
                            <span class="n">haplotypes_by_pop</span><span class="p">[</span><span class="n">pop</span><span class="p">][</span><span class="n">ii</span><span class="p">],</span>
                            <span class="n">haplotypes_by_pop</span><span class="p">[</span><span class="n">pop</span><span class="p">][</span><span class="n">jj</span><span class="p">],</span>
                            <span class="n">ns</span><span class="p">[</span><span class="n">pop</span><span class="p">],</span>
                            <span class="n">any_missing</span><span class="p">,</span>
                        <span class="p">)</span>
                        <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops</span>
                    <span class="p">]</span>
                <span class="p">)</span>

            <span class="k">if</span> <span class="n">use_cache</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                <span class="n">type_counts</span><span class="p">[</span><span class="n">b</span><span class="p">][</span><span class="n">cs</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">pop_ind</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pops</span><span class="p">)):</span>
                    <span class="n">counts_ii</span><span class="p">[</span><span class="n">b</span><span class="p">][</span><span class="n">pop_ind</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cs</span><span class="p">[</span><span class="n">pop_ind</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">use_cache</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bs</span><span class="p">:</span>
                <span class="n">these_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">counts_ii</span><span class="p">[</span><span class="n">b</span><span class="p">])</span>
                <span class="k">if</span> <span class="n">these_counts</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">for</span> <span class="n">stat</span> <span class="ow">in</span> <span class="n">stats_to_compute</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                    <span class="n">sums</span><span class="p">[</span><span class="n">b</span><span class="p">][</span><span class="n">stat</span><span class="p">]</span> <span class="o">+=</span> <span class="n">_call_sgc</span><span class="p">(</span>
                        <span class="n">stat</span><span class="p">,</span> <span class="n">these_counts</span><span class="o">.</span><span class="n">swapaxes</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="n">use_genotypes</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">use_cache</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">type_counts</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">sums</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_call_sgc</span><span class="p">(</span><span class="n">stat</span><span class="p">,</span> <span class="n">Cs</span><span class="p">,</span> <span class="n">use_genotypes</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    stat = &#39;DD&#39;, &#39;Dz&#39;, or &#39;pi2&#39;, with underscore indices (like &#39;DD_1_1&#39;)</span>
<span class="sd">    Cs = L \times n array, L number of count configurations, n = 4 or 9</span>
<span class="sd">    (for haplotypes or genotypes)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="p">(</span>
        <span class="n">ld_extensions</span> <span class="o">==</span> <span class="mi">1</span>
    <span class="p">),</span> <span class="s2">&quot;Need to build LD cython extensions. Install moments with the flag `--ld_extensions`&quot;</span>

    <span class="n">s</span> <span class="o">=</span> <span class="n">stat</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">pop_nums</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">stat</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]]</span>

    <span class="k">if</span> <span class="n">Cs</span><span class="o">.</span><span class="vm">__class__</span> <span class="o">!=</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cs expected to be a numpy array, got&quot;</span><span class="p">,</span> <span class="n">Cs</span><span class="o">.</span><span class="vm">__class__</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">Cs</span><span class="o">.</span><span class="n">dtype</span> <span class="o">!=</span> <span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">:</span>
        <span class="n">Cs</span> <span class="o">=</span> <span class="n">Cs</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">s</span> <span class="o">==</span> <span class="s2">&quot;DD&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">DD</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="n">pop_nums</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">shc</span><span class="o">.</span><span class="n">DD</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="n">pop_nums</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">s</span> <span class="o">==</span> <span class="s2">&quot;Dz&quot;</span><span class="p">:</span>
        <span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">kk</span> <span class="o">=</span> <span class="n">pop_nums</span>
        <span class="k">if</span> <span class="n">jj</span> <span class="o">==</span> <span class="n">kk</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">Dz</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="n">pop_nums</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">shc</span><span class="o">.</span><span class="n">Dz</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="n">pop_nums</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                <span class="k">return</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">Dz</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">kk</span><span class="p">])</span> <span class="o">+</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">Dz</span><span class="p">(</span>
                    <span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">jj</span><span class="p">]</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">shc</span><span class="o">.</span><span class="n">Dz</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">kk</span><span class="p">])</span> <span class="o">+</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">shc</span><span class="o">.</span><span class="n">Dz</span><span class="p">(</span>
                    <span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">jj</span><span class="p">]</span>
                <span class="p">)</span>
    <span class="k">if</span> <span class="n">s</span> <span class="o">==</span> <span class="s2">&quot;pi2&quot;</span><span class="p">:</span>
        <span class="p">(</span>
            <span class="n">ii</span><span class="p">,</span>
            <span class="n">jj</span><span class="p">,</span>
            <span class="n">kk</span><span class="p">,</span>
            <span class="n">ll</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">=</span> <span class="n">pop_nums</span>  <span class="c1">### this doesn&#39;t consider the symmetry between p/q yet...</span>
        <span class="k">if</span> <span class="n">ii</span> <span class="o">==</span> <span class="n">jj</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">kk</span> <span class="o">==</span> <span class="n">ll</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ii</span> <span class="o">==</span> <span class="n">kk</span><span class="p">:</span>  <span class="c1"># all the same</span>
                    <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                        <span class="k">return</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">return</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                <span class="k">else</span><span class="p">:</span>  <span class="c1"># (1, 1; 2, 2)</span>
                    <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                        <span class="k">return</span> <span class="p">(</span>
                            <span class="mf">1.0</span>
                            <span class="o">/</span> <span class="mi">2</span>
                            <span class="o">*</span> <span class="p">(</span>
                                <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                                <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">])</span>
                            <span class="p">)</span>
                        <span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">return</span> <span class="p">(</span>
                            <span class="mf">1.0</span>
                            <span class="o">/</span> <span class="mi">2</span>
                            <span class="o">*</span> <span class="p">(</span>
                                <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                                <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">])</span>
                            <span class="p">)</span>
                        <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># (1, 1; 2, 3) or (1, 1; 1, 2)</span>
                <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                    <span class="k">return</span> <span class="p">(</span>
                        <span class="mf">1.0</span>
                        <span class="o">/</span> <span class="mi">4</span>
                        <span class="o">*</span> <span class="p">(</span>
                            <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">kk</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ll</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">])</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">return</span> <span class="p">(</span>
                        <span class="mf">1.0</span>
                        <span class="o">/</span> <span class="mi">4</span>
                        <span class="o">*</span> <span class="p">(</span>
                            <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">kk</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ll</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">])</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">kk</span> <span class="o">==</span> <span class="n">ll</span><span class="p">:</span>  <span class="c1"># (1, 2; 3, 3) or (1, 2; 2, 2)</span>
                <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                    <span class="k">return</span> <span class="p">(</span>
                        <span class="mf">1.0</span>
                        <span class="o">/</span> <span class="mi">4</span>
                        <span class="o">*</span> <span class="p">(</span>
                            <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">jj</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">ii</span><span class="p">])</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">return</span> <span class="p">(</span>
                        <span class="mf">1.0</span>
                        <span class="o">/</span> <span class="mi">4</span>
                        <span class="o">*</span> <span class="p">(</span>
                            <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">jj</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">ii</span><span class="p">])</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># (1, 2; 3, 4)</span>
                <span class="k">if</span> <span class="n">use_genotypes</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                    <span class="k">return</span> <span class="p">(</span>
                        <span class="mf">1.0</span>
                        <span class="o">/</span> <span class="mi">8</span>
                        <span class="o">*</span> <span class="p">(</span>
                            <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">kk</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">jj</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">jj</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">kk</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ll</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">ii</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">gcs_mp</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ll</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">ii</span><span class="p">])</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">return</span> <span class="p">(</span>
                        <span class="mf">1.0</span>
                        <span class="o">/</span> <span class="mi">8</span>
                        <span class="o">*</span> <span class="p">(</span>
                            <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">kk</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">jj</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">jj</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">kk</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ll</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">ii</span><span class="p">,</span> <span class="n">jj</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">kk</span><span class="p">,</span> <span class="n">ll</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">ii</span><span class="p">])</span>
                            <span class="o">+</span> <span class="n">shc</span><span class="o">.</span><span class="n">pi2</span><span class="p">(</span><span class="n">Cs</span><span class="p">,</span> <span class="p">[</span><span class="n">ll</span><span class="p">,</span> <span class="n">kk</span><span class="p">,</span> <span class="n">jj</span><span class="p">,</span> <span class="n">ii</span><span class="p">])</span>
                        <span class="p">)</span>
                    <span class="p">)</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_cache_ld_statistics</span><span class="p">(</span><span class="n">type_counts</span><span class="p">,</span> <span class="n">ld_stats</span><span class="p">,</span> <span class="n">bins</span><span class="p">,</span> <span class="n">use_genotypes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">report</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="c1">### This function might not be needed anymore if we completely remove caching</span>
    <span class="n">bs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">bins</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">bins</span><span class="p">[</span><span class="mi">1</span><span class="p">:]))</span>

    <span class="n">estimates</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bs</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">cs</span> <span class="ow">in</span> <span class="n">type_counts</span><span class="p">[</span><span class="n">b</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">estimates</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">cs</span><span class="p">,</span> <span class="p">{})</span>

    <span class="n">all_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">estimates</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
    <span class="n">all_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">swapaxes</span><span class="p">(</span><span class="n">all_counts</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">all_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">swapaxes</span><span class="p">(</span><span class="n">all_counts</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">stat</span> <span class="ow">in</span> <span class="n">ld_stats</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">report</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">current_time</span><span class="p">(),</span> <span class="s2">&quot;computing &quot;</span> <span class="o">+</span> <span class="n">stat</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="n">_call_sgc</span><span class="p">(</span><span class="n">stat</span><span class="p">,</span> <span class="n">all_counts</span><span class="p">,</span> <span class="n">use_genotypes</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">ii</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_counts</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])):</span>
            <span class="n">cs</span> <span class="o">=</span> <span class="n">all_counts</span><span class="p">[:,</span> <span class="p">:,</span> <span class="n">ii</span><span class="p">]</span>
            <span class="n">estimates</span><span class="p">[</span><span class="nb">tuple</span><span class="p">([</span><span class="nb">tuple</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">cs</span><span class="p">])][</span><span class="n">stat</span><span class="p">]</span> <span class="o">=</span> <span class="n">vals</span><span class="p">[</span><span class="n">ii</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">estimates</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_get_ld_stat_sums</span><span class="p">(</span><span class="n">type_counts</span><span class="p">,</span> <span class="n">ld_stats</span><span class="p">,</span> <span class="n">bins</span><span class="p">,</span> <span class="n">use_genotypes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">report</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    return sums[b][stat]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">### this is super inefficient, just trying to get around memory issues</span>

    <span class="n">bs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">bins</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">bins</span><span class="p">[</span><span class="mi">1</span><span class="p">:]))</span>
    <span class="n">sums</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">use_genotypes</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">empty_genotypes</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">([</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="mi">9</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">empty_genotypes</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">([</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="mi">4</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">stat</span> <span class="ow">in</span> <span class="n">ld_stats</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">report</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">current_time</span><span class="p">(),</span> <span class="s2">&quot;computing &quot;</span> <span class="o">+</span> <span class="n">stat</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
        <span class="c1"># set counts of non-used stats to zeros, then take set</span>
        <span class="n">pops_in_stat</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">stat</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:])))</span>
        <span class="n">stat_counts</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bs</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">cs</span> <span class="ow">in</span> <span class="n">type_counts</span><span class="p">[</span><span class="n">b</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">this_count</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">cs</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cs</span><span class="p">)):</span>
                    <span class="k">if</span> <span class="n">i</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pops_in_stat</span><span class="p">:</span>
                        <span class="n">this_count</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">empty_genotypes</span>
                <span class="n">stat_counts</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">this_count</span><span class="p">),</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">))</span>
                <span class="n">stat_counts</span><span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="n">this_count</span><span class="p">)][</span><span class="n">b</span><span class="p">]</span> <span class="o">+=</span> <span class="n">type_counts</span><span class="p">[</span><span class="n">b</span><span class="p">][</span><span class="n">cs</span><span class="p">]</span>

        <span class="n">all_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">stat_counts</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
        <span class="n">all_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">swapaxes</span><span class="p">(</span><span class="n">all_counts</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">all_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">swapaxes</span><span class="p">(</span><span class="n">all_counts</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="n">_call_sgc</span><span class="p">(</span><span class="n">stat</span><span class="p">,</span> <span class="n">all_counts</span><span class="p">,</span> <span class="n">use_genotypes</span><span class="p">)</span>

        <span class="n">estimates</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">v</span><span class="p">,</span> <span class="n">ii</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">vals</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_counts</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]))):</span>
            <span class="n">cs</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">all_counts</span><span class="p">[:,</span> <span class="p">:,</span> <span class="n">ii</span><span class="p">])</span>
            <span class="n">estimates</span><span class="p">[</span><span class="n">cs</span><span class="p">]</span> <span class="o">=</span> <span class="n">v</span>

        <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bs</span><span class="p">:</span>
            <span class="n">sums</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">b</span><span class="p">,</span> <span class="p">{})</span>
            <span class="n">sums</span><span class="p">[</span><span class="n">b</span><span class="p">][</span><span class="n">stat</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">cs</span> <span class="ow">in</span> <span class="n">stat_counts</span><span class="p">:</span>
                <span class="n">sums</span><span class="p">[</span><span class="n">b</span><span class="p">][</span><span class="n">stat</span><span class="p">]</span> <span class="o">+=</span> <span class="n">stat_counts</span><span class="p">[</span><span class="n">cs</span><span class="p">][</span><span class="n">b</span><span class="p">]</span> <span class="o">*</span> <span class="n">estimates</span><span class="p">[</span><span class="n">cs</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">sums</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_get_H_statistics</span><span class="p">(</span>
    <span class="n">genotypes</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">,</span> <span class="n">pop_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pops</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ac_filter</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">report</span><span class="o">=</span><span class="kc">True</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Het values are not normalized by sequence length,</span>
<span class="sd">    would need to compute L from bed file.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">pop_file</span> <span class="o">==</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">pops</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">report</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="n">current_time</span><span class="p">(),</span>
                <span class="s2">&quot;No population file or population names given, &quot;</span>
                <span class="s2">&quot;assuming all samples as single pop.&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
    <span class="k">elif</span> <span class="n">pops</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;pop_file given, but not population names...&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
    <span class="k">elif</span> <span class="n">pop_file</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Population names given, but not pop_file...&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">pops</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">pops</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ALL&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">pop_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">samples</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">pop_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">)</span>
        <span class="n">populations</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="s2">&quot;pop&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">samples</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1">### should use this above when counting two locus genotypes</span>
        <span class="n">sample_ids_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)</span>
        <span class="n">subpops</span> <span class="o">=</span> <span class="p">{</span>
            <span class="c1"># for each population, get the list of samples that</span>
            <span class="c1"># belong to the population</span>
            <span class="n">pop_iter</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">sample_ids_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">[</span><span class="n">samples</span><span class="p">[</span><span class="s2">&quot;pop&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">pop_iter</span><span class="p">][</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span>
            <span class="p">]</span>
            <span class="k">for</span> <span class="n">pop_iter</span> <span class="ow">in</span> <span class="n">pops</span>
        <span class="p">}</span>

        <span class="n">ac_subpop</span> <span class="o">=</span> <span class="n">genotypes</span><span class="o">.</span><span class="n">count_alleles_subpops</span><span class="p">(</span><span class="n">subpops</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">subpops</span> <span class="o">=</span> <span class="p">{</span><span class="n">pop_iter</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)))</span> <span class="k">for</span> <span class="n">pop_iter</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">}</span>
        <span class="n">ac_subpop</span> <span class="o">=</span> <span class="n">genotypes</span><span class="o">.</span><span class="n">count_alleles_subpops</span><span class="p">(</span><span class="n">subpops</span><span class="p">)</span>

    <span class="c1"># ensure at least 2 allele counts per pop</span>
    <span class="n">min_ac_filter</span> <span class="o">=</span> <span class="p">[</span><span class="kc">True</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">ac_subpop</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">ac_filter</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">:</span>
            <span class="n">min_ac_filter</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span>
                <span class="n">min_ac_filter</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">ac_subpop</span><span class="p">[</span><span class="n">pop</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">2</span>
            <span class="p">)</span>
    <span class="n">ac_subpop_filt</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">:</span>
        <span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">ac_subpop</span><span class="p">[</span><span class="n">pop</span><span class="p">])</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">min_ac_filter</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">Hs</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">ii</span><span class="p">,</span> <span class="n">pop1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pops</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">pop2</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">[</span><span class="n">ii</span><span class="p">:]:</span>
            <span class="k">if</span> <span class="n">pop1</span> <span class="o">==</span> <span class="n">pop2</span><span class="p">:</span>
                <span class="n">H</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                    <span class="mf">2.0</span>
                    <span class="o">*</span> <span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop1</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span>
                    <span class="o">*</span> <span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop1</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">]</span>
                    <span class="o">/</span> <span class="p">(</span><span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop1</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop1</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">])</span>
                    <span class="o">/</span> <span class="p">(</span><span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop1</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop1</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">H</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                    <span class="mf">1.0</span>
                    <span class="o">*</span> <span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop1</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span>
                    <span class="o">*</span> <span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop2</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">]</span>
                    <span class="o">/</span> <span class="p">(</span><span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop1</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop1</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">])</span>
                    <span class="o">/</span> <span class="p">(</span><span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop2</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop2</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">])</span>
                    <span class="o">+</span> <span class="mf">1.0</span>
                    <span class="o">*</span> <span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop1</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">]</span>
                    <span class="o">*</span> <span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop2</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span>
                    <span class="o">/</span> <span class="p">(</span><span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop1</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop1</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">])</span>
                    <span class="o">/</span> <span class="p">(</span><span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop2</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">ac_subpop_filt</span><span class="p">[</span><span class="n">pop2</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">])</span>
                <span class="p">)</span>
            <span class="n">Hs</span><span class="p">[(</span><span class="n">pop1</span><span class="p">,</span> <span class="n">pop2</span><span class="p">)]</span> <span class="o">=</span> <span class="n">H</span>

    <span class="k">return</span> <span class="n">Hs</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_get_reported_stats</span><span class="p">(</span>
    <span class="n">genotypes</span><span class="p">,</span>
    <span class="n">bins</span><span class="p">,</span>
    <span class="n">sample_ids</span><span class="p">,</span>
    <span class="n">positions</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">pos_rs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">pop_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">pops</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">use_genotypes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">report</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">report_spacing</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">use_cache</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">stats_to_compute</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">ac_filter</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
    <span class="c1">### build wrapping function that can take use_cache = True or False</span>
    <span class="c1"># now if bins is empty, we only return heterozygosity statistics</span>

    <span class="k">if</span> <span class="n">stats_to_compute</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">pops</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">stats_to_compute</span> <span class="o">=</span> <span class="n">Util</span><span class="o">.</span><span class="n">moment_names</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">stats_to_compute</span> <span class="o">=</span> <span class="n">Util</span><span class="o">.</span><span class="n">moment_names</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pops</span><span class="p">))</span>

    <span class="n">bs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">bins</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">bins</span><span class="p">[</span><span class="mi">1</span><span class="p">:]))</span>

    <span class="k">if</span> <span class="n">use_cache</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">type_counts</span> <span class="o">=</span> <span class="n">_count_types_sparse</span><span class="p">(</span>
            <span class="n">genotypes</span><span class="p">,</span>
            <span class="n">bins</span><span class="p">,</span>
            <span class="n">sample_ids</span><span class="p">,</span>
            <span class="n">positions</span><span class="o">=</span><span class="n">positions</span><span class="p">,</span>
            <span class="n">pos_rs</span><span class="o">=</span><span class="n">pos_rs</span><span class="p">,</span>
            <span class="n">pop_file</span><span class="o">=</span><span class="n">pop_file</span><span class="p">,</span>
            <span class="n">pops</span><span class="o">=</span><span class="n">pops</span><span class="p">,</span>
            <span class="n">use_genotypes</span><span class="o">=</span><span class="n">use_genotypes</span><span class="p">,</span>
            <span class="n">report</span><span class="o">=</span><span class="n">report</span><span class="p">,</span>
            <span class="n">report_spacing</span><span class="o">=</span><span class="n">report_spacing</span><span class="p">,</span>
            <span class="n">use_cache</span><span class="o">=</span><span class="n">use_cache</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">sums</span> <span class="o">=</span> <span class="n">_get_ld_stat_sums</span><span class="p">(</span>
            <span class="n">type_counts</span><span class="p">,</span>
            <span class="n">stats_to_compute</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">bins</span><span class="p">,</span>
            <span class="n">use_genotypes</span><span class="o">=</span><span class="n">use_genotypes</span><span class="p">,</span>
            <span class="n">report</span><span class="o">=</span><span class="n">report</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">sums</span> <span class="o">=</span> <span class="n">_count_types_sparse</span><span class="p">(</span>
            <span class="n">genotypes</span><span class="p">,</span>
            <span class="n">bins</span><span class="p">,</span>
            <span class="n">sample_ids</span><span class="p">,</span>
            <span class="n">positions</span><span class="o">=</span><span class="n">positions</span><span class="p">,</span>
            <span class="n">pos_rs</span><span class="o">=</span><span class="n">pos_rs</span><span class="p">,</span>
            <span class="n">pop_file</span><span class="o">=</span><span class="n">pop_file</span><span class="p">,</span>
            <span class="n">pops</span><span class="o">=</span><span class="n">pops</span><span class="p">,</span>
            <span class="n">use_genotypes</span><span class="o">=</span><span class="n">use_genotypes</span><span class="p">,</span>
            <span class="n">report</span><span class="o">=</span><span class="n">report</span><span class="p">,</span>
            <span class="n">report_spacing</span><span class="o">=</span><span class="n">report_spacing</span><span class="p">,</span>
            <span class="n">use_cache</span><span class="o">=</span><span class="n">use_cache</span><span class="p">,</span>
            <span class="n">stats_to_compute</span><span class="o">=</span><span class="n">stats_to_compute</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">report</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">current_time</span><span class="p">(),</span> <span class="s2">&quot;computed sums</span><span class="se">\n</span><span class="s2">getting heterozygosity statistics&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">stats_to_compute</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">Hs</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">Hs</span> <span class="o">=</span> <span class="n">_get_H_statistics</span><span class="p">(</span>
            <span class="n">genotypes</span><span class="p">,</span>
            <span class="n">sample_ids</span><span class="p">,</span>
            <span class="n">pop_file</span><span class="o">=</span><span class="n">pop_file</span><span class="p">,</span>
            <span class="n">pops</span><span class="o">=</span><span class="n">pops</span><span class="p">,</span>
            <span class="n">ac_filter</span><span class="o">=</span><span class="n">ac_filter</span><span class="p">,</span>
            <span class="n">report</span><span class="o">=</span><span class="n">report</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="n">reported_stats</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">reported_stats</span><span class="p">[</span><span class="s2">&quot;bins&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bs</span>
    <span class="n">reported_stats</span><span class="p">[</span><span class="s2">&quot;sums&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">stats_to_compute</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bs</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span>
        <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">stats_to_compute</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
    <span class="p">]</span>
    <span class="k">for</span> <span class="n">ii</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bs</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">stats_to_compute</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="n">reported_stats</span><span class="p">[</span><span class="s2">&quot;sums&quot;</span><span class="p">][</span><span class="n">ii</span><span class="p">][</span><span class="n">stats_to_compute</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">s</span><span class="p">)]</span> <span class="o">=</span> <span class="n">sums</span><span class="p">[</span><span class="n">b</span><span class="p">][</span><span class="n">s</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">pops</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">pops</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ALL&quot;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">stats_to_compute</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
        <span class="n">reported_stats</span><span class="p">[</span><span class="s2">&quot;sums&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">stats_to_compute</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">s</span><span class="p">)]</span> <span class="o">=</span> <span class="n">Hs</span><span class="p">[</span>
            <span class="p">(</span><span class="n">pops</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])],</span> <span class="n">pops</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">2</span><span class="p">])])</span>
        <span class="p">]</span>
    <span class="n">reported_stats</span><span class="p">[</span><span class="s2">&quot;stats&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">stats_to_compute</span>
    <span class="n">reported_stats</span><span class="p">[</span><span class="s2">&quot;pops&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pops</span>
    <span class="k">return</span> <span class="n">reported_stats</span>


<div class="viewcode-block" id="compute_ld_statistics">
<a class="viewcode-back" href="../../../api/api_ld.html#moments.LD.Parsing.compute_ld_statistics">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">compute_ld_statistics</span><span class="p">(</span>
    <span class="n">vcf_file</span><span class="p">,</span>
    <span class="n">bed_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">chromosome</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">rec_map_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">map_name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">map_sep</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">pop_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">pops</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">cM</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">r_bins</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">bp_bins</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">min_bp</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">use_genotypes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">use_h5</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">stats_to_compute</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">ac_filter</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">report</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">report_spacing</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">use_cache</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes LD statistics for a given VCF. Binning can be done by base pair</span>
<span class="sd">    or recombination distances, the latter requiring a recombination map. For</span>
<span class="sd">    more than one population, we include a population file that maps samples</span>
<span class="sd">    to populations, and specify with populations to compute statistics fro.</span>

<span class="sd">    If data is phased, we can set ``use_genotypes`` to False, and there are</span>
<span class="sd">    other options for masking data.</span>

<span class="sd">    .. note::</span>
<span class="sd">        Currently, the recombination map is not given in HapMap format.</span>
<span class="sd">        Future versions will accept HapMap formatted recombination maps</span>
<span class="sd">        and deprecate some of the boutique handling of map options here.</span>

<span class="sd">    :param str vcf_file: The input VCF file name.</span>
<span class="sd">    :param str bed_file: An optional bed file that specifies regions over which</span>
<span class="sd">        to compute LD statistics. If None, computes statistics for all positions</span>
<span class="sd">        in VCF.</span>
<span class="sd">    :param str chromosome: If None, treats all positions in VCF as coming from same</span>
<span class="sd">        chromosome. If multiple chromosomes are reported in the same VCF, we need to</span>
<span class="sd">        specify which chromosome to keep variants from.</span>
<span class="sd">    :param str rec_map_file: The input recombination map. The format is</span>
<span class="sd">        {pos}\t{map (cM)}\t{additional maps}\n</span>
<span class="sd">    :param str map_name: If None, takes the first map column, otherwise takes the</span>
<span class="sd">        specified map column with the name matching the recombination map file header.</span>
<span class="sd">    :param str map_sep: Deprecated! We now read the recombination map, splitting by</span>
<span class="sd">        any white space. Previous behaviour: Tells pandas how to parse the recombination map.</span>
<span class="sd">    :param str pop_file: A file the specifies the population for each sample in the VCF.</span>
<span class="sd">        Each sample is listed on its own line, in the format &quot;{sample}\t{pop}&quot;. The</span>
<span class="sd">        first line must be &quot;sample\tpop&quot;.</span>
<span class="sd">    :param list(str) pops: List of populations to compute statistics for.</span>
<span class="sd">        If none are given, it treates every sample as coming from the same population.</span>
<span class="sd">    :param bool cM: If True, the recombination map is specified in cM. If False,</span>
<span class="sd">        the map is given in units of Morgans.</span>
<span class="sd">    :param list(float) r_bins: A list of raw recombination rate bin edges.</span>
<span class="sd">    :param list(float) bp_bins: If ``r_bins`` are not given, a list of bp bin</span>
<span class="sd">        edges (for use when no recombination map is specified).</span>
<span class="sd">    :param int, float min_bp: The minimum bp allowed for a segment specified</span>
<span class="sd">        by the bed file.</span>
<span class="sd">    :param bool use_genotypes: If True, we assume the data in the VCF is unphased.</span>
<span class="sd">        Otherwise, we use phased information.</span>
<span class="sd">    :param bool use_h5: If True, we use the h5 format.</span>
<span class="sd">    :param list stats_to_compute: If given, we compute only the statistics specified.</span>
<span class="sd">        Otherwise, we compute all possible statistics for the populations given.</span>
<span class="sd">    :param ac_filter: Ensure at least two samples are present per population. This</span>
<span class="sd">        prevents computed heterozygosity statistics from returning NaN when some</span>
<span class="sd">        loci have too few called samples.</span>
<span class="sd">    :param bool report: If True, we report the progress of our parsing.</span>
<span class="sd">    :param int report_spacing: We track the number of &quot;left&quot; variants we compute,</span>
<span class="sd">        and report our progress with the given spacing.</span>
<span class="sd">    :param bool use_cache: If True, cache intermediate results.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">check_imports</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">r_bins</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">bp_bins</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="s2">&quot;Both r_bins and bp_pins are None, so no LD statistics will be computed&quot;</span>
        <span class="p">)</span>
        <span class="n">bins</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">elif</span> <span class="n">r_bins</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">bp_bins</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Can specify only recombination or bp bins, not both&quot;</span><span class="p">)</span>

    <span class="n">positions</span><span class="p">,</span> <span class="n">genotypes</span><span class="p">,</span> <span class="n">counts</span><span class="p">,</span> <span class="n">sample_ids</span> <span class="o">=</span> <span class="n">get_genotypes</span><span class="p">(</span>
        <span class="n">vcf_file</span><span class="p">,</span>
        <span class="n">bed_file</span><span class="o">=</span><span class="n">bed_file</span><span class="p">,</span>
        <span class="n">chromosome</span><span class="o">=</span><span class="n">chromosome</span><span class="p">,</span>
        <span class="n">min_bp</span><span class="o">=</span><span class="n">min_bp</span><span class="p">,</span>
        <span class="n">use_h5</span><span class="o">=</span><span class="n">use_h5</span><span class="p">,</span>
        <span class="n">report</span><span class="o">=</span><span class="n">report</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">report</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">current_time</span><span class="p">(),</span> <span class="s2">&quot;kept </span><span class="si">{0}</span><span class="s2"> total variants&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)))</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">rec_map_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">r_bins</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">report</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">current_time</span><span class="p">(),</span> <span class="s2">&quot;assigning recombination rates to positions&quot;</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
        <span class="n">pos_rs</span> <span class="o">=</span> <span class="n">_assign_recombination_rates</span><span class="p">(</span>
            <span class="n">positions</span><span class="p">,</span> <span class="n">rec_map_file</span><span class="p">,</span> <span class="n">map_name</span><span class="o">=</span><span class="n">map_name</span><span class="p">,</span> <span class="n">cM</span><span class="o">=</span><span class="n">cM</span><span class="p">,</span> <span class="n">report</span><span class="o">=</span><span class="n">report</span>
        <span class="p">)</span>
        <span class="n">bins</span> <span class="o">=</span> <span class="n">r_bins</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">report</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="n">current_time</span><span class="p">(),</span> <span class="s2">&quot;no recombination map provided, using physical distance&quot;</span>
            <span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
        <span class="n">pos_rs</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">bp_bins</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">bins</span> <span class="o">=</span> <span class="n">bp_bins</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">bins</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">([</span><span class="n">b1</span> <span class="o">-</span> <span class="n">b0</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">b0</span><span class="p">,</span> <span class="n">b1</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">bins</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">bins</span><span class="p">[</span><span class="mi">1</span><span class="p">:])]):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;bins must be a monotonically increasing list&quot;</span><span class="p">)</span>

    <span class="n">reported_stats</span> <span class="o">=</span> <span class="n">_get_reported_stats</span><span class="p">(</span>
        <span class="n">genotypes</span><span class="p">,</span>
        <span class="n">bins</span><span class="p">,</span>
        <span class="n">sample_ids</span><span class="p">,</span>
        <span class="n">positions</span><span class="o">=</span><span class="n">positions</span><span class="p">,</span>
        <span class="n">pos_rs</span><span class="o">=</span><span class="n">pos_rs</span><span class="p">,</span>
        <span class="n">pop_file</span><span class="o">=</span><span class="n">pop_file</span><span class="p">,</span>
        <span class="n">pops</span><span class="o">=</span><span class="n">pops</span><span class="p">,</span>
        <span class="n">use_genotypes</span><span class="o">=</span><span class="n">use_genotypes</span><span class="p">,</span>
        <span class="n">report</span><span class="o">=</span><span class="n">report</span><span class="p">,</span>
        <span class="n">stats_to_compute</span><span class="o">=</span><span class="n">stats_to_compute</span><span class="p">,</span>
        <span class="n">report_spacing</span><span class="o">=</span><span class="n">report_spacing</span><span class="p">,</span>
        <span class="n">use_cache</span><span class="o">=</span><span class="n">use_cache</span><span class="p">,</span>
        <span class="n">ac_filter</span><span class="o">=</span><span class="n">ac_filter</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">reported_stats</span></div>



<div class="viewcode-block" id="means_from_region_data">
<a class="viewcode-back" href="../../../api/api_ld.html#moments.LD.Parsing.means_from_region_data">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">means_from_region_data</span><span class="p">(</span><span class="n">all_data</span><span class="p">,</span> <span class="n">stats</span><span class="p">,</span> <span class="n">norm_idx</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get means over all parsed regions.</span>

<span class="sd">    :param dict all_data: A dictionary with keys as unique identifiers of the</span>
<span class="sd">        regions and values as reported stats from ``compute_ld_statistics``.</span>
<span class="sd">    :param list of lists stats: The list of LD and H statistics that are present</span>
<span class="sd">        in the data replicates.</span>
<span class="sd">    :param int, optional norm_idx: The index of the population to normalize by.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">norm_stats</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;pi2_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">norm_idx</span><span class="p">),</span> <span class="s2">&quot;H_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">norm_idx</span><span class="p">)]</span>
    <span class="n">means</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span> <span class="o">*</span> <span class="n">sums</span> <span class="k">for</span> <span class="n">sums</span> <span class="ow">in</span> <span class="n">all_data</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">all_data</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]][</span><span class="s2">&quot;sums&quot;</span><span class="p">]]</span>
    <span class="k">for</span> <span class="n">reg</span> <span class="ow">in</span> <span class="n">all_data</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">ii</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">means</span><span class="p">)):</span>
            <span class="n">means</span><span class="p">[</span><span class="n">ii</span><span class="p">]</span> <span class="o">+=</span> <span class="n">all_data</span><span class="p">[</span><span class="n">reg</span><span class="p">][</span><span class="s2">&quot;sums&quot;</span><span class="p">][</span><span class="n">ii</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">ii</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">means</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">means</span><span class="p">[</span><span class="n">ii</span><span class="p">]</span> <span class="o">/=</span> <span class="n">means</span><span class="p">[</span><span class="n">ii</span><span class="p">][</span><span class="n">stats</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">norm_stats</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>
    <span class="n">means</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">/=</span> <span class="n">means</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">stats</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">norm_stats</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
    <span class="k">return</span> <span class="n">means</span></div>



<div class="viewcode-block" id="get_bootstrap_sets">
<a class="viewcode-back" href="../../../api/api_ld.html#moments.LD.Parsing.get_bootstrap_sets">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_bootstrap_sets</span><span class="p">(</span>
    <span class="n">all_data</span><span class="p">,</span>
    <span class="n">num_bootstraps</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">normalization</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">remove_norm_stats</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">remove_Dz</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    From a dictionary of all the regional data, resample with replacement</span>
<span class="sd">    to construct bootstrap data.</span>

<span class="sd">    Returns a list of bootstrapped datasets of mean statistics.</span>

<span class="sd">    :param dict all_data: Dictionary of regional LD statistics. Keys are region</span>
<span class="sd">        identifiers and must be unique, and the items are the outputs of</span>
<span class="sd">        ``compute_ld_statistics``.</span>
<span class="sd">    :param int num_bootstraps: The number of bootstrap replicates to compute. If</span>
<span class="sd">        None, it computes the same number as the nubmer of regions in ``all_data``.</span>
<span class="sd">    :param int normalization: The index of the population to normalize by. Defaults</span>
<span class="sd">        to 0.</span>
<span class="sd">    :param bool remove_norm_stats: If we should remove the stat used for normalization.</span>
<span class="sd">    :param bool remove_Dz: If we should remove Dz statistics.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">regions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">all_data</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="n">reg</span> <span class="o">=</span> <span class="n">regions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">stats</span> <span class="o">=</span> <span class="n">all_data</span><span class="p">[</span><span class="n">reg</span><span class="p">][</span><span class="s2">&quot;stats&quot;</span><span class="p">]</span>

    <span class="n">num_regions</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">all_data</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">num_bootstraps</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">num_bootstraps</span> <span class="o">=</span> <span class="n">num_regions</span>

    <span class="n">all_boot</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">rep</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_bootstraps</span><span class="p">):</span>
        <span class="n">temp_data</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">choices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">regions</span><span class="p">,</span> <span class="n">num_regions</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">choices</span><span class="p">):</span>
            <span class="n">temp_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_data</span><span class="p">[</span><span class="n">c</span><span class="p">]</span>
        <span class="n">boot_means</span> <span class="o">=</span> <span class="n">means_from_region_data</span><span class="p">(</span><span class="n">temp_data</span><span class="p">,</span> <span class="n">stats</span><span class="p">,</span> <span class="n">normalization</span><span class="p">)</span>

        <span class="n">delete_ld</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">delete_h</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">remove_norm_stats</span><span class="p">:</span>
            <span class="n">delete_ld</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">stats</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s2">&quot;pi2_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">normalization</span><span class="p">))</span>
            <span class="p">)</span>
            <span class="n">delete_h</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">stats</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s2">&quot;H_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">normalization</span><span class="p">)))</span>
        <span class="k">if</span> <span class="n">remove_Dz</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">stat</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">stats</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
                <span class="k">if</span> <span class="n">stat</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Dz&quot;</span><span class="p">:</span>
                    <span class="n">delete_ld</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">delete_ld</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">ii</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">boot_means</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
                <span class="n">boot_means</span><span class="p">[</span><span class="n">ii</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="n">boot_means</span><span class="p">[</span><span class="n">ii</span><span class="p">],</span> <span class="n">delete_ld</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">delete_h</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">boot_means</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="n">boot_means</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">delete_h</span><span class="p">)</span>

        <span class="n">all_boot</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">boot_means</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">all_boot</span></div>



<div class="viewcode-block" id="bootstrap_data">
<a class="viewcode-back" href="../../../api/api_ld.html#moments.LD.Parsing.bootstrap_data">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">bootstrap_data</span><span class="p">(</span><span class="n">all_data</span><span class="p">,</span> <span class="n">normalization</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns bootstrapped variances for LD statistics. This function operates</span>
<span class="sd">    on data that is sums (i.e. the direct output of ``compute_ld_statistics()``),</span>
<span class="sd">    instead of mean statistics.</span>

<span class="sd">    We first check that all &#39;stats&#39;, &#39;bins&#39;, &#39;pops&#39; (if present),</span>
<span class="sd">    match across all regions</span>

<span class="sd">    If there are N total regions, we compute N bootstrap replicates</span>
<span class="sd">    by sampling N times with replacement and summing over all &#39;sums&#39;.</span>

<span class="sd">    :param dict all_data: A dictionary (with arbitrary keys), where each value</span>
<span class="sd">        is LD statistics computed from a distinct region. all_data[reg]</span>
<span class="sd">        stats from each region has keys, &#39;bins&#39;, &#39;sums&#39;, &#39;stats&#39;, and</span>
<span class="sd">        optional &#39;pops&#39;.</span>
<span class="sd">    :param int normalization: we work with :math:`\\sigma_d^2` statistics,</span>
<span class="sd">        and by default we use population 0 to normalize stats</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># check that var-cov matrix will be able to be computed</span>
    <span class="n">k</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">all_data</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">all_data</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="s2">&quot;sums&quot;</span><span class="p">]:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">all_data</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">v</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;There are not enough independent regions to compute &quot;</span>
                <span class="s2">&quot;variance-covariance matrix&quot;</span>
            <span class="p">)</span>

    <span class="n">norm_stats</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;pi2_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">normalization</span><span class="p">),</span>
        <span class="s2">&quot;H_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">normalization</span><span class="p">),</span>
    <span class="p">]</span>

    <span class="n">regions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">all_data</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    <span class="n">reg</span> <span class="o">=</span> <span class="n">regions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">stats</span> <span class="o">=</span> <span class="n">all_data</span><span class="p">[</span><span class="n">reg</span><span class="p">][</span><span class="s2">&quot;stats&quot;</span><span class="p">]</span>
    <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">regions</span><span class="p">)</span>

    <span class="n">means</span> <span class="o">=</span> <span class="n">means_from_region_data</span><span class="p">(</span><span class="n">all_data</span><span class="p">,</span> <span class="n">stats</span><span class="p">,</span> <span class="n">normalization</span><span class="p">)</span>

    <span class="c1"># construct bootstrap data</span>
    <span class="n">bootstrap_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">sums</span><span class="p">),</span> <span class="n">N</span><span class="p">))</span> <span class="k">for</span> <span class="n">sums</span> <span class="ow">in</span> <span class="n">means</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">boot_num</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">):</span>
        <span class="n">boot_means</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span> <span class="o">*</span> <span class="n">sums</span> <span class="k">for</span> <span class="n">sums</span> <span class="ow">in</span> <span class="n">means</span><span class="p">]</span>
        <span class="n">samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">regions</span><span class="p">,</span> <span class="n">N</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">reg</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">ii</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">boot_means</span><span class="p">)):</span>
                <span class="n">boot_means</span><span class="p">[</span><span class="n">ii</span><span class="p">]</span> <span class="o">+=</span> <span class="n">all_data</span><span class="p">[</span><span class="n">reg</span><span class="p">][</span><span class="s2">&quot;sums&quot;</span><span class="p">][</span><span class="n">ii</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">ii</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">boot_means</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
            <span class="n">boot_means</span><span class="p">[</span><span class="n">ii</span><span class="p">]</span> <span class="o">/=</span> <span class="n">boot_means</span><span class="p">[</span><span class="n">ii</span><span class="p">][</span><span class="n">stats</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">norm_stats</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>
        <span class="n">boot_means</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">/=</span> <span class="n">boot_means</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="n">stats</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">norm_stats</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>

        <span class="k">for</span> <span class="n">ii</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">boot_means</span><span class="p">)):</span>
            <span class="n">bootstrap_data</span><span class="p">[</span><span class="n">ii</span><span class="p">][:,</span> <span class="n">boot_num</span><span class="p">]</span> <span class="o">=</span> <span class="n">boot_means</span><span class="p">[</span><span class="n">ii</span><span class="p">]</span>

    <span class="n">varcovs</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">cov</span><span class="p">(</span><span class="n">bootstrap_data</span><span class="p">[</span><span class="n">ii</span><span class="p">])</span> <span class="k">for</span> <span class="n">ii</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bootstrap_data</span><span class="p">))]</span>

    <span class="n">mv</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">mv</span><span class="p">[</span><span class="s2">&quot;bins&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_data</span><span class="p">[</span><span class="n">reg</span><span class="p">][</span><span class="s2">&quot;bins&quot;</span><span class="p">]</span>
    <span class="n">mv</span><span class="p">[</span><span class="s2">&quot;stats&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_data</span><span class="p">[</span><span class="n">reg</span><span class="p">][</span><span class="s2">&quot;stats&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="s2">&quot;pops&quot;</span> <span class="ow">in</span> <span class="n">all_data</span><span class="p">[</span><span class="n">reg</span><span class="p">]:</span>
        <span class="n">mv</span><span class="p">[</span><span class="s2">&quot;pops&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_data</span><span class="p">[</span><span class="n">reg</span><span class="p">][</span><span class="s2">&quot;pops&quot;</span><span class="p">]</span>
    <span class="n">mv</span><span class="p">[</span><span class="s2">&quot;means&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">means</span>
    <span class="n">mv</span><span class="p">[</span><span class="s2">&quot;varcovs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">varcovs</span>

    <span class="k">return</span> <span class="n">mv</span></div>



<div class="viewcode-block" id="subset_data">
<a class="viewcode-back" href="../../../api/api_ld.html#moments.LD.Parsing.subset_data">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">subset_data</span><span class="p">(</span>
    <span class="n">data</span><span class="p">,</span> <span class="n">pops_to</span><span class="p">,</span> <span class="n">normalization</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">r_min</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">r_max</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">remove_Dz</span><span class="o">=</span><span class="kc">False</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Take the output data and get r_edges, ms, vcs, and stats to pass to inference</span>
<span class="sd">    machinery. ``pops_to`` are the subset of the populations to marginalize the data</span>
<span class="sd">    to. ``r_min`` and ``r_max`` trim bins that fall outside of this range, and</span>
<span class="sd">    ``remove_Dz`` allows us to remove all :math:`\\sigma_{Dz}` statistics.</span>

<span class="sd">    :param data: The output of ``bootstrap_data``, which contains</span>
<span class="sd">        bins, statistics, populations, means, and variance-covariance matrices.</span>
<span class="sd">    :param pops_to: A list of populations to subset to.</span>
<span class="sd">    :param normalization: The population index that the original data was</span>
<span class="sd">        normalized by.</span>
<span class="sd">    :param r_min: The minimum recombination distance to keep.</span>
<span class="sd">    :param r_max: The maximum recombination distance to keep.</span>
<span class="sd">    :param remove_Dz: If True, remove all Dz statistics. Otherwise keep them.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pops_from</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;pops&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">([</span><span class="n">p</span> <span class="ow">in</span> <span class="n">pops_from</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">pops_to</span><span class="p">]):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;All pops in pops_to must be in data&quot;</span><span class="p">)</span>

    <span class="n">new_pop_ids</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops_to</span><span class="p">:</span>
        <span class="n">new_pop_ids</span><span class="p">[</span><span class="n">pops_from</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">pop</span><span class="p">)]</span> <span class="o">=</span> <span class="n">pops_to</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span>

    <span class="n">stats</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;stats&quot;</span><span class="p">]</span>

    <span class="n">to_remove</span> <span class="o">=</span> <span class="p">[[],</span> <span class="p">[]]</span>
    <span class="n">new_stats</span> <span class="o">=</span> <span class="p">[[],</span> <span class="p">[]]</span>

    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]:</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">stat</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">stats</span><span class="p">[</span><span class="n">j</span><span class="p">]):</span>
            <span class="k">if</span> <span class="n">stat</span> <span class="ow">in</span> <span class="p">[</span>
                <span class="s2">&quot;pi2_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">normalization</span><span class="p">),</span>
                <span class="s2">&quot;H_</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">normalization</span><span class="p">),</span>
            <span class="p">]:</span>
                <span class="n">to_remove</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">remove_Dz</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">stat</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Dz&quot;</span><span class="p">:</span>
                        <span class="n">to_remove</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
                        <span class="k">continue</span>
                <span class="n">p_inds</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">stat</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]]</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">p_inds</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">new_pop_ids</span><span class="p">))</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">new_stat</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                        <span class="p">[</span><span class="n">stat</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">+</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">new_pop_ids</span><span class="p">[</span><span class="n">x</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">p_inds</span><span class="p">]</span>
                    <span class="p">)</span>
                    <span class="n">new_stats</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_stat</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">to_remove</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="n">means</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">varcovs</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;bins&quot;</span><span class="p">]):</span>
        <span class="k">if</span> <span class="n">r_min</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">b</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">r_min</span><span class="p">:</span>
                <span class="k">continue</span>
        <span class="k">if</span> <span class="n">r_max</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">b</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">r_max</span><span class="p">:</span>
                <span class="k">continue</span>
        <span class="n">means</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;means&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">to_remove</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="n">varcovs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;varcovs&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">to_remove</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
                <span class="n">to_remove</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="n">means</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;means&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">to_remove</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
    <span class="n">varcovs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;varcovs&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">to_remove</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span> <span class="n">to_remove</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="n">r_edges</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;bins&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">flatten</span><span class="p">()))))</span>
    <span class="k">if</span> <span class="n">r_min</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">r_edges</span> <span class="o">=</span> <span class="n">r_edges</span><span class="p">[</span><span class="n">r_edges</span> <span class="o">&gt;=</span> <span class="n">r_min</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">r_max</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">r_edges</span> <span class="o">=</span> <span class="n">r_edges</span><span class="p">[</span><span class="n">r_edges</span> <span class="o">&lt;=</span> <span class="n">r_max</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">r_edges</span><span class="p">,</span> <span class="n">means</span><span class="p">,</span> <span class="n">varcovs</span><span class="p">,</span> <span class="n">new_stats</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020, Aaron Ragsdale.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>