

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API for site frequency spectra &mdash; moments 1.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/jupyter-sphinx.css" />
      <link rel="stylesheet" type="text/css" href="../_static/thebelab.css" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=fc837d61"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/thebelab-helper.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
      <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="API for linkage disequilibrium" href="api_ld.html" />
    <link rel="prev" title="Selection at two loci" href="../modules/two_locus_selection.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            moments
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">SFS</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../sfs/sfs.html">The Site Frequency Spectrum</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sfs/parsing.html">Parsing the SFS from a VCF file</a></li>
<li class="toctree-l1"><a class="reference internal" href="../sfs/inference.html">SFS Inference</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Linkage disequilibrium</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../ld/ld.html">Multi-population LD statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ld/parsing.html">Parsing LD statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ld/inference.html">Inferring demography with LD</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Extensions</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../extensions/demes.html">Specifying models with <code class="docutils literal notranslate"><span class="pre">demes</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../extensions/two_locus.html">Two-locus frequency spectrum</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extensions/triallele.html">Triallele frequency spectrum</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/diversity.html">Demography and genetic diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/dfe.html">DFE inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/recombination.html">Linkage disequilibrium and recombination</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/two_locus_selection.html">Selection at two loci</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Data types and functions</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">API for site frequency spectra</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#the-spectrum-object">The Spectrum object</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#moments.Spectrum"><code class="docutils literal notranslate"><span class="pre">Spectrum</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.Fst"><code class="docutils literal notranslate"><span class="pre">Spectrum.Fst()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.S"><code class="docutils literal notranslate"><span class="pre">Spectrum.S()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.Tajima_D"><code class="docutils literal notranslate"><span class="pre">Spectrum.Tajima_D()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.Watterson_theta"><code class="docutils literal notranslate"><span class="pre">Spectrum.Watterson_theta()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.Zengs_E"><code class="docutils literal notranslate"><span class="pre">Spectrum.Zengs_E()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.admix"><code class="docutils literal notranslate"><span class="pre">Spectrum.admix()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.branch"><code class="docutils literal notranslate"><span class="pre">Spectrum.branch()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.f2"><code class="docutils literal notranslate"><span class="pre">Spectrum.f2()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.f3"><code class="docutils literal notranslate"><span class="pre">Spectrum.f3()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.f4"><code class="docutils literal notranslate"><span class="pre">Spectrum.f4()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.fixed_size_sample"><code class="docutils literal notranslate"><span class="pre">Spectrum.fixed_size_sample()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.fold"><code class="docutils literal notranslate"><span class="pre">Spectrum.fold()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.from_angsd"><code class="docutils literal notranslate"><span class="pre">Spectrum.from_angsd()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.from_data_dict"><code class="docutils literal notranslate"><span class="pre">Spectrum.from_data_dict()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.from_demes"><code class="docutils literal notranslate"><span class="pre">Spectrum.from_demes()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.from_file"><code class="docutils literal notranslate"><span class="pre">Spectrum.from_file()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.from_vcf"><code class="docutils literal notranslate"><span class="pre">Spectrum.from_vcf()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.fromfile"><code class="docutils literal notranslate"><span class="pre">Spectrum.fromfile()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.genotype_matrix"><code class="docutils literal notranslate"><span class="pre">Spectrum.genotype_matrix()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.integrate"><code class="docutils literal notranslate"><span class="pre">Spectrum.integrate()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.log"><code class="docutils literal notranslate"><span class="pre">Spectrum.log()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.marginalize"><code class="docutils literal notranslate"><span class="pre">Spectrum.marginalize()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.mask_corners"><code class="docutils literal notranslate"><span class="pre">Spectrum.mask_corners()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.pi"><code class="docutils literal notranslate"><span class="pre">Spectrum.pi()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.project"><code class="docutils literal notranslate"><span class="pre">Spectrum.project()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.pulse_migrate"><code class="docutils literal notranslate"><span class="pre">Spectrum.pulse_migrate()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.sample"><code class="docutils literal notranslate"><span class="pre">Spectrum.sample()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.scramble_pop_ids"><code class="docutils literal notranslate"><span class="pre">Spectrum.scramble_pop_ids()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.split"><code class="docutils literal notranslate"><span class="pre">Spectrum.split()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.swap_axes"><code class="docutils literal notranslate"><span class="pre">Spectrum.swap_axes()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.theta_L"><code class="docutils literal notranslate"><span class="pre">Spectrum.theta_L()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.to_file"><code class="docutils literal notranslate"><span class="pre">Spectrum.to_file()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.tofile"><code class="docutils literal notranslate"><span class="pre">Spectrum.tofile()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.unfold"><code class="docutils literal notranslate"><span class="pre">Spectrum.unfold()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.unmask_all"><code class="docutils literal notranslate"><span class="pre">Spectrum.unmask_all()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#moments.Spectrum.unmask_corners"><code class="docutils literal notranslate"><span class="pre">Spectrum.unmask_corners()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#miscellaneous-functions">Miscellaneous functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#moments.Misc.perturb_params"><code class="docutils literal notranslate"><span class="pre">perturb_params()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Misc.make_data_dict_vcf"><code class="docutils literal notranslate"><span class="pre">make_data_dict_vcf()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Misc.count_data_dict"><code class="docutils literal notranslate"><span class="pre">count_data_dict()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Misc.bootstrap"><code class="docutils literal notranslate"><span class="pre">bootstrap()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-moments.Demographics1D">Demographic functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#moments.Demographics1D.bottlegrowth"><code class="docutils literal notranslate"><span class="pre">bottlegrowth()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Demographics1D.growth"><code class="docutils literal notranslate"><span class="pre">growth()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Demographics1D.snm"><code class="docutils literal notranslate"><span class="pre">snm()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Demographics1D.three_epoch"><code class="docutils literal notranslate"><span class="pre">three_epoch()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Demographics1D.two_epoch"><code class="docutils literal notranslate"><span class="pre">two_epoch()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Demographics2D.IM"><code class="docutils literal notranslate"><span class="pre">IM()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Demographics2D.IM_pre"><code class="docutils literal notranslate"><span class="pre">IM_pre()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Demographics2D.bottlegrowth"><code class="docutils literal notranslate"><span class="pre">bottlegrowth()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Demographics2D.bottlegrowth_split"><code class="docutils literal notranslate"><span class="pre">bottlegrowth_split()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Demographics2D.bottlegrowth_split_mig"><code class="docutils literal notranslate"><span class="pre">bottlegrowth_split_mig()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Demographics2D.snm"><code class="docutils literal notranslate"><span class="pre">snm()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Demographics2D.split_mig"><code class="docutils literal notranslate"><span class="pre">split_mig()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Demographics3D.out_of_Africa"><code class="docutils literal notranslate"><span class="pre">out_of_Africa()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#inference-functions">Inference functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#moments.Inference.ll"><code class="docutils literal notranslate"><span class="pre">ll()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Inference.ll_multinom"><code class="docutils literal notranslate"><span class="pre">ll_multinom()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Inference.optimal_sfs_scaling"><code class="docutils literal notranslate"><span class="pre">optimal_sfs_scaling()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Inference.optimally_scaled_sfs"><code class="docutils literal notranslate"><span class="pre">optimally_scaled_sfs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Inference.linear_Poisson_residual"><code class="docutils literal notranslate"><span class="pre">linear_Poisson_residual()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Inference.Anscombe_Poisson_residual"><code class="docutils literal notranslate"><span class="pre">Anscombe_Poisson_residual()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Inference.optimize_log"><code class="docutils literal notranslate"><span class="pre">optimize_log()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Inference.optimize_log_fmin"><code class="docutils literal notranslate"><span class="pre">optimize_log_fmin()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Inference.optimize_log_powell"><code class="docutils literal notranslate"><span class="pre">optimize_log_powell()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Inference.optimize_log_lbfgsb"><code class="docutils literal notranslate"><span class="pre">optimize_log_lbfgsb()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-moments.Godambe">Uncertainty functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#moments.Godambe.FIM_uncert"><code class="docutils literal notranslate"><span class="pre">FIM_uncert()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Godambe.GIM_uncert"><code class="docutils literal notranslate"><span class="pre">GIM_uncert()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#moments.Godambe.LRT_adjust"><code class="docutils literal notranslate"><span class="pre">LRT_adjust()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#plotting-features">Plotting features</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#single-population-plotting">Single-population plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multi-population-plotting">Multi-population plotting</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api_ld.html">API for linkage disequilibrium</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_demes.html">API for <code class="docutils literal notranslate"><span class="pre">demes</span></code> functions</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">moments</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">API for site frequency spectra</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/api/api_moments.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="api-for-site-frequency-spectra">
<span id="sec-sfs-api"></span><h1>API for site frequency spectra<a class="headerlink" href="#api-for-site-frequency-spectra" title="Link to this heading"></a></h1>
<section id="the-spectrum-object">
<h2>The Spectrum object<a class="headerlink" href="#the-spectrum-object" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="moments.Spectrum">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">moments.</span></span><span class="sig-name descname"><span class="pre">Spectrum</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask=np.False_</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_corners=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_folded=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_folding=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype=&lt;class</span> <span class="pre">'float'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fill_value=nan</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_mask=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shrink=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids=None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum" title="Link to this definition"></a></dt>
<dd><p>Represents a single-locus biallelic frequency spectrum.</p>
<p>Spectra are represented by masked arrays. The masking allows us to ignore
specific entries in the spectrum. When simulating under the standard infinite
sites model (ISM), the entries we mask are the bins specifying absent or fixed
variants. When using a reversible mutation model (i.e. the finite genome model),
we track the density of variants in fixed bins, setting <code class="docutils literal notranslate"><span class="pre">mask_corners</span></code> to
<code class="docutils literal notranslate"><span class="pre">False</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>array</em>) – An array with dimension equal to the number of populations.
Each dimension has length <span class="math notranslate nohighlight">\(n_i+1\)</span>, where <span class="math notranslate nohighlight">\(n_i\)</span> is the
sample size for the i-th population.</p></li>
<li><p><strong>mask</strong> – An optional array of the same size as data. ‘True’ entries in
this array are masked in the Spectrum. These represent missing
data categories. (For example, you may not trust your singleton
SNP calling.)</p></li>
<li><p><strong>mask_corners</strong> – If True (default), the ‘observed in none’ and ‘observed
in all’ entries of the FS will be masked. Typically these
entries are masked. In the defaul infinite sites model, moments does
not reliably calculate the fixed-bin entries, so you will almost always
want <code class="docutils literal notranslate"><span class="pre">mask_corners=True</span></code>. The exception is if we are simulating under
the finite genome model, in which case we track the probability of
a site to be fixed for either allele.</p></li>
<li><p><strong>data_folded</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, it is assumed that the input data is folded. An
error will be raised if the input data and mask are not
consistent with a folded Spectrum.</p></li>
<li><p><strong>check_folding</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True and data_folded=True, the data and mask will be
checked to ensure they are consistent with a folded
Spectrum. If they are not, a warning will be printed.</p></li>
<li><p><strong>pop_ids</strong> (<em>list</em><em> of </em><em>strings</em><em>, </em><em>optional</em>) – Optional list of strings containing the population labels,
with length equal to the dimension of <code class="docutils literal notranslate"><span class="pre">data</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A frequency spectrum object, as a masked array.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.Fst">
<span class="sig-name descname"><span class="pre">Fst</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pairwise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.Fst"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.Fst" title="Link to this definition"></a></dt>
<dd><p>Wright’s Fst between the populations represented in the fs.</p>
<p>This estimate of Fst assumes random mating, because we don’t have
heterozygote frequencies in the fs.</p>
<p>Calculation is by the method of Weir and Cockerham <em>Evolution</em> 38:1358
(1984). For a single SNP, the relevant formula is at the top of page
1363. To combine results between SNPs, we use the weighted average
indicated by equation 10.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>pairwise</strong> (<em>bool</em>) – Defaults to False. If True, returns a dictionary
of all pairwise Fst within the multi-dimensional spectrum.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.S">
<span class="sig-name descname"><span class="pre">S</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.S"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.S" title="Link to this definition"></a></dt>
<dd><p>Returns the number of segregating sites in the frequency spectrum.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.Tajima_D">
<span class="sig-name descname"><span class="pre">Tajima_D</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.Tajima_D"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.Tajima_D" title="Link to this definition"></a></dt>
<dd><p>Returns Tajima’s D.</p>
<p>Following Gillespie “Population Genetics: A Concise Guide” pg. 45</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.Watterson_theta">
<span class="sig-name descname"><span class="pre">Watterson_theta</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.Watterson_theta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.Watterson_theta" title="Link to this definition"></a></dt>
<dd><p>Returns Watterson’s estimator of theta.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is only sensible for 1-dimensional spectra.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.Zengs_E">
<span class="sig-name descname"><span class="pre">Zengs_E</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.Zengs_E"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.Zengs_E" title="Link to this definition"></a></dt>
<dd><p>Returns Zeng et al.’s E statistic.</p>
<p>From Zeng et al., “Statistical Tests for Detecting Positive Selection by
Utilizing High-Frequency Variants.” <em>Genetics</em>, 2016.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.admix">
<span class="sig-name descname"><span class="pre">admix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idx1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_lineages</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proportion</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.admix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.admix" title="Link to this definition"></a></dt>
<dd><p>Returns a new frequency spectrum with an admixed population that arose through
admixture from indexed populations with given number of lineages and
proportions from parental populations. This serves as a wrapper for
<code class="docutils literal notranslate"><span class="pre">Manips.admix_into_new</span></code>, with the added feature of handling pop_ids.</p>
<p>If the number of lineages that move are equal to the number
of lineages previously present in a source population, that source
population is marginalized.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>idx0</strong> (<em>int</em>) – Index of first source population.</p></li>
<li><p><strong>idx1</strong> (<em>int</em>) – Index of second source population.</p></li>
<li><p><strong>num_lineages</strong> (<em>int</em>) – Number of lineages in the new population. Cannot be
greater than the number of existing lineages in either source
populations.</p></li>
<li><p><strong>proportion</strong> (<em>float</em>) – The proportion of lineages that come from the first
source population (1-proportion acestry comes from the second source
population). Must be a number between 0 and 1.</p></li>
<li><p><strong>new_id</strong> (<em>str</em><em>, </em><em>optional</em>) – The ID of the new population. Can only be used if the
population IDs are specified in the input SFS.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.branch">
<span class="sig-name descname"><span class="pre">branch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.branch"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.branch" title="Link to this definition"></a></dt>
<dd><p>A “branch” event, where a population gives rise to a child population, while
persisting. This is conceptually similar to the split event. The number of
lineages in the new population is provided, and the number of lineages in the
source/parental population is the original sample size minus the number
requested for the branched population. Returns a new frequency spectrum.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>idx</strong> (<em>int</em>) – The index of the population to branch.</p></li>
<li><p><strong>n</strong> (<em>int</em>) – The sample size of the new population.</p></li>
<li><p><strong>new_id</strong> – The population ID of the branch populations. The parental
population retains its original population ID. Can only be
used if pop_ids are given for the input spectrum.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.f2">
<span class="sig-name descname"><span class="pre">f2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.f2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.f2" title="Link to this definition"></a></dt>
<dd><p>Returns <span class="math notranslate nohighlight">\(f_2(X, Y) = (pX-pY)^2\)</span>, where <span class="math notranslate nohighlight">\(p\)</span> are observed
allele frequencies in populations X and Y.</p>
<p>X, and Y can be specified as population ID strings, or as indexes
(but these cannot be mixed).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> – One of the populations, as index or population ID.</p></li>
<li><p><strong>Y</strong> – The other population, as index or population ID.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Patterson’s f2 statistics.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.f3">
<span class="sig-name descname"><span class="pre">f3</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Z</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.f3"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.f3" title="Link to this definition"></a></dt>
<dd><p>Returns <span class="math notranslate nohighlight">\(f_3(X; Y, Z) = (X-Y)(X-Z)\)</span>. A significantly negative
<span class="math notranslate nohighlight">\(f_3\)</span> suggests that population X is the result of admixture
between ancient populations related to Y and Z. A positive value
suggests that X is an outgroup to Y and Z.</p>
<p>X, Y, and Z can be specified as population ID strings, or as indexes
(but these cannot be mixed).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> – The “test” population, as index or population ID.</p></li>
<li><p><strong>Y</strong> – The first reference population, as index or population ID.</p></li>
<li><p><strong>Z</strong> – The second reference population, as index or population ID.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Patterson’s f3 statistic.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.f4">
<span class="sig-name descname"><span class="pre">f4</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Z</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">W</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.f4"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.f4" title="Link to this definition"></a></dt>
<dd><p>Returns <span class="math notranslate nohighlight">\(f_4(X, Y; Z, W) = (X-Y)(Z-W)\)</span>.</p>
<p>X, Y, Z and W can be specified as population ID strings, or as indexes
(but these cannot be mixed).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> – A population index or ID.</p></li>
<li><p><strong>Y</strong> – A population index or ID.</p></li>
<li><p><strong>Z</strong> – A population index or ID.</p></li>
<li><p><strong>W</strong> – A population index or ID.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Patterson’s f4 statistic (pX-pY)*(pZ-pW).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.fixed_size_sample">
<span class="sig-name descname"><span class="pre">fixed_size_sample</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nsamples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_masked</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.fixed_size_sample"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.fixed_size_sample" title="Link to this definition"></a></dt>
<dd><p>Generate a resampled SFS from the current one. Thus, the resampled SFS
follows a multinomial distribution given by the proportion of sites
in each bin in the original SFS.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nsamples</strong> (<em>int</em>) – Number of samples to include in the new SFS.</p></li>
<li><p><strong>include_masked</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, use all bins from the SFS. Otherwise,
use only non-masked bins. Defaults to False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.fold">
<span class="sig-name descname"><span class="pre">fold</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.fold"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.fold" title="Link to this definition"></a></dt>
<dd><p>Returns a folded frequency spectrum.</p>
<p>The folded fs assumes that information on which allele is ancestral or
derived is unavailable. Thus the fs is in terms of minor allele
frequency.  This makes the fs into a “triangular” array. If a masked cell
is folded into non-masked cell, the destination cell is masked as well.</p>
<p>Folding is not done in-place. The return value is a new Spectrum object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.from_angsd">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_angsd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sfs_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_sizes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">folded</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_corners</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.from_angsd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.from_angsd" title="Link to this definition"></a></dt>
<dd><p>Convert ANGSD output to a moments Spectrum object. The sample sizes
are given as number of haploid genome copies (twice the number of
sampled diploid individuals).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sfs_file</strong> (<em>string</em>) – The n-dimensional SFS from ANGSD. This should be a
file with a single line of numbers, as entries in the SFS.</p></li>
<li><p><strong>sample_sizes</strong> (<em>list</em>) – A list of integers with length equal to the number
of population, storing the haploid sample size in each population.
The order must match the population order provided to ANGSD.</p></li>
<li><p><strong>pop_ids</strong> (<em>list</em>) – A list of strings equal with length equal to the number
of population, specifying the population name for each.</p></li>
<li><p><strong>folded</strong> (<em>bool</em>) – If False (default), we assume ancestral states are
known, returning an unfolded SFS. If True, the returned SFS is folded.</p></li>
<li><p><strong>mask_corners</strong> (<em>bool</em>) – If True (default), mask the fixed bins in the
SFS. If False, the fixed bins will remain unmasked.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A <code class="docutils literal notranslate"><span class="pre">moments</span></code> site frequency spectrum.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#moments.Spectrum" title="moments.Spectrum"><code class="xref py py-class docutils literal notranslate"><span class="pre">moments.Spectrum</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.from_data_dict">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_data_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">projections</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_corners</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polarized</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.from_data_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.from_data_dict" title="Link to this definition"></a></dt>
<dd><p>Spectrum from a dictionary of polymorphisms.</p>
<p>The data dictionary should be organized as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">snp_id</span><span class="p">:</span> <span class="p">{</span>
    <span class="s1">&#39;segregating&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">],</span>
    <span class="s1">&#39;calls&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;YRI&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mi">23</span><span class="p">,</span><span class="mi">3</span><span class="p">),</span>
        <span class="s1">&#39;CEU&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mi">7</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
    <span class="p">},</span>
    <span class="s1">&#39;outgroup_allele&#39;</span><span class="p">:</span> <span class="s1">&#39;T&#39;</span>
<span class="p">}}</span>
</pre></div>
</div>
<p>The ‘calls’ entry gives the successful calls in each population, in the
order that the alleles are specified in ‘segregating’.
Non-diallelic polymorphisms are skipped.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pop_ids</strong> – list of which populations to make fs for.</p></li>
<li><p><strong>projections</strong> – list of sample sizes to project down to for each
population.</p></li>
<li><p><strong>polarized</strong> – If True, the data are assumed to be correctly polarized by
‘outgroup_allele’. SNPs in which the ‘outgroup_allele’
information is missing or ‘-’ or not concordant with the
segregating alleles will be ignored.
If False, any ‘outgroup_allele’ info present is ignored,
and the returned spectrum is folded.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.from_demes">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_demes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sampled_demes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_sizes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_times</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unsampled_n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gamma</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">theta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.from_demes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.from_demes" title="Link to this definition"></a></dt>
<dd><p>Takes a deme graph and computes the SFS. <code class="docutils literal notranslate"><span class="pre">demes</span></code> is a package for
specifying demographic models in a user-friendly, human-readable YAML
format. This function automatically parses the demographic description
and returns a SFS for the specified populations and sample sizes.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If a deme sample time is requested that is earlier than the deme’s
end time, for example to simulate ancient samples, we must create a
new population for that ancient sample. This can cause large
slow-downs, as the computation cost of computing the SFS grows
quickly in the number of populations.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>g</strong> (str or <code class="xref py py-class docutils literal notranslate"><span class="pre">demes.DemeGraph</span></code>) – A <code class="docutils literal notranslate"><span class="pre">demes</span></code> DemeGraph from which to compute the SFS. The DemeGraph
can either be specified as a YAML file, in which case <code class="docutils literal notranslate"><span class="pre">g</span></code> is a string,
or as a <code class="docutils literal notranslate"><span class="pre">DemeGraph</span></code> object.</p></li>
<li><p><strong>sampled_demes</strong> (<em>list</em><em> of </em><em>strings</em>) – A list of deme IDs to take samples from. We can repeat
demes, as long as the sampling of repeated deme IDs occurs at distinct
times.</p></li>
<li><p><strong>sample_sizes</strong> (<em>list</em><em> of </em><em>ints</em>) – A list of the same length as <code class="docutils literal notranslate"><span class="pre">sampled_demes</span></code>,
giving the sample sizes for each sampled deme.</p></li>
<li><p><strong>sample_times</strong> (<em>list</em><em> of </em><em>floats</em><em>, </em><em>optional</em>) – If None, assumes all sampling occurs at the end of the
existence of the sampled deme. If there are
ancient samples, <code class="docutils literal notranslate"><span class="pre">sample_times</span></code> must be a list of same length as
<code class="docutils literal notranslate"><span class="pre">sampled_demes</span></code>, giving the sampling times for each sampled
deme. Sampling times are given in time units of the original deme graph,
so might not necessarily be generations (e.g. if <code class="docutils literal notranslate"><span class="pre">g.time_units</span></code> is years)</p></li>
<li><p><strong>unsampled_n</strong> (<em>int</em><em>, </em><em>optional</em>) – The default sample size of unsampled demes, which must be
greater than or equal to 4.</p></li>
<li><p><strong>gamma</strong> (<em>float</em><em> or </em><em>dict</em>) – The scaled selection coefficient(s), 2*Ne*s. Defaults to None,
which implies neutrality. Can be given as a scalar value, in which case
all populations have the same selection coefficient. Alternatively, can
be given as a dictionary, with keys given as population names in the
input Demes model. Any population missing from this dictionary will be
assigned a selection coefficient of zero. A non-zero default selection
coefficient can be provided, using the key <code class="docutils literal notranslate"><span class="pre">_default</span></code>. See the Demes
exension documentation for more details and examples.</p></li>
<li><p><strong>h</strong> (<em>float</em><em> or </em><em>dict</em>) – The dominance coefficient(s). Defaults to additivity (or genic
selection). Can be given as a scalar value, in which case all populations
have the same dominance coefficient. Alternatively, can be given as a
dictionary, with keys given as population names in the input Demes model.
Any population missing from this dictionary will be assigned a dominance
coefficient of 1/2 (additivity). A different default dominance
coefficient can be provided, using the key <code class="docutils literal notranslate"><span class="pre">_default</span></code>. See the Demes
exension documentation for more details and examples.</p></li>
<li><p><strong>theta</strong> (<em>scalar</em>) – The population-size scaled mutation rate (4*Ne*u or 4*Ne*u*L).
The default value is 1. For more control of mutation rates and mutation
models (including using a reversible mutation model), please use
<code class="docutils literal notranslate"><span class="pre">moments.Demes.SFS</span></code>, which has options to specify theta, the per-base
mutation rate u, and/or a reversible mutation model that allows for
different forward and backward mutation rates.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A <code class="docutils literal notranslate"><span class="pre">moments</span></code> site frequency spectrum, with dimension equal to the
length of <code class="docutils literal notranslate"><span class="pre">sampled_demes</span></code>, and shape equal to <code class="docutils literal notranslate"><span class="pre">sample_sizes</span></code> plus one
in each dimension, indexing the allele frequency in each deme from 0
to n[i], where i is the deme index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#moments.Spectrum" title="moments.Spectrum"><code class="xref py py-class docutils literal notranslate"><span class="pre">moments.Spectrum</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.from_file">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_corners</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_comments</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.from_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.from_file" title="Link to this definition"></a></dt>
<dd><p>Read frequency spectrum from file.</p>
<p>See <code class="docutils literal notranslate"><span class="pre">to_file</span></code> for details on the file format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fid</strong> (<em>string</em>) – string with file name to read from or an open file object.</p></li>
<li><p><strong>mask_corners</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, mask the ‘absent in all samples’ and ‘fixed in
all samples’ entries. If False, the corners are unmasked. If unspecified
(left as None), the mask is unchanged from the saved file.</p></li>
<li><p><strong>return_comments</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, the return value is (fs, comments), where
comments is a list of strings containing the comments
from the file (without #’s).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.from_vcf">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vcf_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_mapping</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pops</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anc_seq_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_low_confidence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_AA</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_multiallelic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_sizes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_corners</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ploidy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">folded</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.from_vcf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.from_vcf" title="Link to this definition"></a></dt>
<dd><p>Compute the SFS from genotype data stored in a VCF file. There are
several optional parameters for controlling the assignment of ancestral
states, filtering by quality or annotation, restricting SFS parsing to
certain intervals, and adjusting the shape of the output SFS.</p>
<p>There are several ways to specify estimated ancestral states. By
default, VCF <code class="docutils literal notranslate"><span class="pre">REF</span></code> alleles are interpreted as ancestral alleles. In
this case the output SFS should be folded, because the reference allele
does not in general correspond to the ancestral allele. If <code class="docutils literal notranslate"><span class="pre">use_AA</span></code>
is True, then the <code class="docutils literal notranslate"><span class="pre">INFO/AA</span></code> VCF field specifies the ancestral allele.
Sites where <code class="docutils literal notranslate"><span class="pre">INFO/AA</span></code> is absent or missing data (represented by ‘.’)
are skipped, raising a warning at the first occurence.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">anc_seq_file</span></code> (FASTA format) is given, then ancestral alleles are
read from it. Here also, sites are skipped when they lack a valid
ancestral allele. This behavior is modulated by
<code class="docutils literal notranslate"><span class="pre">allow_low_confidence</span></code>: when True, sites assigned ancestral states
represented in lower-case (e.g. ‘a’) are retained. This usually denotes
a low-confidence assignment. Otherwise such sites are skipped. If there
are sites in the VCF that fall beyond the end of the FASTA sequence
which are not otherwise masked or excluded by the <code class="docutils literal notranslate"><span class="pre">interval</span></code> argument,
an error is raised.</p>
<p>When <code class="docutils literal notranslate"><span class="pre">allow_multiallelic</span></code> is False and the ancestral allele is not
represented as either the reference or alternate allele at a site, that
site is skipped automatically. The relationship between derived alleles
at such sites is not generally clear.</p>
<p>The parameter <code class="docutils literal notranslate"><span class="pre">filters</span></code> allows filtering by quality and/or annotation
at the site and sample level. It should be a flat dictionary with str
keys. Key-value pairs may have the following forms:</p>
<ul class="simple">
<li><p>‘QUAL’ should map to a single number (float or int), imposing a minimum
value for site <code class="docutils literal notranslate"><span class="pre">QUAL</span></code> fields.</p></li>
<li><p>‘FILTER’ should map to a string or a set, tuple or list of strings. For
a site to pass, its <code class="docutils literal notranslate"><span class="pre">FILTER</span></code> field must equal either the value of
‘FILTER’ (if a string) or that of one of its elements (if a set,
tuple or list).</p></li>
<li><p>‘INFO/FIELD’ e.g. ‘INFO/GQ’ may map to a number, a string, or a set,
tuple or list of strings. When it maps to a number, passing sites
must have <code class="docutils literal notranslate"><span class="pre">INFO/FIELD</span></code> greater than or equal to that number to
pass. When it maps to a string, sites must have equal
<code class="docutils literal notranslate"><span class="pre">INFO/FIELD</span></code>. When it maps to a set, tuple or list of strings, the
filter is said to be categorical and a site’s <code class="docutils literal notranslate"><span class="pre">INFO/FIELD</span></code> must be
a member of the set/tuple/list to pass.
The types of values are not explictly checked against the proper
type for their fields- e.g. if ‘INFO/GQ’ maps to a string rather
than a number, no explicit warning is raised, although an error will
typically be thrown once parsing begins.</p></li>
<li><p>‘SAMPLE/FIELD’ e.g. ‘SAMPLE/GQ’ imposes filters at the sample level.
‘FORMAT/FIELD’ is equivalent to ‘SAMPLE/FIELD’. The ‘FIELD’ should
correspond to an entry in the <code class="docutils literal notranslate"><span class="pre">FORMAT</span></code> column of the VCF file.
Typing is the same as for <code class="docutils literal notranslate"><span class="pre">INFO/FIELD</span></code>.</p></li>
</ul>
<p>When fields targeted for filtering are missing (‘.’) or absent in given
lines/samples, those lines/samples are not skipped, but a one-time
alert message is raised. Depending on the context, this may be a sign
of misspecified filters, or it may be unproblematic. Any combination of
valid filter fields is permissible.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vcf_file</strong> (<em>str</em>) – Pathname of the VCF file to parse. The file may be
gzipped, bgzipped or uncompressed.</p></li>
<li><p><strong>pop_mapping</strong> (<em>dict</em><em>, </em><em>optional</em>) – Optional dictionary (default None) mapping
population IDs to lists of VCF sample IDs. Equivalent in function
to, and mutually exclusive with, <code class="docutils literal notranslate"><span class="pre">pop_file</span></code>.</p></li>
<li><p><strong>pop_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Pathname of a whitespace-separated file mapping samples
to populations with the format SAMPLE POPULATION. Sample names must
be unique and there should be one of them on each line (default None
combines all samples into a single population ‘ALL’). Samples
present in the VCF but not included here are ignored.</p></li>
<li><p><strong>pops</strong> (<em>list</em><em> of </em><em>str</em>) – A list of populations from <code class="docutils literal notranslate"><span class="pre">pop_file</span></code> to parse (default
None). Only functions when <code class="docutils literal notranslate"><span class="pre">pop_file</span></code> is given. Populations not in
<code class="docutils literal notranslate"><span class="pre">pops</span></code> are ignored. If None, then all populations in <code class="docutils literal notranslate"><span class="pre">pop_file</span></code> are
included.</p></li>
<li><p><strong>bed_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Pathname of a BED file defining the intervals within
which to parse; useful for applying masks to exclude difficult-to-
call or functionally constrained genomic regions (default None).
BED files represent intervals as 0-indexed and half-open (the ends
of intervals are noninclusive).</p></li>
<li><p><strong>interval</strong> (<em>tuple</em><em> or </em><em>list</em><em> of </em><em>integers</em><em>, </em><em>optional</em>) – 2-tuple or 2-list specifying a 1-indexed, half-open
(upper boundary noninclusive) genomic window to parse (default
None). May be used in conjuction with a BED file.</p></li>
<li><p><strong>anc_seq_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Pathname of a FASTA file defining inferred
ancestral nucleotide states (default None).</p></li>
<li><p><strong>allow_low_confidence</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True (default False) and <code class="docutils literal notranslate"><span class="pre">anc_seq_file</span></code>
is given, allows low-confidence ancestral state assignments-
represented by lower-case nucleotide codes- to stand. If False,
sites with low-confidence assignments are skipped.</p></li>
<li><p><strong>use_AA</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, use entries in the VCF field <code class="docutils literal notranslate"><span class="pre">INFO/AA</span></code> to
assign ancestral alleles (default False).</p></li>
<li><p><strong>filters</strong> (<em>dict</em><em>, </em><em>optional</em>) – A dictionary mapping VCF fields to filter criteria, for
imposing quantitative thresholds on measures of genotype quality and
categorical requirements on annotations. Filtering is discussed
above.</p></li>
<li><p><strong>allow_multiallelic</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True (default False), includes sites with
more than one alternate allele, counting each derived allele at such
sites as a separate entry in the SFS- otherwise multiallelic sites
are skipped. Also allows sites where neither the reference nor any
alternate allelle(s) matches the assigned ancestral state, which are
skipped when False.</p></li>
<li><p><strong>sample_sizes</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary mapping populations to haploid sample
sizes (default None). Determines the shape of the returned SFS.
Any VCF sites with sample sizes greater than <code class="docutils literal notranslate"><span class="pre">sample_sizes</span></code> will be
projected down to match it. This may be useful when some genotype
data is missing or filtered- sites with missing data are otherwise
not included in the output SFS. When not given, output sample sizes
default to the sample sizes implied by <code class="docutils literal notranslate"><span class="pre">ploidy</span></code> and the number of
individuals in each population.</p></li>
<li><p><strong>mask_corners</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True (default), the ‘observed in none’ and
‘observed in all’ entries of the SFS array are masked.</p></li>
<li><p><strong>ploidy</strong> – Optionally defines the ploidy of samples (default 2).
Used to determine the haploid sample size from the number of sampled
individuals when <code class="docutils literal notranslate"><span class="pre">sample_sizes</span></code> is not given.</p></li>
<li><p><strong>verbose</strong> (<em>int</em><em>, </em><em>optional</em>) – If &gt; 0, print a progress message every <code class="docutils literal notranslate"><span class="pre">verbose</span></code> lines
(default 0).</p></li>
<li><p><strong>folded</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, return the folded SFS (default False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The SFS, represented as a <code class="docutils literal notranslate"><span class="pre">moments.Spectrum</span></code> instance.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#moments.Spectrum" title="moments.Spectrum">moments.Spectrum</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.fromfile">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fromfile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_corners</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_comments</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#moments.Spectrum.fromfile" title="Link to this definition"></a></dt>
<dd><p>Read frequency spectrum from file.</p>
<p>See <code class="docutils literal notranslate"><span class="pre">to_file</span></code> for details on the file format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fid</strong> (<em>string</em>) – string with file name to read from or an open file object.</p></li>
<li><p><strong>mask_corners</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, mask the ‘absent in all samples’ and ‘fixed in
all samples’ entries. If False, the corners are unmasked. If unspecified
(left as None), the mask is unchanged from the saved file.</p></li>
<li><p><strong>return_comments</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, the return value is (fs, comments), where
comments is a list of strings containing the comments
from the file (without #’s).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.genotype_matrix">
<span class="sig-name descname"><span class="pre">genotype_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_sizes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">diploid_genotypes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.genotype_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.genotype_matrix" title="Link to this definition"></a></dt>
<dd><p>Generate a genotype matrix of independent loci. For multi-population spectra,
the individual columns are filled in the sample order as the populations
in the SFS.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Sites in the output genotype matrix are necessarily separated by
infinite recombination. The SFS assumes all loci are segregating
independently, so there is no linkage between them.</p>
</div>
<p>Returns a genotype matrix of size number of sites by total sample size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_sites</strong> – Defaults to None, in which case we take a poisson sample from
the SFS. Otherwise, we take a fixed number of sites.</p></li>
<li><p><strong>sample_sizes</strong> – The sample size in each population, as a list with length
of the number of dimension (populations) in the SFS.</p></li>
</ul>
</dd>
<dt class="field-even">Diploid_genotypes<span class="colon">:</span></dt>
<dd class="field-even"><p>Defaults to False, in which case we return a haplotype
matrix of size (num_sites x sum(sample_sizes)). If True, we return a
diploid genotype matrix (filled with 0, 1, 2) of size
(num_sites x sum(sample_sizes)/2).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.integrate">
<span class="sig-name descname"><span class="pre">integrate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">Npop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dt_fac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.02</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gamma</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overdominance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">m</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">theta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adapt_dt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">finite_genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">theta_fd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">theta_bd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frozen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[False]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.integrate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.integrate" title="Link to this definition"></a></dt>
<dd><p>Method to simulate the spectrum’s evolution for a given set of demographic
parameters. The SFS is integrated forward-in-time, and the integration
occurs in-place, meaning you need only call <code class="docutils literal notranslate"><span class="pre">fs.integrate(</span> <span class="pre">)</span></code>, and the
<code class="docutils literal notranslate"><span class="pre">fs</span></code> is updated.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>Npop</strong> (<em>list</em><em> or </em><em>function that returns a list</em>) – List of populations’ relative effective sizes. Can be given
as a list of positive values for constant sizes, or as a function that
returns a list of sizes at a given time.</p></li>
<li><p><strong>tf</strong> (<em>float</em>) – The total integration time in genetic units.</p></li>
<li><p><strong>dt_fac</strong> (<em>float</em><em>, </em><em>optional</em>) – The timestep factor, default is 0.02. This parameter typically
does not need to be adjusted.</p></li>
<li><p><strong>gamma</strong> (<em>float</em><em> or </em><em>list</em><em> of </em><em>floats</em><em> or </em><em>function that returns a float</em><em> or
</em><em>list</em><em> of </em><em>floats</em><em>, </em><em>optional</em>) – Scaled selection coefficient(s), which can be either a number
or a vector gamma = [gamma1,…,gammap], allowing for different selection
coefficients in each population. This can also be given as function
returning a number or such a vector, allowing for selection coefficients
to change over time.</p></li>
<li><p><strong>h</strong> (<em>float</em><em> or </em><em>list</em><em> of </em><em>floats</em><em> or </em><em>function that returns a float</em><em> or
</em><em>list</em><em> of </em><em>floats</em><em>, </em><em>optional</em>) – Dominance coefficient(s) (h=1/2 implies additive selection). Can be
given as a number or a vector h = [h1,…,hp], or a function that returns
a number or vector of coefficients, allowing for dominance coefficients
to change over time.</p></li>
<li><p><strong>overdominance</strong> (<em>float</em><em> or </em><em>list</em><em> of </em><em>floats</em><em> or </em><em>function that returns a float</em><em> or
</em><em>list</em><em> of </em><em>floats</em><em>, </em><em>optional</em>) – Scaled selection coefficient(s) that is applied only to
heterozygotes, in a selection system with fitnesses 1:1+s:1. Underdominance
can be modeled by passing a negative value. Not that this is a symmetric
under/over-dominance model, in which homozygotes for either the ancestral
or derived allele have equal fitness. <code class="docutils literal notranslate"><span class="pre">gamma</span></code>, <code class="docutils literal notranslate"><span class="pre">h</span></code>, and <code class="docutils literal notranslate"><span class="pre">overdominance</span></code>
can be combined (additively) to implement non-symmetric selection
scenarios.</p></li>
<li><p><strong>m</strong> (<em>array-like</em><em>, </em><em>optional</em>) – The migration rates matrix as a 2-D array with shape nxn,
where n is the number of populations. The entry of the migration
matrix m[i,j] is the migration rate from pop j to pop i in genetic
units, that is, normalized by <span class="math notranslate nohighlight">\(2N_e\)</span>. <code class="docutils literal notranslate"><span class="pre">m</span></code> may be either a
2-D array, or a function that returns a 2-D array (with dimensions
equal to (num pops)x(num pops)).</p></li>
<li><p><strong>theta</strong> (<em>float</em><em>, </em><em>optional</em>) – The scaled mutation rate <span class="math notranslate nohighlight">\(4 N_e u\)</span>, which defaults to 1.
<code class="docutils literal notranslate"><span class="pre">theta</span></code> can be used in the reversible model in the case of symmetric
mutation rates. In this case, <code class="docutils literal notranslate"><span class="pre">theta</span></code> must be set to &lt;&lt; 1.</p></li>
<li><p><strong>adapt_dt</strong> (<em>bool</em><em>, </em><em>optional</em>) – flag to allow dt correction avoiding negative entries.</p></li>
<li><p><strong>finite_genome</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, simulate under the finite-genome model with
reversible mutations. If using this model, we can specify the forward
and backward mutation rates, which are per-base rates that are not
scaled by number of mutable loci. If <code class="docutils literal notranslate"><span class="pre">theta_fd</span></code> and <code class="docutils literal notranslate"><span class="pre">theta_bd</span></code>
are not specified, we assume equal forward and backward mutation rates
provided by <code class="docutils literal notranslate"><span class="pre">theta</span></code>, which must be set to less that 1.
Defaults to False.</p></li>
<li><p><strong>theta_fd</strong> (<em>float</em><em>, </em><em>optional</em>) – The forward mutation rate <span class="math notranslate nohighlight">\(4 Ne u\)</span>.</p></li>
<li><p><strong>theta_bd</strong> (<em>float</em><em>, </em><em>optional</em>) – The backward mutation rate <span class="math notranslate nohighlight">\(4 Ne v\)</span>.</p></li>
<li><p><strong>frozen</strong> (<em>list</em><em> of </em><em>bools</em>) – Specifies the populations that are “frozen”, meaning
samples from that population no longer change due or contribute
to migration to other populations. This feature is most often
used to indicate ancient samples, for example, ancient DNA.
The <code class="docutils literal notranslate"><span class="pre">frozen</span></code> parameter is given as a list of same length
as number of pops, with <code class="docutils literal notranslate"><span class="pre">True</span></code> for frozen
populations at the corresponding index, and <code class="docutils literal notranslate"><span class="pre">False</span></code> for
populations that continue to evolve.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.log">
<span class="sig-name descname"><span class="pre">log</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.log"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.log" title="Link to this definition"></a></dt>
<dd><p>Returns the natural logarithm of the entries of the frequency spectrum.</p>
<p>Only necessary because np.ma.log now fails to propagate extra
attributes after np 1.10.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.marginalize">
<span class="sig-name descname"><span class="pre">marginalize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">over</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_corners</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.marginalize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.marginalize" title="Link to this definition"></a></dt>
<dd><p>Reduced dimensionality spectrum summing over the set of populations
given by <code class="docutils literal notranslate"><span class="pre">over</span></code>.</p>
<p><code class="docutils literal notranslate"><span class="pre">marginalize</span></code> does not act in-place, so the input frequency spectrum
will not be altered.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>over</strong> (<em>list</em><em> of </em><em>integers</em>) – List of axes to sum over. For example (0,2) will marginalize
populations 0 and 2.</p></li>
<li><p><strong>mask_corners</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the fixed bins of the resulting spectrum will be
masked. The default behavior is to mask the corners only if at least one
of the corners of the input frequency spectrum is masked. If either
corner is masked, the output frequency spectrum masks the fixed bins.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.mask_corners">
<span class="sig-name descname"><span class="pre">mask_corners</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.mask_corners"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.mask_corners" title="Link to this definition"></a></dt>
<dd><p>Mask the ‘seen in 0 samples’ and ‘seen in all samples’ entries.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.pi">
<span class="sig-name descname"><span class="pre">pi</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.pi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.pi" title="Link to this definition"></a></dt>
<dd><p>Returns the estimated expected number of pairwise differences between two
chromosomes in the population.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This estimate includes a factor of sample_size / (sample_size - 1)
to make <span class="math notranslate nohighlight">\(\mathbb{E}[\pi] = \theta\)</span>.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.project">
<span class="sig-name descname"><span class="pre">project</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ns</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.project"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.project" title="Link to this definition"></a></dt>
<dd><p>Project to smaller sample size.</p>
<p><code class="docutils literal notranslate"><span class="pre">project</span></code> does <em>not</em> act in-place, so that the input frequency
spectrum is not changed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ns</strong> (<em>list</em><em> of </em><em>integers</em>) – Sample sizes for new spectrum.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.pulse_migrate">
<span class="sig-name descname"><span class="pre">pulse_migrate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx_from</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idx_to</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_from</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proportion</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.pulse_migrate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.pulse_migrate" title="Link to this definition"></a></dt>
<dd><p>Mass migration (pulse admixture) between two existing populations. The
target (destination) population has the same number of lineages in the
output SFS, and the source population has <code class="docutils literal notranslate"><span class="pre">keep_from</span></code> number of lineages
after the pulse event. The proportion is the expected ancestry proportion
in the target population that comes from the source population.</p>
<p>This serves as a wrapper for <code class="docutils literal notranslate"><span class="pre">Manips.admix_inplace</span></code>.</p>
<p>Depending on the proportion and number of lineages, because this is an
approximate operation, we often need a large number of lineages from
the source population to maintain accuracy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>idx_from</strong> (<em>int</em>) – Index of source population.</p></li>
<li><p><strong>idx_to</strong> (<em>int</em>) – Index of targeet population.</p></li>
<li><p><strong>keep_from</strong> (<em>int</em>) – Number of lineages to keep in source population.</p></li>
<li><p><strong>proportion</strong> (<em>float</em>) – Ancestry proportion of source population that migrates
to target population.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.sample">
<span class="sig-name descname"><span class="pre">sample</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.sample"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.sample" title="Link to this definition"></a></dt>
<dd><p>Generate a Poisson-sampled fs from the current one.</p>
<p>Entries where the current fs is masked or 0 will be masked in the
output sampled fs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.scramble_pop_ids">
<span class="sig-name descname"><span class="pre">scramble_pop_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mask_corners</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.scramble_pop_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.scramble_pop_ids" title="Link to this definition"></a></dt>
<dd><p>Spectrum corresponding to scrambling individuals among populations.</p>
<p>This is useful for assessing how diverged populations are.
Essentially, it pools all the individuals represented in the fs and
generates new populations of random individuals (without replacement)
from that pool. If this fs is significantly different from the
original, that implies population structure.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.split">
<span class="sig-name descname"><span class="pre">split</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.split"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.split" title="Link to this definition"></a></dt>
<dd><p>Splits a population in the SFS into two populations, with the extra
population placed at the end. Returns a new frequency spectrum.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>idx</strong> (<em>int</em>) – The index of the population to split.</p></li>
<li><p><strong>n0</strong> (<em>int</em>) – The sample size of the first split population.</p></li>
<li><p><strong>n1</strong> (<em>int</em>) – The sample size of the second split population.</p></li>
<li><p><strong>new_ids</strong> (<em>list</em><em> of </em><em>strings</em><em>, </em><em>optional</em>) – The population IDs of the split populations. Can only be
used if pop_ids are given for the input spectrum.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.swap_axes">
<span class="sig-name descname"><span class="pre">swap_axes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.swap_axes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.swap_axes" title="Link to this definition"></a></dt>
<dd><p>Uses np’s swapaxes function, but also swaps pop_ids as appropriate
if pop_ids are given.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><code class="docutils literal notranslate"><span class="pre">fs.swapaxes(ax1,</span> <span class="pre">ax2)</span></code> will still work, but if population
ids are given, it won’t swap the pop_ids entries as expected.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ax1</strong> (<em>int</em>) – The index of the first population to swap.</p></li>
<li><p><strong>ax2</strong> (<em>int</em>) – The index of the second population to swap.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.theta_L">
<span class="sig-name descname"><span class="pre">theta_L</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.theta_L"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.theta_L" title="Link to this definition"></a></dt>
<dd><p>Returns theta_L as defined by Zeng et al. “Statistical Tests for Detecting
Positive Selection by Utilizing High-Frequency Variants” (2006)
Genetics</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is only sensible for 1-dimensional spectra.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.to_file">
<span class="sig-name descname"><span class="pre">to_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">precision</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">16</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comment_lines</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">foldmaskinfo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.to_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.to_file" title="Link to this definition"></a></dt>
<dd><p>Write frequency spectrum to file.</p>
<p>The file format is:</p>
<ul class="simple">
<li><p>Any number of comment lines beginning with a ‘#’</p></li>
<li><p>A single line containing N integers giving the dimensions of the fs
array. So this line would be ‘5 5 3’ for an SFS that was 5x5x3.
(That would be 4x4x2 <em>samples</em>.)</p></li>
<li><p>On the <em>same line</em>, the string ‘folded’ or ‘unfolded’ denoting the
folding status of the array</p></li>
<li><p>On the <em>same line</em>, optional strings each containing the population
labels in quotes separated by spaces, e.g. “pop 1” “pop 2”</p></li>
<li><p>A single line giving the array elements. The order of elements is
e.g.: fs[0,0,0] fs[0,0,1] fs[0,0,2] … fs[0,1,0] fs[0,1,1] …</p></li>
<li><p>A single line giving the elements of the mask in the same order as
the data line. ‘1’ indicates masked, ‘0’ indicates unmasked.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fid</strong> (<em>string</em>) – string with file name to write to or an open file object.</p></li>
<li><p><strong>precision</strong> (<em>int</em><em>, </em><em>optional</em>) – precision with which to write out entries of the SFS. (They
are formated via %.&lt;p&gt;g, where &lt;p&gt; is the precision.) Defaults to 16.</p></li>
<li><p><strong>comment_lines</strong> (<em>list</em><em> of </em><em>strings</em><em>, </em><em>optional</em>) – list of strings to be used as comment lines in the header
of the output file.</p></li>
<li><p><strong>foldmaskinfo</strong> (<em>bool</em><em>, </em><em>optional</em>) – If False, folding and mask and population label
information will not be saved.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.tofile">
<span class="sig-name descname"><span class="pre">tofile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">precision</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">16</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comment_lines</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">foldmaskinfo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#moments.Spectrum.tofile" title="Link to this definition"></a></dt>
<dd><p>Write frequency spectrum to file.</p>
<p>The file format is:</p>
<ul class="simple">
<li><p>Any number of comment lines beginning with a ‘#’</p></li>
<li><p>A single line containing N integers giving the dimensions of the fs
array. So this line would be ‘5 5 3’ for an SFS that was 5x5x3.
(That would be 4x4x2 <em>samples</em>.)</p></li>
<li><p>On the <em>same line</em>, the string ‘folded’ or ‘unfolded’ denoting the
folding status of the array</p></li>
<li><p>On the <em>same line</em>, optional strings each containing the population
labels in quotes separated by spaces, e.g. “pop 1” “pop 2”</p></li>
<li><p>A single line giving the array elements. The order of elements is
e.g.: fs[0,0,0] fs[0,0,1] fs[0,0,2] … fs[0,1,0] fs[0,1,1] …</p></li>
<li><p>A single line giving the elements of the mask in the same order as
the data line. ‘1’ indicates masked, ‘0’ indicates unmasked.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fid</strong> (<em>string</em>) – string with file name to write to or an open file object.</p></li>
<li><p><strong>precision</strong> (<em>int</em><em>, </em><em>optional</em>) – precision with which to write out entries of the SFS. (They
are formated via %.&lt;p&gt;g, where &lt;p&gt; is the precision.) Defaults to 16.</p></li>
<li><p><strong>comment_lines</strong> (<em>list</em><em> of </em><em>strings</em><em>, </em><em>optional</em>) – list of strings to be used as comment lines in the header
of the output file.</p></li>
<li><p><strong>foldmaskinfo</strong> (<em>bool</em><em>, </em><em>optional</em>) – If False, folding and mask and population label
information will not be saved.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.unfold">
<span class="sig-name descname"><span class="pre">unfold</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.unfold"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.unfold" title="Link to this definition"></a></dt>
<dd><p>Returns an unfolded frequency spectrum.</p>
<p>It is assumed that each state of a SNP is equally likely to be
ancestral.</p>
<p>Unfolding is not done in-place. The return value is a new Spectrum object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.unmask_all">
<span class="sig-name descname"><span class="pre">unmask_all</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.unmask_all"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.unmask_all" title="Link to this definition"></a></dt>
<dd><p>Unmask all entires of the frequency spectrum.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="moments.Spectrum.unmask_corners">
<span class="sig-name descname"><span class="pre">unmask_corners</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Spectrum_mod.html#Spectrum.unmask_corners"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Spectrum.unmask_corners" title="Link to this definition"></a></dt>
<dd><p>Unmask the ‘seen in 0 samples’ and ‘seen in all samples’ entries.</p>
</dd></dl>

</dd></dl>

</section>
<section id="miscellaneous-functions">
<h2>Miscellaneous functions<a class="headerlink" href="#miscellaneous-functions" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="moments.Misc.perturb_params">
<span class="sig-prename descclassname"><span class="pre">moments.Misc.</span></span><span class="sig-name descname"><span class="pre">perturb_params</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">upper_bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Misc.html#perturb_params"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Misc.perturb_params" title="Link to this definition"></a></dt>
<dd><p>Generate a perturbed set of parameters. Each element of params is randomly
perturbed <cite>fold</cite> factors of 2 up or down.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fold</strong> (<em>float</em><em>, </em><em>optional</em>) – Number of factors of 2 to perturb by, defaults to 1.</p></li>
<li><p><strong>lower_bound</strong> (<em>list</em><em> of </em><em>floats</em><em>, </em><em>optional</em>) – If not None, the resulting parameter set is adjusted
to have all value greater than lower_bound.</p></li>
<li><p><strong>upper_bound</strong> (<em>list</em><em> of </em><em>floats</em><em>, </em><em>optional</em>) – If not None, the resulting parameter set is adjusted
to have all value less than upper_bound.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Misc.make_data_dict_vcf">
<span class="sig-prename descclassname"><span class="pre">moments.Misc.</span></span><span class="sig-name descname"><span class="pre">make_data_dict_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vcf_filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">popinfo_filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flanking_info</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[None,</span> <span class="pre">None]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_multiallelic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Misc.html#make_data_dict_vcf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Misc.make_data_dict_vcf" title="Link to this definition"></a></dt>
<dd><p>Parse a VCF file containing genomic sequence information, along with a file
identifying the population of each sample, and store the information in
a properly formatted dictionary.</p>
<p>Each file may be zipped (.zip) or gzipped (.gz). If a file is zipped,
it must be the only file in the archive, and the two files cannot be zipped
together. Both files must be present for the function to work.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vcf_filename</strong> (<em>str</em>) – Name of VCF file to work with. The function currently works
for biallelic SNPs only, so if REF or ALT is anything other
than a single base pair (A, C, T, or G), the allele will be
skipped. Additionally, genotype information must be present
in the FORMAT field GT, and genotype info must be known for
every sample, else the SNP will be skipped. If the ancestral
allele is known it should be specified in INFO field ‘AA’.
Otherwise, it will be set to ‘-‘.</p></li>
<li><p><strong>popinfo_filename</strong> (<em>str</em>) – Name of file containing the population assignments for
each sample in the VCF. If a sample in the VCF file does
not have a corresponding entry in this file, it will be
skipped. See _get_popinfo for information on how this
file must be formatted.</p></li>
<li><p><strong>filter</strong> (<em>bool</em><em>, </em><em>optional</em>) – If set to True, alleles will be skipped if they have not passed
all filters (i.e. either ‘PASS’ or ‘.’ must be present in FILTER column.</p></li>
<li><p><strong>flanking_info</strong> (<em>list</em><em> of </em><em>strings</em><em>, </em><em>optional</em>) – Flanking information for the reference and/or ancestral
allele can be provided as field(s) in the INFO column. To
add this information to the dict, flanking_info should
specify the names of the fields that contain this info as a
list (e.g. [‘RFL’, ‘AFL’].) If context info is given for
only one allele, set the other item in the list to None,
(e.g. [‘RFL’, None]). Information can be provided as a 3
base-pair sequence or 2 base-pair sequence, where the first
base-pair is the one immediately preceding the SNP, and the
last base-pair is the one immediately following the SNP.</p></li>
<li><p><strong>skip_multiallelic</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, only keep biallelic sites, and skip sites that
have more than one ALT allele.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Misc.count_data_dict">
<span class="sig-prename descclassname"><span class="pre">moments.Misc.</span></span><span class="sig-name descname"><span class="pre">count_data_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Misc.html#count_data_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Misc.count_data_dict" title="Link to this definition"></a></dt>
<dd><p>Summarize data in data_dict by mapping SNP configurations to counts.</p>
<p>Returns a dictionary with keys (successful_calls, derived_calls,
polarized) mapping to counts of SNPs. Here successful_calls is a tuple
with the number of good calls per population, derived_calls is a tuple
of derived calls per pop, and polarized indicates whether that SNP was
polarized using an ancestral state.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dict</strong> (<em>data dictionary</em>) – data_dict formatted as in Misc.make_data_dict</p></li>
<li><p><strong>pop_ids</strong> (<em>list</em><em> of </em><em>strings</em>) – IDs of populations to collect data for</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Misc.bootstrap">
<span class="sig-prename descclassname"><span class="pre">moments.Misc.</span></span><span class="sig-name descname"><span class="pre">bootstrap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">projections</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_corners</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polarized</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_boots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Misc.html#bootstrap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Misc.bootstrap" title="Link to this definition"></a></dt>
<dd><p>Use a non-parametric bootstrap on SNP information contained in a dictionary
to generate new data sets. The new data is created by sampling with
replacement from independent units of the original data. These units can
simply be chromosomes, or they can be regions specified in a BED file.</p>
<p>This function either returns a list of all the newly created SFS, or writes
them to disk in a specified directory.</p>
<p>See <a class="reference internal" href="#moments.Spectrum.from_data_dict" title="moments.Spectrum.from_data_dict"><code class="xref py py-func docutils literal notranslate"><span class="pre">moments.Spectrum.from_data_dict()</span></code></a> for more details about the options for
creating spectra.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dict</strong> (<em>dict</em><em> of </em><em>SNP information</em>) – Dictionary containing properly formatted SNP information (i.e.
created using one of the make_data_dict methods).</p></li>
<li><p><strong>pop_ids</strong> (<em>list</em><em> of </em><em>strings</em>) – List of population IDs.</p></li>
<li><p><strong>projections</strong> (<em>list</em><em> of </em><em>ints</em>) – Projection sizes for the given population IDs.</p></li>
<li><p><strong>mask_corners</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, mask the invariant bins of the SFS.</p></li>
<li><p><strong>polarized</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, we assume we know the ancestral allele. If False,
return folded spectra.</p></li>
<li><p><strong>bed_filename</strong> (<em>string as path to bed file</em>) – If None, chromosomes will be used as the units for
resampling. Otherwise, this should be the filename of a BED
file specifying the regions to be used as resampling units.
Chromosome names must be consistent between the BED file and
the data dictionary, or bootstrap will not work. For example,
if an entry in the data dict has ID X_Y, then the value in
in the chromosome field of the BED file must also be X (not
chrX, chromosomeX, etc.).
If the name field is provided in the BED file, then any
regions with the same name will be considered to be part of
the same unit. This may be useful for sampling as one unit a
gene that is located across non-continuous regions.</p></li>
<li><p><strong>num_boots</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of resampled SFS to generate.</p></li>
<li><p><strong>save_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – If None, the SFS are returned as a list. Otherwise this should be
a string specifying the name of a new directory under which all
of the new SFS should be saved.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-moments.Demographics1D">
<span id="demographic-functions"></span><h2>Demographic functions<a class="headerlink" href="#module-moments.Demographics1D" title="Link to this heading"></a></h2>
<p>Single-population demographic models.</p>
<dl class="py function">
<dt class="sig sig-object py" id="moments.Demographics1D.bottlegrowth">
<span class="sig-prename descclassname"><span class="pre">moments.Demographics1D.</span></span><span class="sig-name descname"><span class="pre">bottlegrowth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Demographics1D.html#bottlegrowth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Demographics1D.bottlegrowth" title="Link to this definition"></a></dt>
<dd><p>Instantanous size change followed by exponential growth.</p>
<p>params = (nuB, nuF, T)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>params</strong> – <p>Tuple of length three specifying (nuB, nuF, T).</p>
<ul>
<li><p>nuB: Ratio of population size after instantanous change to ancient
population size.</p></li>
<li><p>nuF: Ratio of contemporary to ancient population size.</p></li>
<li><p>T: Time in the past at which instantaneous change happened and growth began
(in units of 2*Na generations).</p></li>
</ul>
</p></li>
<li><p><strong>ns</strong> – Number of samples in resulting Spectrum.</p></li>
<li><p><strong>pop_ids</strong> – Optional list of length one specifying the population ID.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Demographics1D.growth">
<span class="sig-prename descclassname"><span class="pre">moments.Demographics1D.</span></span><span class="sig-name descname"><span class="pre">growth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Demographics1D.html#growth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Demographics1D.growth" title="Link to this definition"></a></dt>
<dd><p>Exponential growth beginning some time ago.</p>
<p>params = (nu, T)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>params</strong> – <p>Tupe of length two, specifying (nu, t).</p>
<ul>
<li><p>nu: the final population size.</p></li>
<li><p>T: the time in the past at which growth began
(in units of 2*Ne generations).</p></li>
</ul>
</p></li>
<li><p><strong>ns</strong> – Number of samples in resulting Spectrum. Must be a list of
length one.</p></li>
<li><p><strong>pop_ids</strong> – Optional list of length one specifying the population ID.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Demographics1D.snm">
<span class="sig-prename descclassname"><span class="pre">moments.Demographics1D.</span></span><span class="sig-name descname"><span class="pre">snm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Demographics1D.html#snm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Demographics1D.snm" title="Link to this definition"></a></dt>
<dd><p>Standard neutral model with theta=1.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ns</strong> – Number of samples in resulting Spectrum. Must be a list
of length one.</p></li>
<li><p><strong>pop_ids</strong> – Optional list of length one specifying the population ID.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Demographics1D.three_epoch">
<span class="sig-prename descclassname"><span class="pre">moments.Demographics1D.</span></span><span class="sig-name descname"><span class="pre">three_epoch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Demographics1D.html#three_epoch"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Demographics1D.three_epoch" title="Link to this definition"></a></dt>
<dd><p>Three epoch model of constant sizes.</p>
<p>params = (nuB, nuF, TB, TF)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>params</strong> – <p>Tuple of length four specifying (nuB, nuF, TB, TF).</p>
<ul>
<li><p>nuB: Ratio of bottleneck population size to ancient pop size.</p></li>
<li><p>nuF: Ratio of contemporary to ancient pop size.</p></li>
<li><p>TB: Length of bottleneck (in units of 2*Na generations).</p></li>
<li><p>TF: Time since bottleneck recovery (in units of 2*Na generations).</p></li>
</ul>
</p></li>
<li><p><strong>ns</strong> – Number of samples in resulting Spectrum.</p></li>
<li><p><strong>pop_ids</strong> – Optional list of length one specifying the population ID.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Demographics1D.two_epoch">
<span class="sig-prename descclassname"><span class="pre">moments.Demographics1D.</span></span><span class="sig-name descname"><span class="pre">two_epoch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Demographics1D.html#two_epoch"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Demographics1D.two_epoch" title="Link to this definition"></a></dt>
<dd><p>Instantaneous size change some time ago.</p>
<p>params = (nu, T)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>params</strong> – <p>Tuple of length two, specifying (nu, T).</p>
<ul>
<li><p>nu: the ratio of contemporary to ancient population size.</p></li>
<li><p>T: the time in the past at which size change happened
(in units of 2*Ne generations).</p></li>
</ul>
</p></li>
<li><p><strong>ns</strong> – Number of samples in resulting Spectrum. Must be a list of
length one.</p></li>
<li><p><strong>pop_ids</strong> – Optional list of length one specifying the population ID.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<p id="module-moments.Demographics2D">Two-population demographic models.</p>
<dl class="py function">
<dt class="sig sig-object py" id="moments.Demographics2D.IM">
<span class="sig-prename descclassname"><span class="pre">moments.Demographics2D.</span></span><span class="sig-name descname"><span class="pre">IM</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Demographics2D.html#IM"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Demographics2D.IM" title="Link to this definition"></a></dt>
<dd><p>Isolation-with-migration model with exponential pop growth.</p>
<p>params = (s, nu1, nu2, T, m12, m21)</p>
<p>ns = [n1, n2]</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>params</strong> – <p>Tuple of length 6.</p>
<ul>
<li><p>s: Size of pop 1 after split. (Pop 2 has size 1-s.)</p></li>
<li><p>nu1: Final size of pop 1.</p></li>
<li><p>nu2: Final size of pop 2.</p></li>
<li><p>T: Time in the past of split (in units of 2*Na generations)</p></li>
<li><p>m12: Migration from pop 2 to pop 1 (2 * Na * m12)</p></li>
<li><p>m21: Migration from pop 1 to pop 2</p></li>
</ul>
</p></li>
<li><p><strong>ns</strong> – List of population sizes in first and second populations.</p></li>
<li><p><strong>pop_ids</strong> – List of population IDs.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Demographics2D.IM_pre">
<span class="sig-prename descclassname"><span class="pre">moments.Demographics2D.</span></span><span class="sig-name descname"><span class="pre">IM_pre</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Demographics2D.html#IM_pre"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Demographics2D.IM_pre" title="Link to this definition"></a></dt>
<dd><p>params = (nuPre, TPre, s, nu1, nu2, T, m12, m21)</p>
<p>ns = [n1, n2]</p>
<p>Isolation-with-migration model with exponential pop growth and a size change
prior to split.</p>
<ul class="simple">
<li><p>nuPre: Size after first size change</p></li>
<li><p>TPre: Time before split of first size change.</p></li>
<li><p>s: Fraction of nuPre that goes to pop1. (Pop 2 has size nuPre*(1-s).)</p></li>
<li><p>nu1: Final size of pop 1.</p></li>
<li><p>nu2: Final size of pop 2.</p></li>
<li><p>T: Time in the past of split (in units of 2*Na generations)</p></li>
<li><p>m12: Migration from pop 2 to pop 1 (2*Na*m12)</p></li>
<li><p>m21: Migration from pop 1 to pop 2</p></li>
<li><p>n1, n2: Sample sizes of resulting Spectrum.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ns</strong> – List of population sizes in first and second populations.</p></li>
<li><p><strong>pop_ids</strong> – List of population IDs.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Demographics2D.bottlegrowth">
<span class="sig-prename descclassname"><span class="pre">moments.Demographics2D.</span></span><span class="sig-name descname"><span class="pre">bottlegrowth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Demographics2D.html#bottlegrowth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Demographics2D.bottlegrowth" title="Link to this definition"></a></dt>
<dd><p>params = (nuB, nuF, T)</p>
<p>ns = [n1, n2]</p>
<p>Instantanous size change followed by exponential growth with no population
split.</p>
<ul class="simple">
<li><p>nuB: Ratio of population size after instantanous change to ancient
population size</p></li>
<li><p>nuF: Ratio of contempoary to ancient population size</p></li>
<li><p>T: Time in the past at which instantaneous change happened and growth began
(in units of 2*Na generations)</p></li>
<li><p>n1, n2: Sample sizes of resulting Spectrum.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>params</strong> – List of parameters, (nuB, nuF, T).</p></li>
<li><p><strong>ns</strong> – List of population sizes in first and second populations.</p></li>
<li><p><strong>pop_ids</strong> – List of population IDs.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Demographics2D.bottlegrowth_split">
<span class="sig-prename descclassname"><span class="pre">moments.Demographics2D.</span></span><span class="sig-name descname"><span class="pre">bottlegrowth_split</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Demographics2D.html#bottlegrowth_split"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Demographics2D.bottlegrowth_split" title="Link to this definition"></a></dt>
<dd><p>params = (nuB, nuF, T, Ts)</p>
<p>ns = [n1, n2]</p>
<p>Instantanous size change followed by exponential growth then split.</p>
<ul class="simple">
<li><p>nuB: Ratio of population size after instantanous change to ancient
population size</p></li>
<li><p>nuF: Ratio of contempoary to ancient population size</p></li>
<li><p>T: Time in the past at which instantaneous change happened and growth began
(in units of 2*Na generations)</p></li>
<li><p>Ts: Time in the past at which the two populations split.</p></li>
<li><p>n1, n2: Sample sizes of resulting Spectrum.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ns</strong> – List of population sizes in first and second populations.</p></li>
<li><p><strong>pop_ids</strong> – List of population IDs.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Demographics2D.bottlegrowth_split_mig">
<span class="sig-prename descclassname"><span class="pre">moments.Demographics2D.</span></span><span class="sig-name descname"><span class="pre">bottlegrowth_split_mig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Demographics2D.html#bottlegrowth_split_mig"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Demographics2D.bottlegrowth_split_mig" title="Link to this definition"></a></dt>
<dd><p>params = (nuB, nuF, m, T, Ts)
ns = [n1, n2]</p>
<p>Instantanous size change followed by exponential growth then split with
migration.</p>
<ul class="simple">
<li><p>nuB: Ratio of population size after instantanous change to ancient
population size</p></li>
<li><p>nuF: Ratio of contempoary to ancient population size</p></li>
<li><p>m: Migration rate between the two populations (2*Na*m).</p></li>
<li><p>T: Time in the past at which instantaneous change happened and growth began
(in units of 2*Na generations)</p></li>
<li><p>Ts: Time in the past at which the two populations split.</p></li>
<li><p>n1, n2: Sample sizes of resulting Spectrum.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ns</strong> – List of population sizes in first and second populations.</p></li>
<li><p><strong>pop_ids</strong> – List of population IDs.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Demographics2D.snm">
<span class="sig-prename descclassname"><span class="pre">moments.Demographics2D.</span></span><span class="sig-name descname"><span class="pre">snm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Demographics2D.html#snm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Demographics2D.snm" title="Link to this definition"></a></dt>
<dd><p>ns = [n1, n2]</p>
<p>Standard neutral model with a split but no divergence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ns</strong> – List of population sizes in first and second populations.</p></li>
<li><p><strong>pop_ids</strong> – List of population IDs.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Demographics2D.split_mig">
<span class="sig-prename descclassname"><span class="pre">moments.Demographics2D.</span></span><span class="sig-name descname"><span class="pre">split_mig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Demographics2D.html#split_mig"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Demographics2D.split_mig" title="Link to this definition"></a></dt>
<dd><p>Split into two populations of specifed size, with migration.</p>
<p>params = (nu1, nu2, T, m)</p>
<p>ns = [n1, n2]</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>params</strong> – <p>Tuple of length 4.</p>
<ul>
<li><p>nu1: Size of population 1 after split.</p></li>
<li><p>nu2: Size of population 2 after split.</p></li>
<li><p>T: Time in the past of split (in units of 2*Na generations)</p></li>
<li><p>m: Migration rate between populations (2*Na*m)</p></li>
</ul>
</p></li>
<li><p><strong>ns</strong> – List of length two specifying sample sizes n1 and n2.</p></li>
<li><p><strong>pop_ids</strong> – List of population IDs.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<p id="module-moments.Demographics3D">Three-population demographic models.</p>
<dl class="py function">
<dt class="sig sig-object py" id="moments.Demographics3D.out_of_Africa">
<span class="sig-prename descclassname"><span class="pre">moments.Demographics3D.</span></span><span class="sig-name descname"><span class="pre">out_of_Africa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['YRI',</span> <span class="pre">'CEU',</span> <span class="pre">'CHB']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Demographics3D.html#out_of_Africa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Demographics3D.out_of_Africa" title="Link to this definition"></a></dt>
<dd><p>The Gutenkunst et al (2009) out-of-Africa that has been reinferred a
number of times.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>params</strong> (<em>list</em><em> of </em><em>floats</em>) – List of parameters, in the order (nuA, TA, nuB, TB, nuEu0,
nuEuF, nuAs0, nuAsF, TF, mAfB, mAfEu, mAfAs, mEuAs).</p></li>
<li><p><strong>ns</strong> (<em>list</em><em> of </em><em>ints</em>) – List of population sizes in each population, in order given
by <cite>pop_ids</cite>.</p></li>
<li><p><strong>pop_ids</strong> (<em>list</em><em> of </em><em>strings</em><em>, </em><em>optional</em>) – List of population IDs, defaults to [“YRI”, “CEU”, “CHB”].</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="inference-functions">
<h2>Inference functions<a class="headerlink" href="#inference-functions" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="moments.Inference.ll">
<span class="sig-prename descclassname"><span class="pre">moments.Inference.</span></span><span class="sig-name descname"><span class="pre">ll</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Inference.html#ll"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Inference.ll" title="Link to this definition"></a></dt>
<dd><p>The log-likelihood of the data given the model sfs.</p>
<p>Evaluate the log-likelihood of the data given the model. This is based on
Poisson statistics, where the probability of observing k entries in a cell
given that the mean number is given by the model is
<span class="math notranslate nohighlight">\(P(k) = exp(-model) * model^k / k!\)</span>.</p>
<p>Note: If either the model or the data is a masked array, the return ll will
ignore any elements that are masked in <em>either</em> the model or the data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – The model Spectrum object.</p></li>
<li><p><strong>data</strong> – The data Spectrum object, with same size as <code class="docutils literal notranslate"><span class="pre">model</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Inference.ll_multinom">
<span class="sig-prename descclassname"><span class="pre">moments.Inference.</span></span><span class="sig-name descname"><span class="pre">ll_multinom</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Inference.html#ll_multinom"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Inference.ll_multinom" title="Link to this definition"></a></dt>
<dd><p>Log-likelihood of the data given the model, with optimal rescaling.</p>
<p>Evaluate the log-likelihood of the data given the model. This is based on
Poisson statistics, where the probability of observing k entries in a cell
given that the mean number is given by the model is
<span class="math notranslate nohighlight">\(P(k) = exp(-model) * model^k / k!\)</span>.</p>
<p>model is optimally scaled to maximize ll before calculation.</p>
<p>Note: If either the model or the data is a masked array, the return ll will
ignore any elements that are masked in <em>either</em> the model or the data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – The model Spectrum object.</p></li>
<li><p><strong>data</strong> – The data Spectrum object, with same size as <code class="docutils literal notranslate"><span class="pre">model</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Inference.optimal_sfs_scaling">
<span class="sig-prename descclassname"><span class="pre">moments.Inference.</span></span><span class="sig-name descname"><span class="pre">optimal_sfs_scaling</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Inference.html#optimal_sfs_scaling"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Inference.optimal_sfs_scaling" title="Link to this definition"></a></dt>
<dd><p>Optimal multiplicative scaling factor between model and data.</p>
<p>This scaling is based on only those entries that are masked in neither
model nor data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – The model Spectrum object.</p></li>
<li><p><strong>data</strong> – The data Spectrum object, with same size as <code class="docutils literal notranslate"><span class="pre">model</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Inference.optimally_scaled_sfs">
<span class="sig-prename descclassname"><span class="pre">moments.Inference.</span></span><span class="sig-name descname"><span class="pre">optimally_scaled_sfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Inference.html#optimally_scaled_sfs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Inference.optimally_scaled_sfs" title="Link to this definition"></a></dt>
<dd><p>Optimially scale model sfs to data sfs.</p>
<p>Returns a new scaled model sfs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – The model Spectrum object.</p></li>
<li><p><strong>data</strong> – The data Spectrum object, with same size as <code class="docutils literal notranslate"><span class="pre">model</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Inference.linear_Poisson_residual">
<span class="sig-prename descclassname"><span class="pre">moments.Inference.</span></span><span class="sig-name descname"><span class="pre">linear_Poisson_residual</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Inference.html#linear_Poisson_residual"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Inference.linear_Poisson_residual" title="Link to this definition"></a></dt>
<dd><p>Return the Poisson residuals, (model - data)/sqrt(model), of model and data.</p>
<p>mask sets the level in model below which the returned residual array is
masked. The default of 0 excludes values where the residuals are not
defined.</p>
<p>In the limit that the mean of the Poisson distribution is large, these
residuals are normally distributed. (If the mean is small, the Anscombe
residuals are better.)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – The model Spectrum object.</p></li>
<li><p><strong>data</strong> – The data Spectrum object, with same size as <code class="docutils literal notranslate"><span class="pre">model</span></code>.</p></li>
<li><p><strong>mask</strong> – Optional mask, with same size as <code class="docutils literal notranslate"><span class="pre">model</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Inference.Anscombe_Poisson_residual">
<span class="sig-prename descclassname"><span class="pre">moments.Inference.</span></span><span class="sig-name descname"><span class="pre">Anscombe_Poisson_residual</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Inference.html#Anscombe_Poisson_residual"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Inference.Anscombe_Poisson_residual" title="Link to this definition"></a></dt>
<dd><p>Return the Anscombe Poisson residuals between model and data.</p>
<p>mask sets the level in model below which the returned residual array is
masked. This excludes very small values where the residuals are not normal.
1e-2 seems to be a good default for the NIEHS human data. (model = 1e-2,
data = 0, yields a residual of ~1.5.)</p>
<p>Residuals defined in this manner are more normally distributed than the
linear residuals when the mean is small. See this reference below for
justification: Pierce DA and Schafer DW, “Residuals in generalized linear
models” Journal of the American Statistical Association, 81(396)977-986
(1986).</p>
<p>Note that I tried implementing the “adjusted deviance” residuals, but they
always looked very biased for the cases where the data was 0.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – The model Spectrum object.</p></li>
<li><p><strong>data</strong> – The data Spectrum object, with same size as <code class="docutils literal notranslate"><span class="pre">model</span></code>.</p></li>
<li><p><strong>mask</strong> – Optional mask, with same size as <code class="docutils literal notranslate"><span class="pre">model</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Inference.optimize_log">
<span class="sig-prename descclassname"><span class="pre">moments.Inference.</span></span><span class="sig-name descname"><span class="pre">optimize_log</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_func</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">upper_bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flush_delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epsilon</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gtol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multinom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_output</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fixed_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ll_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Inference.html#optimize_log"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Inference.optimize_log" title="Link to this definition"></a></dt>
<dd><p>Optimize log(params) to fit model to data using the BFGS method. This optimization
method works well when we start reasonably close to the optimum.</p>
<p>Because this works in log(params), it cannot explore values of params &lt; 0.
However, it should perform well when parameters range over different orders
of magnitude.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>p0</strong> – Initial parameters.</p></li>
<li><p><strong>data</strong> – Data SFS.</p></li>
<li><p><strong>model_func</strong> – Function to evaluate model spectrum. Should take arguments
<code class="docutils literal notranslate"><span class="pre">model_func(params,</span> <span class="pre">(n1,n2...))</span></code>.</p></li>
<li><p><strong>lower_bound</strong> – Lower bound on parameter values. If not None, must be of same
length as p0.</p></li>
<li><p><strong>upper_bound</strong> – Upper bound on parameter values. If not None, must be of same
length as p0.</p></li>
<li><p><strong>verbose</strong> – If &gt; 0, print optimization status every <code class="docutils literal notranslate"><span class="pre">verbose</span></code> steps.</p></li>
<li><p><strong>output_file</strong> – Stream verbose output into this filename. If None, stream to
standard out.</p></li>
<li><p><strong>flush_delay</strong> – Standard output will be flushed once every &lt;flush_delay&gt;
minutes. This is useful to avoid overloading I/O on clusters.</p></li>
<li><p><strong>epsilon</strong> – Step-size to use for finite-difference derivatives.</p></li>
<li><p><strong>gtol</strong> – Convergence criterion for optimization. For more info,
see help(scipy.optimize.fmin_bfgs)</p></li>
<li><p><strong>multinom</strong> – If True, do a multinomial fit where model is optimially scaled to
data at each step. If False, assume theta is a parameter and do
no scaling.</p></li>
<li><p><strong>maxiter</strong> – Maximum iterations to run for.</p></li>
<li><p><strong>full_output</strong> – If True, return full outputs as in described in
help(scipy.optimize.fmin_bfgs)</p></li>
<li><p><strong>func_args</strong> – Additional arguments to model_func. It is assumed that
model_func’s first argument is an array of parameters to
optimize, that its second argument is an array of sample sizes
for the sfs, and that its last argument is the list of grid
points to use in evaluation.
Using func_args.
For example, you could define your model function as
<code class="docutils literal notranslate"><span class="pre">def</span> <span class="pre">func((p1,p2),</span> <span class="pre">ns,</span> <span class="pre">f1,</span> <span class="pre">f2):</span> <span class="pre">...</span></code>.
If you wanted to fix f1=0.1 and f2=0.2 in the optimization, you
would pass func_args = [0.1,0.2] (and ignore the fixed_params
argument).</p></li>
<li><p><strong>func_kwargs</strong> – Additional keyword arguments to model_func.</p></li>
<li><p><strong>fixed_params</strong> – If not None, should be a list used to fix model parameters at
particular values. For example, if the model parameters
are (nu1,nu2,T,m), then fixed_params = [0.5,None,None,2]
ll hold nu1=0.5 and m=2. The optimizer will only change
T and m. Note that the bounds lists must include all
parameters. Optimization will fail if the fixed values
lie outside their bounds. A full-length p0 should be passed
in; values corresponding to fixed parameters are ignored.
For example, suppose your model function is
<code class="docutils literal notranslate"><span class="pre">def</span> <span class="pre">func((p1,f1,p2,f2),</span> <span class="pre">ns):</span> <span class="pre">...</span></code>
If you wanted to fix f1=0.1 and f2=0.2 in the optimization,
you would pass fixed_params = [None,0.1,None,0.2] (and ignore
the func_args argument).</p></li>
<li><p><strong>ll_scale</strong> – The bfgs algorithm may fail if your initial log-likelihood is
too large. (This appears to be a flaw in the scipy
implementation.) To overcome this, pass ll_scale &gt; 1, which will
simply reduce the magnitude of the log-likelihood. Once in a
region of reasonable likelihood, you’ll probably want to
re-optimize with ll_scale=1.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Inference.optimize_log_fmin">
<span class="sig-prename descclassname"><span class="pre">moments.Inference.</span></span><span class="sig-name descname"><span class="pre">optimize_log_fmin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_func</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">upper_bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flush_delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multinom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxfun</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_output</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fixed_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Inference.html#optimize_log_fmin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Inference.optimize_log_fmin" title="Link to this definition"></a></dt>
<dd><p>Optimize log(params) to fit model to data using Nelder-Mead.
This optimization method may work better than BFGS when far from a
minimum. It is much slower, but more robust, because it doesn’t use
gradient information.</p>
<p>Because this works in log(params), it cannot explore values of params &lt; 0.
It should also perform better when parameters range over large scales.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>p0</strong> – Initial parameters.</p></li>
<li><p><strong>data</strong> – Spectrum with data.</p></li>
<li><p><strong>model_function</strong> – Function to evaluate model spectrum. Should take arguments
(params, (n1,n2…))</p></li>
<li><p><strong>lower_bound</strong> – Lower bound on parameter values. If not None, must be of same
length as p0. A parameter can be declared unbound by assigning
a bound of None.</p></li>
<li><p><strong>upper_bound</strong> – Upper bound on parameter values. If not None, must be of same
length as p0. A parameter can be declared unbound by assigning
a bound of None.</p></li>
<li><p><strong>verbose</strong> – If True, print optimization status every &lt;verbose&gt; steps.</p></li>
<li><p><strong>output_file</strong> – Stream verbose output into this filename. If None, stream to
standard out.</p></li>
<li><p><strong>flush_delay</strong> – Standard output will be flushed once every &lt;flush_delay&gt;
minutes. This is useful to avoid overloading I/O on clusters.</p></li>
<li><p><strong>multinom</strong> – If True, do a multinomial fit where model is optimially scaled to
data at each step. If False, assume theta is a parameter and do
no scaling.</p></li>
<li><p><strong>maxiter</strong> – Maximum number of iterations to run optimization.</p></li>
<li><p><strong>maxfun</strong> – Maximum number of objective function calls to perform.</p></li>
<li><p><strong>full_output</strong> – If True, return full outputs as in described in
help(scipy.optimize.fmin_bfgs)</p></li>
<li><p><strong>func_args</strong> – Additional arguments to model_func. It is assumed that
model_func’s first argument is an array of parameters to
optimize, that its second argument is an array of sample sizes
for the sfs, and that its last argument is the list of grid
points to use in evaluation.</p></li>
<li><p><strong>func_kwargs</strong> – Additional keyword arguments to model_func.</p></li>
<li><p><strong>fixed_params</strong> – If not None, should be a list used to fix model parameters at
particular values. For example, if the model parameters
are (nu1,nu2,T,m), then fixed_params = [0.5,None,None,2]
will hold nu1=0.5 and m=2. The optimizer will only change
T and m. Note that the bounds lists must include all
parameters. Optimization will fail if the fixed values
lie outside their bounds. A full-length p0 should be passed
in; values corresponding to fixed parameters are ignored.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Inference.optimize_log_powell">
<span class="sig-prename descclassname"><span class="pre">moments.Inference.</span></span><span class="sig-name descname"><span class="pre">optimize_log_powell</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_func</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">upper_bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flush_delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multinom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_output</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fixed_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Inference.html#optimize_log_powell"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Inference.optimize_log_powell" title="Link to this definition"></a></dt>
<dd><p>Optimize log(params) to fit model to data using Powell’s conjugate direction method.</p>
<p>This method works without calculating any derivatives, and optimizes along
one direction at a time. May be useful as an initial search for an approximate
solution, followed by further optimization using a gradient optimizer.</p>
<p>Because this works in log(params), it cannot explore values of params &lt; 0.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>p0</strong> – Initial parameters.</p></li>
<li><p><strong>data</strong> – Spectrum with data.</p></li>
<li><p><strong>model_function</strong> – Function to evaluate model spectrum. Should take arguments
(params, (n1,n2…))</p></li>
<li><p><strong>lower_bound</strong> – Lower bound on parameter values. If not None, must be of same
length as p0. A parameter can be declared unbound by assigning
a bound of None.</p></li>
<li><p><strong>upper_bound</strong> – Upper bound on parameter values. If not None, must be of same
length as p0. A parameter can be declared unbound by assigning
a bound of None.</p></li>
<li><p><strong>verbose</strong> – If True, print optimization status every &lt;verbose&gt; steps.
output_file: Stream verbose output into this filename. If None, stream to
standard out.</p></li>
<li><p><strong>flush_delay</strong> – Standard output will be flushed once every &lt;flush_delay&gt;
minutes. This is useful to avoid overloading I/O on clusters.
multinom: If True, do a multinomial fit where model is optimially scaled to
data at each step. If False, assume theta is a parameter and do
no scaling.</p></li>
<li><p><strong>maxiter</strong> – Maximum iterations to run for.</p></li>
<li><p><strong>full_output</strong> – If True, return full outputs as in described in
help(scipy.optimize.fmin_bfgs)</p></li>
<li><p><strong>func_args</strong> – Additional arguments to model_func. It is assumed that
model_func’s first argument is an array of parameters to
optimize, that its second argument is an array of sample sizes
for the sfs, and that its last argument is the list of grid
points to use in evaluation.</p></li>
<li><p><strong>func_kwargs</strong> – Additional keyword arguments to model_func.</p></li>
<li><p><strong>fixed_params</strong> – If not None, should be a list used to fix model parameters at
particular values. For example, if the model parameters
are (nu1,nu2,T,m), then fixed_params = [0.5,None,None,2]
will hold nu1=0.5 and m=2. The optimizer will only change
T and m. Note that the bounds lists must include all
parameters. Optimization will fail if the fixed values
lie outside their bounds. A full-length p0 should be passed
in; values corresponding to fixed parameters are ignored.
(See help(moments.Inference.optimize_log for examples of func_args and
fixed_params usage.)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Inference.optimize_log_lbfgsb">
<span class="sig-prename descclassname"><span class="pre">moments.Inference.</span></span><span class="sig-name descname"><span class="pre">optimize_log_lbfgsb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_func</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">upper_bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flush_delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epsilon</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pgtol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multinom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100000.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_output</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fixed_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ll_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Inference.html#optimize_log_lbfgsb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Inference.optimize_log_lbfgsb" title="Link to this definition"></a></dt>
<dd><p>Optimize log(params) to fit model to data using the L-BFGS-B method.</p>
<p>This optimization method works well when we start reasonably close to the
optimum. It is best at burrowing down a single minimum. This method is
better than optimize_log if the optimum lies at one or more of the
parameter bounds. However, if your optimum is not on the bounds, this
method may be much slower.</p>
<p>Because this works in log(params), it cannot explore values of params &lt; 0.
It should also perform better when parameters range over scales.</p>
<p>The L-BFGS-B method was developed by Ciyou Zhu, Richard Byrd, and Jorge
Nocedal. The algorithm is described in:</p>
<ul class="simple">
<li><p>R. H. Byrd, P. Lu and J. Nocedal. A Limited Memory Algorithm for Bound
Constrained Optimization, (1995), SIAM Journal on Scientific and
Statistical Computing , 16, 5, pp. 1190-1208.</p></li>
<li><p>C. Zhu, R. H. Byrd and J. Nocedal. L-BFGS-B: Algorithm 778: L-BFGS-B,
FORTRAN routines for large scale bound constrained optimization (1997),
ACM Transactions on Mathematical Software, Vol 23, Num. 4, pp. 550-560.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>p0</strong> – Initial parameters.</p></li>
<li><p><strong>data</strong> – Spectrum with data.</p></li>
<li><p><strong>model_function</strong> – Function to evaluate model spectrum. Should take arguments
(params, (n1,n2…))</p></li>
<li><p><strong>lower_bound</strong> – Lower bound on parameter values. If not None, must be of same
length as p0. A parameter can be declared unbound by assigning
a bound of None.</p></li>
<li><p><strong>upper_bound</strong> – Upper bound on parameter values. If not None, must be of same
length as p0. A parameter can be declared unbound by assigning
a bound of None.</p></li>
<li><p><strong>verbose</strong> – If &gt; 0, print optimization status every &lt;verbose&gt; steps.</p></li>
<li><p><strong>output_file</strong> – Stream verbose output into this filename. If None, stream to
standard out.</p></li>
<li><p><strong>flush_delay</strong> – Standard output will be flushed once every &lt;flush_delay&gt;
minutes. This is useful to avoid overloading I/O on clusters.</p></li>
<li><p><strong>epsilon</strong> – Step-size to use for finite-difference derivatives.</p></li>
<li><p><strong>pgtol</strong> – Convergence criterion for optimization. For more info,
see help(scipy.optimize.fmin_l_bfgs_b)</p></li>
<li><p><strong>multinom</strong> – If True, do a multinomial fit where model is optimially scaled to
data at each step. If False, assume theta is a parameter and do
no scaling.</p></li>
<li><p><strong>maxiter</strong> – Maximum algorithm iterations to run.</p></li>
<li><p><strong>full_output</strong> – If True, return full outputs as in described in
help(scipy.optimize.fmin_bfgs)</p></li>
<li><p><strong>func_args</strong> – Additional arguments to model_func. It is assumed that
model_func’s first argument is an array of parameters to
optimize, that its second argument is an array of sample sizes
for the sfs, and that its last argument is the list of grid
points to use in evaluation.</p></li>
<li><p><strong>func_kwargs</strong> – Additional keyword arguments to model_func.</p></li>
<li><p><strong>fixed_params</strong> – If not None, should be a list used to fix model parameters at
particular values. For example, if the model parameters
are (nu1,nu2,T,m), then fixed_params = [0.5,None,None,2]
will hold nu1=0.5 and m=2. The optimizer will only change
T and m. Note that the bounds lists must include all
parameters. Optimization will fail if the fixed values
lie outside their bounds. A full-length p0 should be passed
in; values corresponding to fixed parameters are ignored.</p></li>
<li><p><strong>ll_scale</strong> – The bfgs algorithm may fail if your initial log-likelihood is
too large. (This appears to be a flaw in the scipy
implementation.) To overcome this, pass ll_scale &gt; 1, which will
simply reduce the magnitude of the log-likelihood. Once in a
region of reasonable likelihood, you’ll probably want to
re-optimize with ll_scale=1.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-moments.Godambe">
<span id="uncertainty-functions"></span><span id="sec-sfs-api-plotting"></span><h2>Uncertainty functions<a class="headerlink" href="#module-moments.Godambe" title="Link to this heading"></a></h2>
<p>Parameter uncertainties and likelihood ratio tests using Godambe information.</p>
<dl class="py function">
<dt class="sig sig-object py" id="moments.Godambe.FIM_uncert">
<span class="sig-prename descclassname"><span class="pre">moments.Godambe.</span></span><span class="sig-name descname"><span class="pre">FIM_uncert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">func_ex</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multinom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Godambe.html#FIM_uncert"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Godambe.FIM_uncert" title="Link to this definition"></a></dt>
<dd><p>Parameter uncertainties from Fisher Information Matrix.
Returns standard deviations of parameter values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>func_ex</strong> (<em>demographic model</em>) – Model function</p></li>
<li><p><strong>p0</strong> (<em>list-like</em>) – Best-fit parameters for func_ex</p></li>
<li><p><strong>data</strong> (<em>spectrum object</em>) – Original data frequency spectrum</p></li>
<li><p><strong>log</strong> (<em>bool</em>) – If True, assume log-normal distribution of parameters.
Returned values are then the standard deviations of the <em>logs</em>
of the parameter values, which can be interpreted as relative
parameter uncertainties.</p></li>
<li><p><strong>multinom</strong> (<em>bool</em>) – If True, assume model is defined without an explicit
parameter for theta. Because uncertainty in theta must be accounted
for to get correct uncertainties for other parameters, this function
will automatically consider theta if multinom=True. In that case, the
final entry of the returned uncertainties will correspond to theta.</p></li>
<li><p><strong>eps</strong> (<em>float</em>) – Fractional stepsize to use when taking
finite-difference derivatives</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Godambe.GIM_uncert">
<span class="sig-prename descclassname"><span class="pre">moments.Godambe.</span></span><span class="sig-name descname"><span class="pre">GIM_uncert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">func_ex</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_boot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multinom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_GIM</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Godambe.html#GIM_uncert"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Godambe.GIM_uncert" title="Link to this definition"></a></dt>
<dd><p>Parameter uncertainties from Godambe Information Matrix (GIM).
Returns standard deviations of parameter values. Bootstrap data
is typically generated by splitting the genome into <em>N</em> chunks and
sampling with replacement from those chunks <em>N</em> times.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>func_ex</strong> (<em>demographic model</em>) – Model function</p></li>
<li><p><strong>all_boot</strong> (<em>list</em><em> of </em><em>spectra</em>) – List of bootstrap frequency spectra</p></li>
<li><p><strong>p0</strong> (<em>list-like</em>) – Best-fit parameters for func_ex</p></li>
<li><p><strong>data</strong> (<em>spectrum object</em>) – Original data frequency spectrum</p></li>
<li><p><strong>log</strong> (<em>bool</em>) – If True, assume log-normal distribution of parameters.
Returned values are then the standard deviations of the <em>logs</em>
of the parameter values, which can be interpreted as relative
parameter uncertainties.</p></li>
<li><p><strong>multinom</strong> (<em>bool</em>) – If True, assume model is defined without an explicit
parameter for theta. Because uncertainty in theta must be accounted
for to get correct uncertainties for other parameters, this function
will automatically consider theta if multinom=True. In that case, the
final entry of the returned uncertainties will correspond to
theta.</p></li>
<li><p><strong>eps</strong> (<em>float</em>) – Fractional stepsize to use when taking finite-difference derivatives</p></li>
<li><p><strong>return_GIM</strong> – If True, also return the full GIM.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="moments.Godambe.LRT_adjust">
<span class="sig-prename descclassname"><span class="pre">moments.Godambe.</span></span><span class="sig-name descname"><span class="pre">LRT_adjust</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">func_ex</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_boot</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nested_indices</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multinom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Godambe.html#LRT_adjust"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#moments.Godambe.LRT_adjust" title="Link to this definition"></a></dt>
<dd><p>First-order moment matching adjustment factor for likelihood ratio test.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>func_ex</strong> (<em>demographic model</em>) – Model function for complex model</p></li>
<li><p><strong>all_boot</strong> (<em>list</em><em> of </em><em>spectra</em>) – List of bootstrap frequency spectra</p></li>
<li><p><strong>p0</strong> (<em>list-like</em>) – Best-fit parameters for the simple model, with nested parameter
explicity defined.  Although equal to values for simple model, should
be in a list form that can be taken in by the complex model you’d like
to evaluate.</p></li>
<li><p><strong>data</strong> (<em>spectrum object</em>) – Original data frequency spectrum</p></li>
<li><p><strong>nested_indices</strong> (<em>list</em><em> of </em><em>ints</em>) – List of positions of nested parameters in complex model
parameter list</p></li>
<li><p><strong>multinom</strong> (<em>bool</em>) – If True, assume model is defined without an explicit parameter
for theta. Because uncertainty in theta must be accounted for to get
correct uncertainties for other parameters, this function will
automatically consider theta if multinom=True.</p></li>
<li><p><strong>eps</strong> (<em>float</em>) – Fractional stepsize to use when taking finite-difference derivatives</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="plotting-features">
<h2>Plotting features<a class="headerlink" href="#plotting-features" title="Link to this heading"></a></h2>
<section id="single-population-plotting">
<h3>Single-population plotting<a class="headerlink" href="#single-population-plotting" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">moments.Plotting.</span></span><span class="sig-name descname"><span class="pre">plot_1d_comp_Poisson</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_num</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Anscombe'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_masked</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['Model',</span> <span class="pre">'Data']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Plotting.html#plot_1d_comp_Poisson"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Poisson comparison between 1d model and data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – 1-dimensional model SFS</p></li>
<li><p><strong>data</strong> – 1-dimensional data SFS</p></li>
<li><p><strong>fig_num</strong> – Clear and use figure fig_num for display. If None, an new figure
window is created.</p></li>
<li><p><strong>residual</strong> – ‘Anscombe’ for Anscombe residuals, which are more normally
distributed for Poisson sampling. ‘linear’ for the linear
residuals, which can be less biased.</p></li>
<li><p><strong>plot_masked</strong> – Additionally plots (in open circles) results for points in the
model or data that were masked.</p></li>
<li><p><strong>out</strong> – Output filename to save figure, if given.</p></li>
<li><p><strong>show</strong> – If True, execute pylab.show command to make sure plot displays.</p></li>
<li><p><strong>labels</strong> – A list of strings of length two, labels for the first and second
input frequency spectra. Defaults to “Model” and “Data”.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">moments.Plotting.</span></span><span class="sig-name descname"><span class="pre">plot_1d_comp_multinom</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_num</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Anscombe'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_masked</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['Model',</span> <span class="pre">'Data']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Plotting.html#plot_1d_comp_multinom"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Multinomial comparison between 1d model and data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – 1-dimensional model SFS</p></li>
<li><p><strong>data</strong> – 1-dimensional data SFS</p></li>
<li><p><strong>fig_num</strong> – Clear and use figure fig_num for display. If None, an new figure
window is created.</p></li>
<li><p><strong>residual</strong> – ‘Anscombe’ for Anscombe residuals, which are more normally
distributed for Poisson sampling. ‘linear’ for the linear
residuals, which can be less biased.</p></li>
<li><p><strong>plot_masked</strong> – Additionally plots (in open circles) results for points in the
model or data that were masked.</p></li>
<li><p><strong>out</strong> – Output filename to save figure, if given.</p></li>
<li><p><strong>show</strong> – If True, displays figure. Set to False to supress.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">moments.Plotting.</span></span><span class="sig-name descname"><span class="pre">plot_1d_fs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_num</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Plotting.html#plot_1d_fs"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Plot a 1-dimensional frequency spectrum.</p>
<p>Note that all the plotting is done with pylab. To see additional pylab
methods: “import pylab; help(pylab)”. Pylab’s many functions are documented
at <a class="reference external" href="http://matplotlib.sourceforge.net/contents.html">http://matplotlib.sourceforge.net/contents.html</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fs</strong> – A single-population Spectrum</p></li>
<li><p><strong>fig_num</strong> – If used, clear and use figure fig_num for display.
If None, a new figure window is created.</p></li>
<li><p><strong>show</strong> – If True, execute pylab.show command to make sure plot displays.</p></li>
<li><p><strong>ax</strong> – If None, uses new or specified figure. Otherwise plots in axes object
that is given after clearing.</p></li>
<li><p><strong>out</strong> – If file name is given, saves before showing.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="multi-population-plotting">
<h3>Multi-population plotting<a class="headerlink" href="#multi-population-plotting" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">moments.Plotting.</span></span><span class="sig-name descname"><span class="pre">plot_2d_comp_Poisson</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resid_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_num</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Anscombe'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adjust</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Plotting.html#plot_2d_comp_Poisson"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Poisson comparison between 2d model and data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – 2-dimensional model SFS</p></li>
<li><p><strong>data</strong> – 2-dimensional data SFS</p></li>
<li><p><strong>vmin</strong> – Minimum value plotted.</p></li>
<li><p><strong>vmax</strong> – Maximum value plotted.</p></li>
<li><p><strong>resid_range</strong> – Residual plot saturates at +- resid_range.</p></li>
<li><p><strong>fig_num</strong> – Clear and use figure fig_num for display. If None, an new figure
window is created.</p></li>
<li><p><strong>pop_ids</strong> – If not None, override pop_ids stored in Spectrum.</p></li>
<li><p><strong>residual</strong> – ‘Anscombe’ for Anscombe residuals, which are more normally
distributed for Poisson sampling. ‘linear’ for the linear
residuals, which can be less biased.</p></li>
<li><p><strong>adjust</strong> – Should method use automatic ‘subplots_adjust’? For advanced
manipulation of plots, it may be useful to make this False.</p></li>
<li><p><strong>out</strong> – Output filename to save figure, if given.</p></li>
<li><p><strong>show</strong> – If True, execute pylab.show command to make sure plot displays.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">moments.Plotting.</span></span><span class="sig-name descname"><span class="pre">plot_2d_comp_multinom</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resid_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_num</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Anscombe'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adjust</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Plotting.html#plot_2d_comp_multinom"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Multinomial comparison between 2d model and data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – 2-dimensional model SFS</p></li>
<li><p><strong>data</strong> – 2-dimensional data SFS</p></li>
<li><p><strong>vmin</strong> – Minimum value plotted.</p></li>
<li><p><strong>vmax</strong> – Maximum value plotted.</p></li>
<li><p><strong>resid_range</strong> – Residual plot saturates at +- resid_range.</p></li>
<li><p><strong>fig_num</strong> – Clear and use figure fig_num for display. If None, an new figure
window is created.</p></li>
<li><p><strong>pop_ids</strong> – If not None, override pop_ids stored in Spectrum.</p></li>
<li><p><strong>residual</strong> – ‘Anscombe’ for Anscombe residuals, which are more normally
distributed for Poisson sampling. ‘linear’ for the linear
residuals, which can be less biased.</p></li>
<li><p><strong>adjust</strong> – Should method use automatic ‘subplots_adjust’? For advanced
manipulation of plots, it may be useful to make this False.</p></li>
<li><p><strong>out</strong> – Output filename to save figure, if given.</p></li>
<li><p><strong>show</strong> – If True, execute pylab.show command to make sure plot displays.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">moments.Plotting.</span></span><span class="sig-name descname"><span class="pre">plot_2d_resid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resid_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'neither'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colorbar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Plotting.html#plot_2d_resid"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Linear heatmap of 2d residual array.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sfs</strong> – Residual array to plot.</p></li>
<li><p><strong>resid_range</strong> – Values &gt; resid range or &lt; resid_range saturate the color
spectrum.</p></li>
<li><p><strong>ax</strong> – Axes object to plot into. If None, the result of pylab.gca() is used.</p></li>
<li><p><strong>pop_ids</strong> – If not None, override pop_ids stored in Spectrum.</p></li>
<li><p><strong>extend</strong> – Whether the colorbar should have ‘extension’ arrows. See
help(pylab.colorbar) for more details.</p></li>
<li><p><strong>colorbar</strong> – Should we plot a colorbar?</p></li>
<li><p><strong>out</strong> – Output filename to save figure, if given.</p></li>
<li><p><strong>show</strong> – If True, execute pylab.show command to make sure plot displays.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">moments.Plotting.</span></span><span class="sig-name descname"><span class="pre">plot_3d_comp_Poisson</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resid_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_num</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Anscombe'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adjust</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Plotting.html#plot_3d_comp_Poisson"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Poisson comparison between 3d model and data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – 3-dimensional model SFS</p></li>
<li><p><strong>data</strong> – 3-dimensional data SFS</p></li>
<li><p><strong>vmin</strong> – Minimum value plotted.</p></li>
<li><p><strong>vmax</strong> – Maximum value plotted.</p></li>
<li><p><strong>resid_range</strong> – Residual plot saturates at +- resid_range.</p></li>
<li><p><strong>fig_num</strong> – Clear and use figure fig_num for display. If None, an new figure
window is created.</p></li>
<li><p><strong>pop_ids</strong> – If not None, override pop_ids stored in Spectrum.</p></li>
<li><p><strong>residual</strong> – ‘Anscombe’ for Anscombe residuals, which are more normally
distributed for Poisson sampling. ‘linear’ for the linear
residuals, which can be less biased.</p></li>
<li><p><strong>adjust</strong> – Should method use automatic ‘subplots_adjust’? For advanced
manipulation of plots, it may be useful to make this False.</p></li>
<li><p><strong>out</strong> – Output filename to save figure, if given.</p></li>
<li><p><strong>show</strong> – If True, execute pylab.show command to make sure plot displays.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">moments.Plotting.</span></span><span class="sig-name descname"><span class="pre">plot_3d_comp_multinom</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resid_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_num</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Anscombe'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adjust</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Plotting.html#plot_3d_comp_multinom"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Multinomial comparison between 3d model and data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – 3-dimensional model SFS</p></li>
<li><p><strong>data</strong> – 3-dimensional data SFS</p></li>
<li><p><strong>vmin</strong> – Minimum value plotted.</p></li>
<li><p><strong>vmax</strong> – Maximum value plotted.</p></li>
<li><p><strong>resid_range</strong> – Residual plot saturates at +- resid_range.</p></li>
<li><p><strong>fig_num</strong> – Clear and use figure fig_num for display. If None, an new figure
window is created.</p></li>
<li><p><strong>pop_ids</strong> – If not None, override pop_ids stored in Spectrum.</p></li>
<li><p><strong>residual</strong> – ‘Anscombe’ for Anscombe residuals, which are more normally
distributed for Poisson sampling. ‘linear’ for the linear
residuals, which can be less biased.</p></li>
<li><p><strong>adjust</strong> – Should method use automatic ‘subplots_adjust’? For advanced
manipulation of plots, it may be useful to make this False.</p></li>
<li><p><strong>out</strong> – Output filename to save figure, if given.</p></li>
<li><p><strong>show</strong> – If True, execute pylab.show command to make sure plot displays.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">moments.Plotting.</span></span><span class="sig-name descname"><span class="pre">plot_3d_spectrum</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fignum</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Plotting.html#plot_3d_spectrum"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Logarithmic heatmap of single 3d FS.</p>
<p>Note that this method is slow, because it relies on matplotlib’s software
rendering. For faster and better looking plots, use plot_3d_spectrum_mayavi.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fs</strong> – FS to plot</p></li>
<li><p><strong>vmin</strong> – Values in fs below vmin are masked in plot.</p></li>
<li><p><strong>vmax</strong> – Values in fs above vmax saturate the color spectrum.</p></li>
<li><p><strong>fignum</strong> – Figure number to plot into. If None, a new figure will be created.</p></li>
<li><p><strong>pop_ids</strong> – If not None, override pop_ids stored in Spectrum.</p></li>
<li><p><strong>out</strong> – Output filename to save figure, if given.</p></li>
<li><p><strong>show</strong> – If True, execute pylab.show command to make sure plot displays.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">moments.Plotting.</span></span><span class="sig-name descname"><span class="pre">plot_4d_comp_Poisson</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resid_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_num</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Anscombe'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adjust</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Plotting.html#plot_4d_comp_Poisson"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Poisson comparison between 4d model and data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – 4-dimensional model SFS</p></li>
<li><p><strong>data</strong> – 4-dimensional data SFS</p></li>
<li><p><strong>vmin</strong> – Minimum value plotted.</p></li>
<li><p><strong>vmax</strong> – Maximum value plotted.</p></li>
<li><p><strong>resid_range</strong> – Residual plot saturates at +- resid_range.</p></li>
<li><p><strong>fig_num</strong> – Clear and use figure fig_num for display. If None, an new figure
window is created.</p></li>
<li><p><strong>pop_ids</strong> – If not None, override pop_ids stored in Spectrum.</p></li>
<li><p><strong>residual</strong> – ‘Anscombe’ for Anscombe residuals, which are more normally
distributed for Poisson sampling. ‘linear’ for the linear
residuals, which can be less biased.</p></li>
<li><p><strong>adjust</strong> – Should method use automatic ‘subplots_adjust’? For advanced
manipulation of plots, it may be useful to make this False.</p></li>
<li><p><strong>out</strong> – Output filename to save figure, if given.</p></li>
<li><p><strong>show</strong> – If True, execute pylab.show command to make sure plot displays.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">moments.Plotting.</span></span><span class="sig-name descname"><span class="pre">plot_4d_comp_multinom</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resid_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_num</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Anscombe'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adjust</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Plotting.html#plot_4d_comp_multinom"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Multinomial comparison between 4d model and data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – 4-dimensional model SFS</p></li>
<li><p><strong>data</strong> – 4-dimensional data SFS</p></li>
<li><p><strong>vmin</strong> – Minimum value plotted.</p></li>
<li><p><strong>vmax</strong> – Maximum value plotted.</p></li>
<li><p><strong>resid_range</strong> – Residual plot saturates at +- resid_range.</p></li>
<li><p><strong>fig_num</strong> – Clear and use figure fig_num for display. If None, an new figure
window is created.</p></li>
<li><p><strong>pop_ids</strong> – If not None, override pop_ids stored in Spectrum.</p></li>
<li><p><strong>residual</strong> – ‘Anscombe’ for Anscombe residuals, which are more normally
distributed for Poisson sampling. ‘linear’ for the linear
residuals, which can be less biased.</p></li>
<li><p><strong>adjust</strong> – Should method use automatic ‘subplots_adjust’? For advanced
manipulation of plots, it may be useful to make this False.</p></li>
<li><p><strong>out</strong> – Output filename to save figure, if given.</p></li>
<li><p><strong>show</strong> – If True, execute pylab.show command to make sure plot displays.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">moments.Plotting.</span></span><span class="sig-name descname"><span class="pre">plot_single_2d_sfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sfs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_ids=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend='neither'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colorbar=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap=&lt;matplotlib.colors.LinearSegmentedColormap</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show=True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/moments/Plotting.html#plot_single_2d_sfs"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Heatmap of single 2d SFS.</p>
<p>If vmax is greater than a factor of 10, plot on log scale.</p>
<p>Returns colorbar that is created.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sfs</strong> – SFS to plot</p></li>
<li><p><strong>vmin</strong> – Values in sfs below vmin are masked in plot.</p></li>
<li><p><strong>vmax</strong> – Values in sfs above vmax saturate the color spectrum.</p></li>
<li><p><strong>ax</strong> – Axes object to plot into. If None, the result of pylab.gca() is used.</p></li>
<li><p><strong>pop_ids</strong> – If not None, override pop_ids stored in Spectrum.</p></li>
<li><p><strong>extend</strong> – Whether the colorbar should have ‘extension’ arrows. See
help(pylab.colorbar) for more details.</p></li>
<li><p><strong>colorbar</strong> – Should we plot a colorbar?</p></li>
<li><p><strong>cmap</strong> – Pylab colormap to use for plotting.</p></li>
<li><p><strong>out</strong> – Output filename to save figure, if given.</p></li>
<li><p><strong>show</strong> – If True, execute pylab.show command to make sure plot displays.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../modules/two_locus_selection.html" class="btn btn-neutral float-left" title="Selection at two loci" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="api_ld.html" class="btn btn-neutral float-right" title="API for linkage disequilibrium" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020, Aaron Ragsdale.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>